
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Japan eDNA bootcamp 2025 &#8212; Japan eDNA bootcamp 2025</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'intro';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Setting up the environment" href="setup.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="#">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Japan eDNA bootcamp 2025 - Home"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Japan eDNA bootcamp 2025 - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1 current active">
                <a class="reference internal" href="#">
                    Japan eDNA bootcamp 2025
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setting up the environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinformaticanalysis.html">Bioinformatic analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="taxonomyassignment.html">Taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="datacuration.html">Data curation</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogenetictree.html">Phylogenetic tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="statisticalanalysis.html">Statistical analysis</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fintro.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/intro.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Japan eDNA bootcamp 2025</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#experimental-setup">2. Experimental setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-note-on-software-programs">3. A note on software programs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#general-overview-of-the-bioinformatic-pipeline">4. General overview of the bioinformatic pipeline</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-note-on-scripts-and-reproducible-code">5. A note on scripts and reproducible code</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="japan-edna-bootcamp-2025">
<h1>Japan eDNA bootcamp 2025<a class="headerlink" href="#japan-edna-bootcamp-2025" title="Link to this heading">#</a></h1>
<p>Welcome to the environmental DNA (eDNA) bootcamp held at the University of the Ryukyus!</p>
<p>The primary goal of this course is to introduce you to the analysis of metabarcoding sequence data, with a specific focus on environmental DNA or eDNA, and ultimately make you comfortable to explore your own next-generation sequencing data. The sessions are split up into three days, with the bioinformatic analysis on Wednesday 16 April, taxonomic assignment and data curation on Thursday 17 April, and statistical analysis and visualisations on Friday 18 April. At the end of the three days, we hope you will be able to understand and execute the main steps in a bioinformatic pipeline, build your own custom reference database, assign a taxonomic ID to your sequences using different algorithms, further curate your data, and analyse data in a statistically correct manner.</p>
<section id="introduction">
<h2>1. Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<p>Based on yesterday’s excellent presentations, we can get an understanding of how broadly applicable <a class="reference external" href="https://en.wikipedia.org/wiki/Environmental_DNA#eDNA_metabarcoding">eDNA metabarcoding</a> is and how many research fields are using this technology to answer ecological questions.</p>
<figure class="align-default" id="edna-applications">
<img alt="_images/Applications_of_eDNA_metabarcoding.png" src="_images/Applications_of_eDNA_metabarcoding.png" />
<figcaption>
<p><span class="caption-number">Fig. 1 </span><span class="caption-text">: a general overview of the different applications of eDNA metabarcoding research.</span><a class="headerlink" href="#edna-applications" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Before we get started with coding, we will first introduce the data we will be analysing, some of the software programs available (including the ones we will use in this bootcamp), and the general structure of a bioinformatic pipeline to process metabarcoding data. Finally, we will go over the folder structure we suggest you use during the processing of metabarcoding data to help keep an overview of the files and ensure reproducibility of the analysis.</p>
</section>
<section id="experimental-setup">
<h2>2. Experimental setup<a class="headerlink" href="#experimental-setup" title="Link to this heading">#</a></h2>
<p>The data we will be analyzing is part of a larger eDNA metabarcoding project jointly conducted by the team of Prof. James Reimer, Molecular Invertebrate Systematics and Ecology Laboratory (MISE, University of the Ryukyus) and Prof. Timothy Ravasi, Marine Climate Change Unit (Okinawa Institute of Science and Technology; OIST). One of the goals of this collaborative research is to examine how fish community composition changes across a latitudinal gradient in Japan, with a particular interest in assessing evidence for tropicalization—i.e., the poleward spread of tropical species due to ocean warming.</p>
<p>To investigate this, surface seawater samples were collected from various sites throughout Japan, ranging from temperate to subtropical regions. Samples were taken either from boats or directly from manmade structures such as seawalls and harbors. At each site, three replicate 2-liter surface water samples were collected using a sterilized bucket and transferred into sterile Mighty Bags (Maruemu Corp., Osaka, Japan). To account for potential contamination during sampling and processing, two controls were included per sampling day. A field control was prepared by opening a bottle of sterile water for five minutes at the sampling location, while a filter control was done using the same procedure immediately prior to filtration.</p>
<figure class="align-default" id="edna-sampling">
<img alt="_images/eDNA_sampling.png" src="_images/eDNA_sampling.png" />
<figcaption>
<p><span class="caption-number">Fig. 2 </span><span class="caption-text">: eDNA water sample collection Amami-Oshima, November 2022</span><a class="headerlink" href="#edna-sampling" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Water samples and field controls were filtered using 0.45 µm Sterivex filters (Merck, Darmstadt, Germany) using an Aspirator (As One, Osaka, Japan). Filters were preserved in RNAlater (Sigma-Aldrich, St. Louis, USA) and stored at –80°C until extraction. DNA was extracted following a modified protocol based on the Qiagen DNeasy Blood &amp; Tissue Kit (Açıkbaş et al., 2024). Metabarcoding was carried out using a ~313 bp fragment of the mitochondrial cytochrome c oxidase subunit I (COI) gene, using primers and protocols described by Leray et al. (2013). The resulting data will be used to evaluate spatial biodiversity patterns along Japan’s coastline.</p>
<figure class="align-default" id="edna-sampling-locations">
<img alt="_images/eDNA_locations.png" src="_images/eDNA_locations.png" />
<figcaption>
<p><span class="caption-number">Fig. 3 </span><span class="caption-text">: Sampling locations in Japan</span><a class="headerlink" href="#edna-sampling-locations" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>The subset of samples we will be working with during the workshop were collected during a research trip to Amami-Oshima in November 2022. This expedition was initiated by Megumi Nakano (then with the Nature Conservation Society of Japan) in response to concerns raised by local island residents regarding perceived changes in their surrounding reef ecosystems. The trip aimed to document current biodiversity patterns and provide baseline data that could help assess potential shifts in reef community composition.</p>
<figure class="align-default" id="edna-sampling-locations-subset">
<img alt="_images/eDNA_locations_subset.png" src="_images/eDNA_locations_subset.png" />
<figcaption>
<p><span class="caption-number">Fig. 4 </span><span class="caption-text">: Sampling locations Amami-Ōshima. a) Location of Amami-Ōshima in the island chains of the Ryukyus. b,c) Sample locations of eDNA water samples on Amami-Ōshima</span><a class="headerlink" href="#edna-sampling-locations-subset" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="a-note-on-software-programs">
<h2>3. A note on software programs<a class="headerlink" href="#a-note-on-software-programs" title="Link to this heading">#</a></h2>
<p>The popularity of metabarcoding research, both in the bacterial and eukaryote kingdom, has resulted in the development of a myriad of software packages and pipelines. Furthermore, software packages are continuously updated with new features, as well as novel software programs being designed and published. We will be using several software programs and R packages during this bootcamp. However, we would like to stress that the pipeline used in the workshop is by no means better than other pipelines and we urge you to explore alternative software programs and trial them with your own data after this workshop.</p>
<p>Metabarcoding software programs can generally be split up into two categories, including:</p>
<ol class="arabic simple">
<li><p><em>stand-alone programs that are self-contained</em>: These programs usually incorporate novel functions or alterations on already-existing functions. Programs can be developed for a specific task within the bioinformatic pipeline or can execute multiple or all steps in the bioinformatic pipeline. Examples of such programs are: Mothur, USEARCH, VSEARCH, cutadapt, dada2, OBITools3.</p></li>
<li><p><em>wrappers around existing software</em>: These programs attempt to provide an ecosystem for the user to complete all steps within the bioinformatic pipeline without needing to reformat documents or change the coding language. Examples of such programs are: QIIME2 and JAMP.</p></li>
</ol>
<p>Below, we can find a list of some of the software programs and R packages that are available for analysing metabarcoding data (those marked with a ‘*’ will be covered in this workshop; please note that this list is not exhaustive and many other programs are available):</p>
<ol class="arabic simple">
<li><p>Bioinformatic processing:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a>: assess quality of .fastq files.</p></li>
<li><p><a class="reference external" href="https://multiqc.info">MULTIQC</a>: concatenate FastQC reports to enable comparisons between samples.</p></li>
<li><p><a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a>*: remove adapter regions from sequence data.</p></li>
<li><p><a class="reference external" href="https://github.com/torognes/vsearch">VSEARCH</a>: bioinformatic processing of metabarcoding data.</p></li>
<li><p><a class="reference external" href="https://benjjneb.github.io/dada2/">dada2</a>*: bioinformatic processing of metabarcoding data in R.</p></li>
<li><p><a class="reference external" href="https://docs.qiime2.org/2024.10/">Qiime2</a>: a collection of different programs and packages, all compiled into a single ecosystem for bioinformatic analysis.</p></li>
<li><p><a class="reference external" href="https://mothur.org/">Mothur</a>: one of the first software programs to analyse metabarcoding data.</p></li>
<li><p><a class="reference external" href="https://github.com/VascoElbrecht/JAMP">JAMP</a>: Just Another Metabarcoding Pipeline in R.</p></li>
</ol>
</li>
<li><p>Taxonomy assignment:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://github.com/gjeunen/reference_database_creator">CRABS</a>*: build custom curated reference databases.</p></li>
<li><p><a class="reference external" href="https://www.drive5.com/usearch/manual/sintax_algo.html">SINTAX</a>: k-mer based classifier to assign a taxonomic ID to sequences.</p></li>
<li><p><a class="reference external" href="http://www2.decipher.codes/Classification.html">IDTAXA (RStudio)</a>: machine learning classifier to assign a taxonomic ID to sequences.</p></li>
<li><p><a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a>*: local alignment classifier to assign a taxonomic ID to sequences.</p></li>
<li><p><a class="reference external" href="https://github.com/DominikBuchner/BOLDigger3">BOLDigger3</a>: a Python program to query .fasta files against the <a class="reference external" href="https://www.boldsystems.org">BOLD V5</a> databases.</p></li>
</ol>
</li>
<li><p>Data curation:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://github.com/gjeunen/tombRaider">tombRaider</a>*: an algorithm to identify artefacts from metabarcoding data.</p></li>
<li><p><a class="reference external" href="https://github.com/tobiasgf/lulu">LULU</a>: an R-package for distribution-based post clustering curation of amplicon data.</p></li>
<li><p><a class="reference external" href="https://github.com/donaldtmcknight/microDecon">microDecon</a>: an R-package for removing contamination from metabarcoding datasets post-sequencing.</p></li>
<li><p><a class="reference external" href="https://benjjneb.github.io/decontam/vignettes/decontam_intro.html">Decontam</a>: an R-package for identifying contaminants in marker-gene and metagenomics data.</p></li>
</ol>
</li>
<li><p>Statistical analysis:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://joey711.github.io/phyloseq/">Phyloseq</a>: R package to explore metabarcoding data.</p></li>
<li><p><a class="reference external" href="https://github.com/KaiHsiangHu/iNEXT.3D">iNEXT.3D</a>: R package to calculate inter- and extrapolation of alpha diversity measures.</p></li>
<li><p><a class="reference external" href="https://cran.r-project.org/web/packages/ape/index.html">ape</a>: R package to build multiple sequence alignments and phylogenetic trees.</p></li>
<li><p><a class="reference external" href="https://microbiome.github.io/tutorials/">microbiome</a>: R package as an extension to phyloseq.</p></li>
<li><p><a class="reference external" href="https://cran.r-project.org/web/packages/indicspecies/vignettes/IndicatorSpeciesAnalysis.html">indicspecies</a>: R package to calculate indicator species values</p></li>
<li><p><a class="reference external" href="https://cran.r-project.org/web/packages/vegan/vegan.pdf">vegan</a>: R package to calculate various alpha and beta diversity measures.</p></li>
<li><p><a class="reference external" href="https://yulab-smu.top/treedata-book/index.html">ggtree</a>: build phylogenetic tree graphics.</p></li>
</ol>
</li>
</ol>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Some software programs change the standard structure of your file to be compatible with the functions implemented within the program. When implementing several software programs in your bioinformatic pipeline, such modifications in the file structure could lead to incompatability issues when switching to the next program in your pipeline. It is, therefore, important to understand how a specific program changes the structure of the sequence file and learn how these modifications can be reverted back (through simple python, bash, or R scripts). We will talk a little bit more about file structure and conversion over the coming days.</p>
</div>
<p>In a recent meta-analysis our lab conducted, which included ~600 scientific publications on metabarcoding from 2023, we can see that dada2 and USEARCH-VSEARCH are the most prominently-used software programs in the metabarcoding research community.</p>
<figure class="align-default" id="alluvial-diagram">
<img alt="_images/alluvial_diagram.png" src="_images/alluvial_diagram.png" />
<figcaption>
<p><span class="caption-number">Fig. 5 </span><span class="caption-text">: an alluvial diagram depicting the prevalence of different bioinformatic pipelines.</span><a class="headerlink" href="#alluvial-diagram" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>During this bootcamp, we will be using <a class="reference external" href="https://benjjneb.github.io/dada2/">dada2</a>, the most-frequently used program for the bioinformatic analysis of metabarcoding data.</p>
</section>
<section id="general-overview-of-the-bioinformatic-pipeline">
<h2>4. General overview of the bioinformatic pipeline<a class="headerlink" href="#general-overview-of-the-bioinformatic-pipeline" title="Link to this heading">#</a></h2>
<p>Although software options are numerous, each pipeline follows more or less the same basic steps and accomplishes these steps using similar tools. Before we get started with coding, let’s quickly go over the main steps of the pipeline to give you a clear overview of what we will cover during this bootcamp. For each of these steps, we will provide more information when we cover those sections during the workshop.</p>
<figure class="align-default" id="bioinformatic-pipeline">
<img alt="_images/bioinformatic_pipeline_workflow.png" src="_images/bioinformatic_pipeline_workflow.png" />
<figcaption>
<p><span class="caption-number">Fig. 6 </span><span class="caption-text">: The general workflow of the bioinformatic pipeline. Copyright: <a class="reference external" href="https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.13847">Hakimzadeh <em>et al</em>., 2023.</a></span><a class="headerlink" href="#bioinformatic-pipeline" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="a-note-on-scripts-and-reproducible-code">
<h2>5. A note on scripts and reproducible code<a class="headerlink" href="#a-note-on-scripts-and-reproducible-code" title="Link to this heading">#</a></h2>
<p>One final topic we need to cover before we get started with coding is the use of scripts to execute code. During this workshop, we will mainly be working in RStudio. This environment lends itself to writing code in R scripts, while still remaining interactive by enabling line-by-line code execution. While it is possible to copy-paste the provided commands/code directly in the R console or R Terminal window without using scripts, it is good practice to use scripts for the following three reasons:</p>
<ol class="arabic simple">
<li><p>By using scripts to execute the code, there is a written record of how the data was processed. This will make it easier for you to remember how the data was processed in the future.</p></li>
<li><p>While the sample data provided in this tutorial is small and computing steps take up a maximum of several minutes, processing large data sets can be time consuming. It is, therefore, recommended to process a small portion of the data first to test the code, modify the filenames once everything is set up, and run it on the full data set.</p></li>
<li><p>Scripts can be reused in the future by changing file names, saving you time by not having to constantly write every line of code in the Console or Terminal window. Minimising the amount needing to be written in the Console or Terminal will, also, minimise the risk of receiving error messages due to typo’s, which is the most common error message seen during bioinformatic analysis.</p></li>
</ol>
<p><strong>That is it for the intro, let’s get started with coding!</strong></p>
</section>
<div class="toctree-wrapper compound">
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="right-next"
       href="setup.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Setting up the environment</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#experimental-setup">2. Experimental setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-note-on-software-programs">3. A note on software programs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#general-overview-of-the-bioinformatic-pipeline">4. General overview of the bioinformatic pipeline</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-note-on-scripts-and-reproducible-code">5. A note on scripts and reproducible code</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gert-Jan Jeunen
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025-2030.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>