
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Bioinformatic analysis &#8212; Japan eDNA bootcamp 2025</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'bioinformaticanalysis';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Taxonomy assignment" href="taxonomyassignment.html" />
    <link rel="prev" title="Setting up the environment" href="setup.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Japan eDNA bootcamp 2025 - Home"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Japan eDNA bootcamp 2025 - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Japan eDNA bootcamp 2025
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setting up the environment</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Bioinformatic analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="taxonomyassignment.html">Taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="datacuration.html">Data curation</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogenetictree.html">Phylogenetic tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="statisticalanalysis.html">Statistical analysis</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fbioinformaticanalysis.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/bioinformaticanalysis.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Bioinformatic analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-the-r-environment">1. Setting up the R environment</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exploring-the-starting-files">2. Exploring the starting files</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-metadata-file">2.1 Sample metadata file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-files">2.2 Sequence files</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#data-format">2.2.1 Data format</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#renaming-files">2.2.2 Renaming files</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#initial-data-checks">3. Initial data checks</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#raw-data-quality-scores">3.1 Raw data quality scores</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#read-count-per-sample">3.2 Read count per sample</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#removing-primer-sequences">4. Removing primer sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-single-sample-example">4.1 A single sample example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-trimming">4.2 Batch trimming</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summarising-results">4.3 Summarising results</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-scores-after-primer-trimming">4.3.1 Quality scores after primer trimming</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#reads-retained-after-primer-trimming">4.3.2 Reads retained after primer trimming</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#visualise-read-count-after-primer-trimming">4.3.3 Visualise read count after primer trimming</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-filtering">5. Quality filtering</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-processing">5.1 Sample processing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-results">5.2 Visualising results</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-scores">5.2.1 Quality scores</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#visualise-read-counts">5.2.2 Visualise read counts</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-error-rates">6. Learning error rates</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-asvs">7. Generating ASVs</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#denoising-the-data">7.1 Denoising the data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">7.2 Visualising results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merging-forward-and-reverse-reads">8. Merging forward and reverse reads</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reconstruct-the-full-amplicon">8.1 Reconstruct the full amplicon</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">8.2 Visualising results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-table">9. Count table</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chimera-removal">10. Chimera removal</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#read-count-tracking">11. Read count tracking</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#saving-output-files">12. Saving output files</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="bioinformatic-analysis">
<h1>Bioinformatic analysis<a class="headerlink" href="#bioinformatic-analysis" title="Link to this heading">#</a></h1>
<section id="setting-up-the-r-environment">
<h2>1. Setting up the R environment<a class="headerlink" href="#setting-up-the-r-environment" title="Link to this heading">#</a></h2>
<p>Before beginning our bioinformatic analysis, we need to prepare the R environment by setting the working directory and loading the required <strong>R packages</strong>. While base R provides essential functions, many specialised tools, like those for processing metabarcoding data, are developed by the community and distributed as <strong>packages</strong>. These packages are collections of functions, documentation, and datasets published by researchers to address specific analytical needs. For example, we’ll use packages like <a class="reference external" href="https://benjjneb.github.io/dada2/">dada2</a> for quality filtering and denoising during the bioinformatic pipeline. Installing and loading these packages extends R’s capabilities, leveraging open-source collaboration to make cutting-edge methods accessible. This modularity is one of R’s greatest strengths, enabling reproducible and efficient workflows.</p>
<p>Let’s open a new R script, save it as <code class="docutils literal notranslate"><span class="pre">bioinformatic_pipeline.R</span></code> in the <code class="docutils literal notranslate"><span class="pre">bootcamp</span></code> folder, set the current working directory to the <code class="docutils literal notranslate"><span class="pre">bootcamp</span></code> folder, and, finally, load all necessary packages we will be using during the bioinformatic analysis.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## set up the working environment</span>
<span class="nf">setwd</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="nf">Sys.getenv</span><span class="p">(</span><span class="s">&quot;HOME&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;Desktop&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;bootcamp&quot;</span><span class="p">))</span>
<span class="c1"># package ShortRead needs to be installed, but not loaded</span>
<span class="n">required_packages</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;dada2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ggplot2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;gridExtra&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;dplyr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;readr&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="s">&quot;ggpubr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stringr&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;tidyverse&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;scales&quot;</span><span class="p">)</span>
<span class="nf">invisible</span><span class="p">(</span><span class="nf">sapply</span><span class="p">(</span><span class="n">required_packages</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">pkg</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="n">character.only</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">install.packages</span><span class="p">(</span><span class="n">pkg</span><span class="p">)</span>
<span class="w">    </span><span class="nf">library</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="n">character.only</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span><span class="nf">message</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="s">&quot; version: &quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="n">pkg</span><span class="p">))</span>
<span class="p">}))</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Loading required package: dada2
Loading required package: Rcpp
dada2 version: 1.28.0
Loading required package: ggplot2
ggplot2 version: 3.5.1
Loading required package: gridExtra
gridExtra version: 2.3
Loading required package: dplyr

Attaching package: ‘dplyr’

The following object is masked from ‘package:gridExtra’:

    combine

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

dplyr version: 1.1.4
Loading required package: readr
readr version: 2.1.5
Loading required package: ggpubr
ggpubr version: 0.6.0
Loading required package: stringr
stringr version: 1.5.1
Loading required package: tidyverse
── Attaching core tidyverse packages ────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──
✔ forcats   1.0.0     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ──────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::combine() masks gridExtra::combine()
✖ dplyr::filter()  masks stats::filter()
✖ dplyr::lag()     masks stats::lag()
ℹ Use the conflicted package to force all conflicts to become errors
tidyverse version: 2.0.0
Loading required package: scales

Attaching package: ‘scales’

The following object is masked from ‘package:purrr’:

    discard

The following object is masked from ‘package:readr’:

    col_factor

scales version: 1.3.0
</pre></div>
</div>
</div>
<p>A bit more information about the code block above (as it might look a bit daunting at first):</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">#</span></code>: We can use this symbol to “comment out” code lines. This method is frequently used to provide more information about the code.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">required_packages</span></code>: is a character vector containing all the package names we want to load.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sapply()</span></code>: is a function loop applying a function to each element of <code class="docutils literal notranslate"><span class="pre">required_packages</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">invisible()</span></code>: suppresses extra output from the function.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">!require()</span></code>: tries to load a package (<code class="docutils literal notranslate"><span class="pre">library</span></code>), but if not successfully loaded, the package will be installed first (<code class="docutils literal notranslate"><span class="pre">install.packages</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">message()</span></code>: prints a formatted message for each package.</p></li>
</ol>
<div class="tip dropdown admonition">
<p class="admonition-title">Understanding loops for efficient coding</p>
<p>Writing loops, such as the <code class="docutils literal notranslate"><span class="pre">sapply()</span></code> loop above or for loops, are very common in bioinformatic pipelines, as they avoid repetitive code writing. For example, the <code class="docutils literal notranslate"><span class="pre">sapply()</span></code> loop above can also be written as (without the necessary check for the need to install a package first):</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">dada2</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;dada2&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;ggplot2&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;gridExtra&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;dplyr&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">readr</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;readr&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggpubr</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;ggpubr&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;stringr&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;tidyverse&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">scales</span><span class="p">);</span><span class="w"> </span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;scales&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Loops require a bit of practise to understand and write, as they contain very specific syntax. We recommend practicing often to get the hang of it :)</p>
</div>
<p>Now that we have created a variable (<code class="docutils literal notranslate"><span class="pre">required_packages</span></code>), we can see this appear in the Environment window in the top-right of RStudio.</p>
<figure class="align-default" id="our-first-variable">
<img alt="_images/our_first_variable.png" src="_images/our_first_variable.png" />
<figcaption>
<p><span class="caption-number">Fig. 12 </span><span class="caption-text">: Our first variable in the Environment window</span><a class="headerlink" href="#our-first-variable" title="Link to this image">#</a></p>
</figcaption>
</figure>
<div class="tip dropdown admonition">
<p class="admonition-title">R help documentation</p>
<p>Consulting the R help documentation will help with understanding all aspects of the function. The help documentation for specific functions can be called using the following code:</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="o">?</span><span class="nf">sapply</span><span class="p">()</span>
</pre></div>
</div>
<figure class="align-default" id="r-help-documentation">
<img alt="_images/r_help_documentation.png" src="_images/r_help_documentation.png" />
<figcaption>
<p><span class="caption-number">Fig. 13 </span><span class="caption-text">: Bringing up the R help documentation using the <code class="docutils literal notranslate"><span class="pre">?function()</span></code> command</span><a class="headerlink" href="#r-help-documentation" title="Link to this image">#</a></p>
</figcaption>
</figure>
</div>
</section>
<section id="exploring-the-starting-files">
<h2>2. Exploring the starting files<a class="headerlink" href="#exploring-the-starting-files" title="Link to this heading">#</a></h2>
<section id="sample-metadata-file">
<h3>2.1 Sample metadata file<a class="headerlink" href="#sample-metadata-file" title="Link to this heading">#</a></h3>
<p>Now that we have set the correct working directory and loaded all necessary packages, we can read in some of the starting files and explore their contents. We will start with the metadata file. Note that the code below reads in the metadata file and pipes (<code class="docutils literal notranslate"><span class="pre">%&gt;%</span></code>) the file into a new function (<code class="docutils literal notranslate"><span class="pre">arrange()</span></code>) to sort the table based on the sampleID column.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Read metadata file into memory</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_tsv</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;sample_metadata.txt&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">arrange</span><span class="p">(</span><span class="n">sampleID</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Rows: 45 Columns: 8                                                                                                 
── Column specification ──────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: &quot;\t&quot;
chr (5): sampleID, collectionDate, region, sampleType, sampleLocation
dbl (3): latitude, longitude, replicate

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
</pre></div>
</div>
</div>
<p>The output shows that the metadata table consists of 45 rows and 8 columns. The output also lists the column headers and type of data contained within each column. We can view the metadata dataframe as a spreadsheet in the Source Editor window by clicking on the name of the variable in the Environment window.</p>
<figure class="align-default" id="viewing-dataframes">
<img alt="_images/viewing_dataframes.png" src="_images/viewing_dataframes.png" />
<figcaption>
<p><span class="caption-number">Fig. 14 </span><span class="caption-text">: Viewing dataframes as spreadsheets in the Source Editor window by clicking on the name in the Environment window</span><a class="headerlink" href="#viewing-dataframes" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Now, let’s look at the naming structure of the sample ID’s, as well as the different locations at which the samples were collected. We can select a specific column from an R dataframe by using the <code class="docutils literal notranslate"><span class="pre">$</span></code> symbol. Furthermore, we can use indexing <code class="docutils literal notranslate"><span class="pre">[]</span></code> to specify the number of items. So, let’s print out the first 10 sample ID’s.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Explore and organise data</span>
<span class="c1"># Print sample and file names to Console</span>
<span class="nf">writeLines</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleID</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">])</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; writeLines(metadata$sampleID[1:10])
Akakina_1
Akakina_2
Akakina_3
Asan_1
Asan_2
Asan_3
Ashiken_1
Ashiken_2
Ashiken_3
Edateku_1
</pre></div>
</div>
</div>
<p>From the output, we can see that our sample ID’s are structured as <code class="docutils literal notranslate"><span class="pre">Location</span></code> + <code class="docutils literal notranslate"><span class="pre">_</span></code> + <code class="docutils literal notranslate"><span class="pre">replicate</span> <span class="pre">number</span></code>. It also seems that we have 3 replicates per location, which is a standard replication approach for eDNA metabarcoding surveys.</p>
<p>Finally, let’s print out the location ID’s within <code class="docutils literal notranslate"><span class="pre">metadata</span></code>. Since we have 3 replicates, it will be best to only print the unique names (function: <code class="docutils literal notranslate"><span class="pre">unique()</span></code>), and not repeat each location 3 times.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeLines</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleLocation</span><span class="p">))</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; writeLines(unique(metadata$sampleLocation))
Akakina
Asan
Ashiken
Edateku
Field_control
Filter_control
Gamo
Heda
Imaizaki
Kurasaki
Sakibaru
Sani
Taen
Tatsugo
Yadon
</pre></div>
</div>
</div>
<p>The output shows that we have 15 unique location names, including 2 different control samples (<code class="docutils literal notranslate"><span class="pre">Field_control</span></code> and <code class="docutils literal notranslate"><span class="pre">Filter_control</span></code>). Multiplying 15 (number of locations) times 3 (number of replicates per location) gives us 45 samples, which is the same number as the rows in our metadata file.</p>
</section>
<section id="sequence-files">
<h3>2.2 Sequence files<a class="headerlink" href="#sequence-files" title="Link to this heading">#</a></h3>
<section id="data-format">
<h4>2.2.1 Data format<a class="headerlink" href="#data-format" title="Link to this heading">#</a></h4>
<p>Besides the metadata file, we also have access to our raw sequence files. These are stored in the <code class="docutils literal notranslate"><span class="pre">1.raw</span></code> subdirectory. Since we have 45 samples, let’s count the number of raw sequence files we have.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">length</span><span class="p">(</span><span class="nf">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;001.fastq.gz&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">length</span><span class="p">(</span><span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;1.raw&quot;</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="s2">&quot;001.fastq.gz&quot;</span><span class="p">,</span> <span class="n">full</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="mi">90</span>
</pre></div>
</div>
</div>
<p>We seem to have 90 files, double the number of samples. This tells us two things:</p>
<ol class="arabic simple">
<li><p>Samples are most likely already demultiplexed, i.e., reads have been assigned to their respective samples.</p></li>
<li><p>Samples are most likely sequenced using a paired-end Illumina sequencing strategy.</p></li>
</ol>
<figure class="align-default" id="paired-end-sequencing">
<img alt="_images/paired_end_sequencing.png" src="_images/paired_end_sequencing.png" />
<figcaption>
<p><span class="caption-number">Fig. 15 </span><span class="caption-text">: Illustration of Illumina paired-end sequencing</span><a class="headerlink" href="#paired-end-sequencing" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Let’s verify this by printing the first 10 filenames.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">writeLines</span><span class="p">(</span><span class="nf">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;001.fastq.gz&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">])</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">writeLines</span><span class="p">(</span><span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;1.raw&quot;</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="s2">&quot;001.fastq.gz&quot;</span><span class="p">,</span> <span class="n">full</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">])</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_1_S103_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_1_S103_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_2_S104_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_2_S104_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_3_S105_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_3_S105_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Asan_1_S108_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Asan_1_S108_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Asan_2_S109_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Asan_2_S109_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
<p>Indeed, we seem to have two files, an <code class="docutils literal notranslate"><span class="pre">R1</span></code> (forward) and <code class="docutils literal notranslate"><span class="pre">R2</span></code> (reverse) file, for each sample, <code class="docutils literal notranslate"><span class="pre">Akakina_1</span></code>, <code class="docutils literal notranslate"><span class="pre">Akakina_2</span></code>, <code class="docutils literal notranslate"><span class="pre">Akakina_3</span></code>, etc.</p>
<div class="warning dropdown admonition">
<p class="admonition-title">Data formats</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>The file structure observed for this data is the most common format for metabarcoding data. However, please note that your data format might deviate from this standard and will be dependent on your laboratory protocols and sequencing specifications. Other formats that are frequently observed are:

1. Demultiplexed single end: rather than paired-end, the library was sequenced single end. For this format, you will end up with a single file (`R1`) per sample, rather than two. The bioinformatic pipeline can be followed with minor modifications for this type of data, e.g., merging forward and reverse reads can be omitted.
2. Combined paired end: For this format, you will receive a single forward (`R1`) and reverse (`R2`) file for your full library. This format is usually observed when the library was constructed without following Illumina specifications. You will need to conduct demultiplexing prior to following this bioinformatic pipeline.
3. Combined single end: For this format, you will receive a single forward (`R1`) file for your full library. Besides the need to conduct demultiplexing prior to following this bioinformatic pipeline, minor modifications to the pipeline are necessary, e.g., omit the merging of forward and reverse reads step.

More recently, scientists have been exploring third-generation sequencing approaches (e.g., Oxford Nanopore Technologies) for eDNA and metabarcoding purposes. This type of data, however, requires a completely different bioinformatic pipeline for analysis. Please come have a chat with me after class if you would like some help with this type of sequencing data.
</pre></div>
</div>
</div>
</section>
<section id="renaming-files">
<h4>2.2.2 Renaming files<a class="headerlink" href="#renaming-files" title="Link to this heading">#</a></h4>
<p>During our bioinformatic pipeline, we will use sample ID’s to identify and process sequencing files. For ease-of-use, it would be best if our sequence filenames consist only out of sample IDs plus a suffix that is shared between all forward and all reverse files (don’t worry if this sounds abstract at the moment, it will become clear once we start processing the files). However, our sequence filenames include a unique identifier in between the sample ID and common suffix. When we inspect the previous output, we can see that for sample ID <code class="docutils literal notranslate"><span class="pre">Akakina_1</span></code>, for example, the unique identifier is <code class="docutils literal notranslate"><span class="pre">_S103</span></code>, while the unique identifier is <code class="docutils literal notranslate"><span class="pre">_S104</span></code> for <code class="docutils literal notranslate"><span class="pre">Akakina_2</span></code>. To make our lives a bit easier when processing the files, let’s remove this unique identifier now. We can print the first 10 files again to verify our code worked as expected.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rename sequence files for consistency</span>
<span class="nf">file.rename</span><span class="p">(</span><span class="nf">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;_001.fastq.gz&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span>
<span class="w">            </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_[A-Z]\\d{2,3}&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;_001.fastq.gz&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)))</span>
<span class="c1"># print the first 10 sequence filenames</span>
<span class="nf">writeLines</span><span class="p">(</span><span class="nf">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;001.fastq.gz&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">])</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">writeLines</span><span class="p">(</span><span class="nb">list</span><span class="o">.</span><span class="n">files</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;1.raw&quot;</span><span class="p">,</span> <span class="n">pattern</span> <span class="o">=</span> <span class="s2">&quot;001.fastq.gz&quot;</span><span class="p">,</span> <span class="n">full</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">])</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_1_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_1_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_2_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_2_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_3_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_3_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Asan_1_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Asan_1_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Asan_2_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Asan_2_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
<p>Looking at the output, we can see that our sequence filenames now only contain the sample ID, followed by either <code class="docutils literal notranslate"><span class="pre">_R1_001.fastq.gz</span></code> for the forward or <code class="docutils literal notranslate"><span class="pre">_R2_001.fastq.gz</span></code> for the reverse sequence file. We can, also, check if the sample IDs and sequence filenames match using the <code class="docutils literal notranslate"><span class="pre">setdiff()</span></code> function, i.e., ensure all sample IDs have sequence files and no sequence files exist for which no sample ID metadata is provided.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Determine if sample and file names match</span>
<span class="nf">setdiff</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="nf">str_extract</span><span class="p">(</span><span class="nf">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;R1_001.fastq.gz&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;^[A-za-z_]+_[0-9]+&quot;</span><span class="p">))</span>
<span class="nf">setdiff</span><span class="p">(</span><span class="nf">str_extract</span><span class="p">(</span><span class="nf">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;R1_001.fastq.gz&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;^[A-za-z_]+_[0-9]+&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleID</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; setdiff(metadata$sampleID, str_extract(list.files(path = &quot;1.raw&quot;, pattern = &quot;R1_001.fastq.gz&quot;), &quot;^[A-za-z_]+_[0-9]+&quot;))
character(0)
&gt; setdiff(str_extract(list.files(path = &quot;1.raw&quot;, pattern = &quot;R1_001.fastq.gz&quot;), &quot;^[A-za-z_]+_[0-9]+&quot;), metadata$sampleID)
character(0)
</pre></div>
</div>
</div>
</section>
</section>
</section>
<section id="initial-data-checks">
<h2>3. Initial data checks<a class="headerlink" href="#initial-data-checks" title="Link to this heading">#</a></h2>
<p>When receiving your sequencing files, it is always best to check the data prior to processing to ensure everything is as you expect or within the parameters of your chosen sequencing specifications.</p>
<section id="raw-data-quality-scores">
<h3>3.1 Raw data quality scores<a class="headerlink" href="#raw-data-quality-scores" title="Link to this heading">#</a></h3>
<p>One aspect to check is if the raw data quality scores conform to Illumina standards. We can use the <code class="docutils literal notranslate"><span class="pre">plotQualityProfile()</span></code> function within <a class="reference external" href="https://benjjneb.github.io/dada2/">dada2</a> to explore the quality scores.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step one: check raw sequence files</span>
<span class="c1"># Set variables</span>
<span class="n">raw_forward_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sort</span><span class="p">(</span><span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;R1_001.fastq.gz&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>
<span class="n">raw_reverse_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sort</span><span class="p">(</span><span class="nf">list.files</span><span class="p">(</span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;R2_001.fastq.gz&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span>
<span class="c1"># Determine raw quality scores</span>
<span class="n">forward_raw_qual_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="n">raw_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">aggregate</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Forward Raw Reads&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">theme</span><span class="p">(</span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span>
<span class="n">reverse_raw_qual_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="n">raw_reverse_reads</span><span class="p">,</span><span class="w"> </span><span class="n">aggregate</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Reverse Raw Reads&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">theme</span><span class="p">(</span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span>
<span class="n">raw_qual_plots</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggarrange</span><span class="p">(</span><span class="n">forward_raw_qual_plot</span><span class="p">,</span><span class="w"> </span><span class="n">reverse_raw_qual_plot</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;raw_read_quality_plot_aggregated.pdf&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_qual_plots</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="raw-quality-scores">
<img alt="_images/raw_quality_scores.png" src="_images/raw_quality_scores.png" />
<figcaption>
<p><span class="caption-number">Fig. 16 </span><span class="caption-text">: A plot of the combined raw quality scores of all forward and reverse sequence files (note the standard Illumina pattern of reduced quality at the end of the reads)</span><a class="headerlink" href="#raw-quality-scores" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>For these plots, the cycle number (or read length) is on the x-axis, while the quality score is on the y-axis. The black heatmap shows the frequency of each score at a specific base position. The green line is the mean quality score at a specific base position, while the orange line is the median, and the dashed orange lines depict the quartiles. The red line at the bottom shows the percentage of reads with this length.</p>
<p>For our data, the plots show that we have a total number of 5,626,415 reads for the forward and reverse sequence files, nearly all reads obtaining a length of 300 basepairs (which should match the cycle number of the kit you purchased), and an overall high quality score that conforms to the standard Illumina pattern. The number of total reads does not conform to the Illumina specifications of a V3 2x300 bp Illumina MiSeq kit. However, as this data is a subset of the sequencing run, we do not need to be alarmed.</p>
</section>
<section id="read-count-per-sample">
<h3>3.2 Read count per sample<a class="headerlink" href="#read-count-per-sample" title="Link to this heading">#</a></h3>
<p>Besides investigating the quality scores, total read count, and read length, it is also useful to investigate read count per sample and determine if the forward and reverse sequencing files contain the same number of reads. Determining read count per sample can help identify low read count samples or an uneven distribution of reads across samples, which might indicate the need for resequencing some samples. Determining identical read count for forward and reverse sequencing files could help determine issues during file transfer and data integrity issues.</p>
<p>We will make use of our metadata file and the fact that our sequence filenames contain sample ID information. This link will help us store the read count per sample in new columns in <code class="docutils literal notranslate"><span class="pre">metadata</span></code>, enabling us to visualise results using the <a class="reference external" href="https://ggplot2.tidyverse.org/">ggplot2</a> R package.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Determine raw read count per sample</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">    </span><span class="c1"># Forward reads</span>
<span class="w">    </span><span class="n">raw_reads_f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ShortRead</span><span class="o">::</span><span class="nf">countFastq</span><span class="p">(</span>
<span class="w">      </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;_R1_001.fastq.gz&quot;</span><span class="p">))</span>
<span class="w">    </span><span class="p">)</span><span class="o">$</span><span class="n">records</span><span class="p">,</span>
<span class="w">    </span><span class="c1"># Reverse reads</span>
<span class="w">    </span><span class="n">raw_reads_r</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ShortRead</span><span class="o">::</span><span class="nf">countFastq</span><span class="p">(</span>
<span class="w">      </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;_R2_001.fastq.gz&quot;</span><span class="p">))</span>
<span class="w">    </span><span class="p">)</span><span class="o">$</span><span class="n">records</span>
<span class="w">  </span><span class="p">)</span>

<span class="c1"># Check if forward and reverse have the same number of reads</span>
<span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">nrow</span><span class="p">(</span><span class="n">metadata</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_r</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;\nERROR: Mismatch in&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleID</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span>
<span class="w">        </span><span class="s">&quot;- F:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span>
<span class="w">        </span><span class="s">&quot;R:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_r</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">}</span>

<span class="c1"># Plot raw read count per sample</span>
<span class="n">mean_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">sampleType</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">summarise</span><span class="p">(</span><span class="n">mean_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">mean</span><span class="p">(</span><span class="n">raw_reads_f</span><span class="p">))</span>
<span class="n">type_colors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">scales</span><span class="o">::</span><span class="nf">hue_pal</span><span class="p">()(</span><span class="nf">length</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleType</span><span class="p">)))</span>
<span class="nf">names</span><span class="p">(</span><span class="n">type_colors</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleType</span><span class="p">)</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">reorder</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sampleType</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_hline</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mean_reads</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">yintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mean_reads</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sampleType</span><span class="p">),</span>
<span class="w">    </span><span class="n">linetype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;dashed&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">show.legend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mean_reads</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">-</span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mean_reads</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">0.02</span><span class="p">,</span>
<span class="w">      </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Mean:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">comma</span><span class="p">(</span><span class="nf">round</span><span class="p">(</span><span class="n">mean_reads</span><span class="p">))),</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sampleType</span><span class="p">),</span>
<span class="w">    </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.1</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.5</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">show.legend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">type_colors</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">type_colors</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">comma</span><span class="p">,</span><span class="w"> </span><span class="n">expand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">expansion</span><span class="p">(</span><span class="n">mult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="m">0.15</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Sample ID (ordered by read count)&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Raw Read Count&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w"> </span><span class="n">plot.margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">margin</span><span class="p">(</span><span class="n">r</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">40</span><span class="p">),</span><span class="w"> </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">)</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;per_sample_raw_read_count.png&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">last_plot</span><span class="p">(),</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>

<span class="c1"># Calculate total read count across all samples</span>
<span class="nf">sum</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span>
<span class="nf">sum</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_r</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; sum(metadata$raw_reads_f)
[1] 5626415
&gt; sum(metadata$raw_reads_r)
[1] 5626415
</pre></div>
</div>
</div>
<figure class="align-default" id="raw-per-sample-read-count">
<img alt="_images/per_sample_raw_read_count.png" src="_images/per_sample_raw_read_count.png" />
<figcaption>
<p><span class="caption-number">Fig. 17 </span><span class="caption-text">: A plot showing the per sample read count, filled based on sample type</span><a class="headerlink" href="#raw-per-sample-read-count" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
</section>
<section id="removing-primer-sequences">
<h2>4. Removing primer sequences<a class="headerlink" href="#removing-primer-sequences" title="Link to this heading">#</a></h2>
<p>While our data has been demultiplexed, i.e., barcodes removed and reads assigned to sample IDs, our data still contains the primer sequences. Since these regions are artefacts from the PCR amplification and not biological, they will need to be removed from the reads before continuing with the bioinformatic pipeline. For this library and experiment, we have used the mlCOIintF/jgHCO2198 primer set (<a class="reference external" href="https://frontiersinzoology.biomedcentral.com/articles/10.1186/1742-9994-10-34">Leray et al., 2013</a>). The forward primer corresponds to 5’-GGWACWGGWTGAACWGTWTAYCCYCC-3’ and the reverse primer sequence is 5’-TAIACYTCIGGRTGICCRAARAAYCA-3’. Before batch processing every sample, let’s test our code on a single sample to start with. For primer or adapter removal, we can use the prgram <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a>. Cutadapt, unfortunately, is not an R package, but rather a CLI program written in Python. Normally, this means that we have to run cutadapt through the Terminal window. However, we can make use of the <code class="docutils literal notranslate"><span class="pre">system2()</span></code> function in R, which allows us to run shell commands like cutadapt in an R environment.</p>
<p>Before running cutadapt on a single sample, let’s specify the necessary variables, including where R can find the software program (<code class="docutils literal notranslate"><span class="pre">Sys.which()</span></code>), the primer sequences in a cutadapt format (note that we need to specify the R2 primers in the opposite direction as the R1 primers), and a list of forward and reverse output files.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step two: trim primer sequences</span>
<span class="c1"># Set variables</span>
<span class="n">cutadapt</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">Sys.which</span><span class="p">(</span><span class="s">&quot;cutadapt&quot;</span><span class="p">)</span>
<span class="nf">system2</span><span class="p">(</span><span class="n">cutadapt</span><span class="p">,</span><span class="w"> </span><span class="n">args</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;--version&quot;</span><span class="p">)</span>
<span class="kr">if</span><span class="p">(</span><span class="o">!</span><span class="nf">dir.exists</span><span class="p">(</span><span class="s">&quot;2.trimmed&quot;</span><span class="p">))</span><span class="w"> </span><span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;2.trimmed&quot;</span><span class="p">)</span>
<span class="n">FWD</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;GGWACWGGWTGAACWGTWTAYCCYCC&quot;</span>
<span class="n">REV</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;TANACYTCNGGRTGNCCRAARAAYCA&quot;</span>
<span class="n">FWD_RC</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rc</span><span class="p">(</span><span class="n">FWD</span><span class="p">)</span>
<span class="n">REV_RC</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rc</span><span class="p">(</span><span class="n">REV</span><span class="p">)</span>
<span class="n">R1_primers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;-g &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">FWD</span><span class="p">,</span><span class="w"> </span><span class="s">&quot; -a &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">REV_RC</span><span class="p">)</span>
<span class="n">R2_primers</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;-G &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">REV</span><span class="p">,</span><span class="w"> </span><span class="s">&quot; -A &quot;</span><span class="p">,</span><span class="w"> </span><span class="n">FWD_RC</span><span class="p">)</span>
<span class="n">trimmed_forward_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;2.trimmed&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">basename</span><span class="p">(</span><span class="n">raw_forward_reads</span><span class="p">))</span>
<span class="n">trimmed_reverse_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;2.trimmed&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">basename</span><span class="p">(</span><span class="n">raw_reverse_reads</span><span class="p">))</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">system2</span><span class="p">(</span><span class="n">cutadapt</span><span class="p">,</span> <span class="n">args</span> <span class="o">=</span> <span class="s2">&quot;--version&quot;</span><span class="p">)</span>
<span class="mf">5.0</span>
</pre></div>
</div>
</div>
<section id="a-single-sample-example">
<h3>4.1 A single sample example<a class="headerlink" href="#a-single-sample-example" title="Link to this heading">#</a></h3>
<p>Now that we have all necessary variables loaded, we can test our <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a> code on a single sample. When using the <code class="docutils literal notranslate"><span class="pre">system2()</span></code> function, we specify the program, followed by a list of parameters (<code class="docutils literal notranslate"><span class="pre">args</span> <span class="pre">=</span> <span class="pre">c()</span></code>). For cutadapt, we need to specify which primers to remove. To only keep reds for which both primers were found and removed, we need to specify the <code class="docutils literal notranslate"><span class="pre">--discard-untrimmed</span></code> option. The <code class="docutils literal notranslate"><span class="pre">--no-indels</span></code> and <code class="docutils literal notranslate"><span class="pre">-e</span> <span class="pre">2</span></code> parameters allow us to tell cutadapt to not include insertions and deletions in the search and allow a maximum of 2 errors in the primer sequence. We can specify the <code class="docutils literal notranslate"><span class="pre">--revcomp</span></code> parameter to search for the primer sequences in both directions, while <code class="docutils literal notranslate"><span class="pre">--overlap</span> <span class="pre">15</span></code> requires at least a 15 basepair length match with the primer sequence. Finally, we can use the <code class="docutils literal notranslate"><span class="pre">--cores=0</span></code> parameter to automatically detect the number of available cores.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="nf">system2</span><span class="p">(</span><span class="n">cutadapt</span><span class="p">,</span><span class="w"> </span><span class="n">args</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">R1_primers</span><span class="p">,</span><span class="w"> </span><span class="n">R2_primers</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;--discard-untrimmed&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                           </span><span class="s">&quot;--overlap 15&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;--no-indels&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;-e 2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;--cores=0&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                           </span><span class="s">&quot;-o&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">trimmed_forward_reads</span><span class="p">[</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;-p&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">trimmed_reverse_reads</span><span class="p">[</span><span class="m">1</span><span class="p">],</span><span class="w"> </span>
<span class="w">                           </span><span class="n">raw_forward_reads</span><span class="p">[</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">raw_reverse_reads</span><span class="p">[</span><span class="m">1</span><span class="p">]))</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">system2</span><span class="p">(</span><span class="n">cutadapt</span><span class="p">,</span> <span class="n">args</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">R1_primers</span><span class="p">,</span> <span class="n">R2_primers</span><span class="p">,</span> <span class="s2">&quot;--discard-untrimmed&quot;</span><span class="p">,</span> 
<span class="o">+</span>                            <span class="s2">&quot;--overlap 15&quot;</span><span class="p">,</span> <span class="s2">&quot;--no-indels&quot;</span><span class="p">,</span> <span class="s2">&quot;-e 2&quot;</span><span class="p">,</span> <span class="s2">&quot;--cores=0&quot;</span><span class="p">,</span> 
<span class="o">+</span>                            <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">trimmed_forward_reads</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;-p&quot;</span><span class="p">,</span> <span class="n">trimmed_reverse_reads</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> 
<span class="o">+</span>                            <span class="n">raw_forward_reads</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">raw_reverse_reads</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
<span class="n">This</span> <span class="ow">is</span> <span class="n">cutadapt</span> <span class="mf">5.0</span> <span class="k">with</span> <span class="n">Python</span> <span class="mf">3.11.4</span>
<span class="n">Command</span> <span class="n">line</span> <span class="n">parameters</span><span class="p">:</span> <span class="o">-</span><span class="n">g</span> <span class="n">GGWACWGGWTGAACWGTWTAYCCYCC</span> <span class="o">-</span><span class="n">a</span> <span class="n">TGRTTYTTYGGNCAYCCNGARGTNTA</span> <span class="o">-</span><span class="n">G</span> <span class="n">TANACYTCNGGRTGNCCRAARAAYCA</span> <span class="o">-</span><span class="n">A</span> <span class="n">GGRGGRTAWACWGTTCAWCCWGTWCC</span> <span class="o">--</span><span class="n">discard</span><span class="o">-</span><span class="n">untrimmed</span> <span class="o">--</span><span class="n">overlap</span> <span class="mi">15</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">indels</span> <span class="o">-</span><span class="n">e</span> <span class="mi">2</span> <span class="o">--</span><span class="n">cores</span><span class="o">=</span><span class="mi">0</span> <span class="o">-</span><span class="n">o</span> <span class="mf">2.</span><span class="n">trimmed</span><span class="o">/</span><span class="n">Akakina_1_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">p</span> <span class="mf">2.</span><span class="n">trimmed</span><span class="o">/</span><span class="n">Akakina_1_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_1_R1_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="mf">1.</span><span class="n">raw</span><span class="o">/</span><span class="n">Akakina_1_R2_001</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="n">Processing</span> <span class="n">paired</span><span class="o">-</span><span class="n">end</span> <span class="n">reads</span> <span class="n">on</span> <span class="mi">14</span> <span class="n">cores</span> <span class="o">...</span>

<span class="o">===</span> <span class="n">Summary</span> <span class="o">===</span>

<span class="n">Total</span> <span class="n">read</span> <span class="n">pairs</span> <span class="n">processed</span><span class="p">:</span>             <span class="mi">94</span><span class="p">,</span><span class="mi">059</span>
  <span class="n">Read</span> <span class="mi">1</span> <span class="k">with</span> <span class="n">adapter</span><span class="p">:</span>                  <span class="mi">89</span><span class="p">,</span><span class="mi">697</span> <span class="p">(</span><span class="mf">95.4</span><span class="o">%</span><span class="p">)</span>
  <span class="n">Read</span> <span class="mi">2</span> <span class="k">with</span> <span class="n">adapter</span><span class="p">:</span>                  <span class="mi">86</span><span class="p">,</span><span class="mi">851</span> <span class="p">(</span><span class="mf">92.3</span><span class="o">%</span><span class="p">)</span>

<span class="o">==</span> <span class="n">Read</span> <span class="n">fate</span> <span class="n">breakdown</span> <span class="o">==</span>
<span class="n">Pairs</span> <span class="n">discarded</span> <span class="k">as</span> <span class="n">untrimmed</span><span class="p">:</span>           <span class="mi">11</span><span class="p">,</span><span class="mi">235</span> <span class="p">(</span><span class="mf">11.9</span><span class="o">%</span><span class="p">)</span>
<span class="n">Pairs</span> <span class="n">written</span> <span class="p">(</span><span class="n">passing</span> <span class="n">filters</span><span class="p">):</span>        <span class="mi">82</span><span class="p">,</span><span class="mi">824</span> <span class="p">(</span><span class="mf">88.1</span><span class="o">%</span><span class="p">)</span>

<span class="n">Total</span> <span class="n">basepairs</span> <span class="n">processed</span><span class="p">:</span>    <span class="mi">56</span><span class="p">,</span><span class="mi">591</span><span class="p">,</span><span class="mi">056</span> <span class="n">bp</span>
  <span class="n">Read</span> <span class="mi">1</span><span class="p">:</span>    <span class="mi">28</span><span class="p">,</span><span class="mi">287</span><span class="p">,</span><span class="mi">800</span> <span class="n">bp</span>
  <span class="n">Read</span> <span class="mi">2</span><span class="p">:</span>    <span class="mi">28</span><span class="p">,</span><span class="mi">303</span><span class="p">,</span><span class="mi">256</span> <span class="n">bp</span>
<span class="n">Total</span> <span class="n">written</span> <span class="p">(</span><span class="n">filtered</span><span class="p">):</span>     <span class="mi">45</span><span class="p">,</span><span class="mi">532</span><span class="p">,</span><span class="mi">657</span> <span class="n">bp</span> <span class="p">(</span><span class="mf">80.5</span><span class="o">%</span><span class="p">)</span>
  <span class="n">Read</span> <span class="mi">1</span><span class="p">:</span>    <span class="mi">22</span><span class="p">,</span><span class="mi">758</span><span class="p">,</span><span class="mi">418</span> <span class="n">bp</span>
  <span class="n">Read</span> <span class="mi">2</span><span class="p">:</span>    <span class="mi">22</span><span class="p">,</span><span class="mi">774</span><span class="p">,</span><span class="mi">239</span> <span class="n">bp</span>

<span class="o">===</span> <span class="n">First</span> <span class="n">read</span><span class="p">:</span> <span class="n">Adapter</span> <span class="mi">1</span> <span class="o">===</span>

<span class="n">Sequence</span><span class="p">:</span> <span class="n">GGWACWGGWTGAACWGTWTAYCCYCC</span><span class="p">;</span> <span class="n">Type</span><span class="p">:</span> <span class="n">regular</span> <span class="mi">5</span><span class="s1">&#39;; Length: 26; Trimmed: 89689 times</span>

<span class="n">Minimum</span> <span class="n">overlap</span><span class="p">:</span> <span class="mi">15</span>
<span class="n">No</span><span class="o">.</span> <span class="n">of</span> <span class="n">allowed</span> <span class="n">errors</span><span class="p">:</span>
<span class="mi">1</span><span class="o">-</span><span class="mi">12</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">0</span><span class="p">;</span> <span class="mi">13</span><span class="o">-</span><span class="mi">25</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">1</span><span class="p">;</span> <span class="mi">26</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">2</span>

<span class="n">Overview</span> <span class="n">of</span> <span class="n">removed</span> <span class="n">sequences</span>
<span class="n">length</span>	<span class="n">count</span>	<span class="n">expect</span>	<span class="nb">max</span><span class="o">.</span><span class="n">err</span>	<span class="n">error</span> <span class="n">counts</span>
<span class="mi">15</span>	<span class="mi">4</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">2</span> <span class="mi">2</span>
<span class="mi">16</span>	<span class="mi">3</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">3</span>
<span class="mi">17</span>	<span class="mi">10</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">1</span> <span class="mi">9</span>
<span class="mi">19</span>	<span class="mi">39</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">28</span> <span class="mi">11</span>
<span class="mi">20</span>	<span class="mi">10</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">4</span> <span class="mi">6</span>
<span class="mi">21</span>	<span class="mi">42</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">23</span> <span class="mi">19</span>
<span class="mi">22</span>	<span class="mi">48</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">33</span> <span class="mi">15</span>
<span class="mi">23</span>	<span class="mi">44</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">17</span> <span class="mi">27</span>
<span class="mi">24</span>	<span class="mi">136</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">69</span> <span class="mi">67</span>
<span class="mi">25</span>	<span class="mi">1590</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">767</span> <span class="mi">823</span>
<span class="mi">26</span>	<span class="mi">87619</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">62938</span> <span class="mi">23850</span> <span class="mi">831</span>
<span class="mi">27</span>	<span class="mi">140</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">52</span> <span class="mi">57</span> <span class="mi">31</span>
<span class="mi">28</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span>
<span class="mi">30</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span>
<span class="mi">33</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">0</span> <span class="mi">1</span>
<span class="mi">34</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">0</span> <span class="mi">1</span>


<span class="o">===</span> <span class="n">First</span> <span class="n">read</span><span class="p">:</span> <span class="n">Adapter</span> <span class="mi">2</span> <span class="o">===</span>

<span class="n">Sequence</span><span class="p">:</span> <span class="n">TGRTTYTTYGGNCAYCCNGARGTNTA</span><span class="p">;</span> <span class="n">Type</span><span class="p">:</span> <span class="n">regular</span> <span class="mi">3</span><span class="s1">&#39;; Length: 26; Trimmed: 8 times</span>

<span class="n">Minimum</span> <span class="n">overlap</span><span class="p">:</span> <span class="mi">15</span>
<span class="n">No</span><span class="o">.</span> <span class="n">of</span> <span class="n">allowed</span> <span class="n">errors</span><span class="p">:</span>
<span class="mi">1</span><span class="o">-</span><span class="mi">10</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">0</span><span class="p">;</span> <span class="mi">11</span><span class="o">-</span><span class="mi">22</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">1</span><span class="p">;</span> <span class="mi">23</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">2</span>

<span class="n">Bases</span> <span class="n">preceding</span> <span class="n">removed</span> <span class="n">adapters</span><span class="p">:</span>
  <span class="n">A</span><span class="p">:</span> <span class="mf">0.0</span><span class="o">%</span>
  <span class="n">C</span><span class="p">:</span> <span class="mf">37.5</span><span class="o">%</span>
  <span class="n">G</span><span class="p">:</span> <span class="mf">12.5</span><span class="o">%</span>
  <span class="n">T</span><span class="p">:</span> <span class="mf">50.0</span><span class="o">%</span>
  <span class="n">none</span><span class="o">/</span><span class="n">other</span><span class="p">:</span> <span class="mf">0.0</span><span class="o">%</span>

<span class="n">Overview</span> <span class="n">of</span> <span class="n">removed</span> <span class="n">sequences</span>
<span class="n">length</span>	<span class="n">count</span>	<span class="n">expect</span>	<span class="nb">max</span><span class="o">.</span><span class="n">err</span>	<span class="n">error</span> <span class="n">counts</span>
<span class="mi">26</span>	<span class="mi">8</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">6</span> <span class="mi">2</span>


<span class="o">===</span> <span class="n">Second</span> <span class="n">read</span><span class="p">:</span> <span class="n">Adapter</span> <span class="mi">3</span> <span class="o">===</span>

<span class="n">Sequence</span><span class="p">:</span> <span class="n">TANACYTCNGGRTGNCCRAARAAYCA</span><span class="p">;</span> <span class="n">Type</span><span class="p">:</span> <span class="n">regular</span> <span class="mi">5</span><span class="s1">&#39;; Length: 26; Trimmed: 86836 times</span>

<span class="n">Minimum</span> <span class="n">overlap</span><span class="p">:</span> <span class="mi">15</span>
<span class="n">No</span><span class="o">.</span> <span class="n">of</span> <span class="n">allowed</span> <span class="n">errors</span><span class="p">:</span>
<span class="mi">1</span><span class="o">-</span><span class="mi">10</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">0</span><span class="p">;</span> <span class="mi">11</span><span class="o">-</span><span class="mi">22</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">1</span><span class="p">;</span> <span class="mi">23</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">2</span>

<span class="n">Overview</span> <span class="n">of</span> <span class="n">removed</span> <span class="n">sequences</span>
<span class="n">length</span>	<span class="n">count</span>	<span class="n">expect</span>	<span class="nb">max</span><span class="o">.</span><span class="n">err</span>	<span class="n">error</span> <span class="n">counts</span>
<span class="mi">18</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">0</span> <span class="mi">1</span>
<span class="mi">19</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">0</span> <span class="mi">1</span>
<span class="mi">20</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">1</span>
<span class="mi">21</span>	<span class="mi">10</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">8</span> <span class="mi">2</span>
<span class="mi">22</span>	<span class="mi">5</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">3</span> <span class="mi">2</span>
<span class="mi">23</span>	<span class="mi">29</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">25</span> <span class="mi">4</span>
<span class="mi">24</span>	<span class="mi">81</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">52</span> <span class="mi">29</span>
<span class="mi">25</span>	<span class="mi">1137</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">397</span> <span class="mi">740</span>
<span class="mi">26</span>	<span class="mi">85503</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">83445</span> <span class="mi">1575</span> <span class="mi">483</span>
<span class="mi">27</span>	<span class="mi">67</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">28</span> <span class="mi">12</span> <span class="mi">27</span>
<span class="mi">170</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">1</span>


<span class="o">===</span> <span class="n">Second</span> <span class="n">read</span><span class="p">:</span> <span class="n">Adapter</span> <span class="mi">4</span> <span class="o">===</span>

<span class="n">Sequence</span><span class="p">:</span> <span class="n">GGRGGRTAWACWGTTCAWCCWGTWCC</span><span class="p">;</span> <span class="n">Type</span><span class="p">:</span> <span class="n">regular</span> <span class="mi">3</span><span class="s1">&#39;; Length: 26; Trimmed: 15 times</span>

<span class="n">Minimum</span> <span class="n">overlap</span><span class="p">:</span> <span class="mi">15</span>
<span class="n">No</span><span class="o">.</span> <span class="n">of</span> <span class="n">allowed</span> <span class="n">errors</span><span class="p">:</span>
<span class="mi">1</span><span class="o">-</span><span class="mi">12</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">0</span><span class="p">;</span> <span class="mi">13</span><span class="o">-</span><span class="mi">25</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">1</span><span class="p">;</span> <span class="mi">26</span> <span class="n">bp</span><span class="p">:</span> <span class="mi">2</span>

<span class="n">Bases</span> <span class="n">preceding</span> <span class="n">removed</span> <span class="n">adapters</span><span class="p">:</span>
  <span class="n">A</span><span class="p">:</span> <span class="mf">13.3</span><span class="o">%</span>
  <span class="n">C</span><span class="p">:</span> <span class="mf">26.7</span><span class="o">%</span>
  <span class="n">G</span><span class="p">:</span> <span class="mf">26.7</span><span class="o">%</span>
  <span class="n">T</span><span class="p">:</span> <span class="mf">33.3</span><span class="o">%</span>
  <span class="n">none</span><span class="o">/</span><span class="n">other</span><span class="p">:</span> <span class="mf">0.0</span><span class="o">%</span>

<span class="n">Overview</span> <span class="n">of</span> <span class="n">removed</span> <span class="n">sequences</span>
<span class="n">length</span>	<span class="n">count</span>	<span class="n">expect</span>	<span class="nb">max</span><span class="o">.</span><span class="n">err</span>	<span class="n">error</span> <span class="n">counts</span>
<span class="mi">15</span>	<span class="mi">3</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">2</span> <span class="mi">1</span>
<span class="mi">19</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">1</span>	<span class="mi">1</span>
<span class="mi">26</span>	<span class="mi">9</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">5</span> <span class="mi">4</span>
<span class="mi">47</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">1</span>
<span class="mi">103</span>	<span class="mi">1</span>	<span class="mf">0.0</span>	<span class="mi">2</span>	<span class="mi">0</span> <span class="mi">0</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<p>The cutadapt output provides information on how many reads were analysed, how many reads were trimmed, plus a detailed overview of where the primers were cut in the sequence. For our sample <code class="docutils literal notranslate"><span class="pre">Akakina_1</span></code>, 94,059 reads were processed and 82,824 (88.1%) reads were found to contain the primer sequences. As the options we specified in our command managed to remove the primer sequences from nearly all reads, we will use these settings to process all samples.</p>
<div class="tip dropdown admonition">
<p class="admonition-title">Checking results</p>
<p>We strongly suggest to frequently check results and confirm the expected output, especially when you are trying new code or testing parameters. In this case, for example, we can look at the number of raw reads for sample <code class="docutils literal notranslate"><span class="pre">Akakina_1</span></code> that we have stored in our metadata file (<code class="docutils literal notranslate"><span class="pre">metadata$raw_reads_f[1]</span></code>) and see that this number matches the cutadapt report.</p>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Optimising parameters</p>
<p>While the correct parameters are provided during this workshop for this data set, we suggest you spend quite a bit of time to test different parameter settings for your own data, ensuring optimal results.</p>
</div>
</section>
<section id="batch-trimming">
<h3>4.2 Batch trimming<a class="headerlink" href="#batch-trimming" title="Link to this heading">#</a></h3>
<p>To batch process our cutadapt command and trim primers for all samples, we can use a for loop combined with vector indexing. Before executing the for loop with the cutadapt command, we can run through the loop and print out all the sample and file names, ensuring we point to the correct files at each stage.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">seq_along</span><span class="p">(</span><span class="n">raw_forward_reads</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Sample ID:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleID</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;\nForward In:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">raw_forward_reads</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;\nReverse In:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">raw_reverse_reads</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span>
<span class="w">      </span><span class="s">&quot;\nForward Out:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">trimmed_forward_reads</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;\nReverse Out:&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">trimmed_reverse_reads</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;\n\n&quot;</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; for (i in seq_along(raw_forward_reads)) {
+   cat(&quot;Sample ID:&quot;, metadata$sampleID[i], &quot;\nForward In:&quot;, raw_forward_reads[i], &quot;\nReverse In:&quot;, raw_reverse_reads[i], 
+       &quot;\nForward Out:&quot;, trimmed_forward_reads[i], &quot;\nReverse Out:&quot;, trimmed_reverse_reads[i], &quot;\n\n&quot;)
+ }
Sample ID: Akakina_1 
Forward In: 1.raw/Akakina_1_R1_001.fastq.gz 
Reverse In: 1.raw/Akakina_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Akakina_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Akakina_1_R2_001.fastq.gz 

Sample ID: Akakina_2 
Forward In: 1.raw/Akakina_2_R1_001.fastq.gz 
Reverse In: 1.raw/Akakina_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Akakina_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Akakina_2_R2_001.fastq.gz 

Sample ID: Akakina_3 
Forward In: 1.raw/Akakina_3_R1_001.fastq.gz 
Reverse In: 1.raw/Akakina_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Akakina_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Akakina_3_R2_001.fastq.gz 

Sample ID: Asan_1 
Forward In: 1.raw/Asan_1_R1_001.fastq.gz 
Reverse In: 1.raw/Asan_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Asan_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Asan_1_R2_001.fastq.gz 

Sample ID: Asan_2 
Forward In: 1.raw/Asan_2_R1_001.fastq.gz 
Reverse In: 1.raw/Asan_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Asan_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Asan_2_R2_001.fastq.gz 

Sample ID: Asan_3 
Forward In: 1.raw/Asan_3_R1_001.fastq.gz 
Reverse In: 1.raw/Asan_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Asan_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Asan_3_R2_001.fastq.gz 

Sample ID: Ashiken_1 
Forward In: 1.raw/Ashiken_1_R1_001.fastq.gz 
Reverse In: 1.raw/Ashiken_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Ashiken_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Ashiken_1_R2_001.fastq.gz 

Sample ID: Ashiken_2 
Forward In: 1.raw/Ashiken_2_R1_001.fastq.gz 
Reverse In: 1.raw/Ashiken_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Ashiken_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Ashiken_2_R2_001.fastq.gz 

Sample ID: Ashiken_3 
Forward In: 1.raw/Ashiken_3_R1_001.fastq.gz 
Reverse In: 1.raw/Ashiken_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Ashiken_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Ashiken_3_R2_001.fastq.gz 

Sample ID: Edateku_1 
Forward In: 1.raw/Edateku_1_R1_001.fastq.gz 
Reverse In: 1.raw/Edateku_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Edateku_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Edateku_1_R2_001.fastq.gz 

Sample ID: Edateku_2 
Forward In: 1.raw/Edateku_2_R1_001.fastq.gz 
Reverse In: 1.raw/Edateku_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Edateku_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Edateku_2_R2_001.fastq.gz 

Sample ID: Edateku_3 
Forward In: 1.raw/Edateku_3_R1_001.fastq.gz 
Reverse In: 1.raw/Edateku_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Edateku_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Edateku_3_R2_001.fastq.gz 

Sample ID: FieldControl_112122 
Forward In: 1.raw/FieldControl_112122_R1_001.fastq.gz 
Reverse In: 1.raw/FieldControl_112122_R2_001.fastq.gz 
Forward Out: 2.trimmed/FieldControl_112122_R1_001.fastq.gz 
Reverse Out: 2.trimmed/FieldControl_112122_R2_001.fastq.gz 

Sample ID: FieldControl_112222 
Forward In: 1.raw/FieldControl_112222_R1_001.fastq.gz 
Reverse In: 1.raw/FieldControl_112222_R2_001.fastq.gz 
Forward Out: 2.trimmed/FieldControl_112222_R1_001.fastq.gz 
Reverse Out: 2.trimmed/FieldControl_112222_R2_001.fastq.gz 

Sample ID: FieldControl_112422 
Forward In: 1.raw/FieldControl_112422_R1_001.fastq.gz 
Reverse In: 1.raw/FieldControl_112422_R2_001.fastq.gz 
Forward Out: 2.trimmed/FieldControl_112422_R1_001.fastq.gz 
Reverse Out: 2.trimmed/FieldControl_112422_R2_001.fastq.gz 

Sample ID: FilterControl_112122 
Forward In: 1.raw/FilterControl_112122_R1_001.fastq.gz 
Reverse In: 1.raw/FilterControl_112122_R2_001.fastq.gz 
Forward Out: 2.trimmed/FilterControl_112122_R1_001.fastq.gz 
Reverse Out: 2.trimmed/FilterControl_112122_R2_001.fastq.gz 

Sample ID: FilterControl_112222 
Forward In: 1.raw/FilterControl_112222_R1_001.fastq.gz 
Reverse In: 1.raw/FilterControl_112222_R2_001.fastq.gz 
Forward Out: 2.trimmed/FilterControl_112222_R1_001.fastq.gz 
Reverse Out: 2.trimmed/FilterControl_112222_R2_001.fastq.gz 

Sample ID: FilterControl_112422 
Forward In: 1.raw/FilterControl_112422_R1_001.fastq.gz 
Reverse In: 1.raw/FilterControl_112422_R2_001.fastq.gz 
Forward Out: 2.trimmed/FilterControl_112422_R1_001.fastq.gz 
Reverse Out: 2.trimmed/FilterControl_112422_R2_001.fastq.gz 

Sample ID: Gamo_1 
Forward In: 1.raw/Gamo_1_R1_001.fastq.gz 
Reverse In: 1.raw/Gamo_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Gamo_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Gamo_1_R2_001.fastq.gz 

Sample ID: Gamo_2 
Forward In: 1.raw/Gamo_2_R1_001.fastq.gz 
Reverse In: 1.raw/Gamo_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Gamo_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Gamo_2_R2_001.fastq.gz 

Sample ID: Gamo_3 
Forward In: 1.raw/Gamo_3_R1_001.fastq.gz 
Reverse In: 1.raw/Gamo_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Gamo_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Gamo_3_R2_001.fastq.gz 

Sample ID: Heda_1 
Forward In: 1.raw/Heda_1_R1_001.fastq.gz 
Reverse In: 1.raw/Heda_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Heda_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Heda_1_R2_001.fastq.gz 

Sample ID: Heda_2 
Forward In: 1.raw/Heda_2_R1_001.fastq.gz 
Reverse In: 1.raw/Heda_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Heda_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Heda_2_R2_001.fastq.gz 

Sample ID: Heda_3 
Forward In: 1.raw/Heda_3_R1_001.fastq.gz 
Reverse In: 1.raw/Heda_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Heda_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Heda_3_R2_001.fastq.gz 

Sample ID: Imaizaki_1 
Forward In: 1.raw/Imaizaki_1_R1_001.fastq.gz 
Reverse In: 1.raw/Imaizaki_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Imaizaki_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Imaizaki_1_R2_001.fastq.gz 

Sample ID: Imaizaki_2 
Forward In: 1.raw/Imaizaki_2_R1_001.fastq.gz 
Reverse In: 1.raw/Imaizaki_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Imaizaki_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Imaizaki_2_R2_001.fastq.gz 

Sample ID: Imaizaki_3 
Forward In: 1.raw/Imaizaki_3_R1_001.fastq.gz 
Reverse In: 1.raw/Imaizaki_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Imaizaki_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Imaizaki_3_R2_001.fastq.gz 

Sample ID: Kurasaki_1 
Forward In: 1.raw/Kurasaki_1_R1_001.fastq.gz 
Reverse In: 1.raw/Kurasaki_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Kurasaki_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Kurasaki_1_R2_001.fastq.gz 

Sample ID: Kurasaki_2 
Forward In: 1.raw/Kurasaki_2_R1_001.fastq.gz 
Reverse In: 1.raw/Kurasaki_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Kurasaki_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Kurasaki_2_R2_001.fastq.gz 

Sample ID: Kurasaki_3 
Forward In: 1.raw/Kurasaki_3_R1_001.fastq.gz 
Reverse In: 1.raw/Kurasaki_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Kurasaki_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Kurasaki_3_R2_001.fastq.gz 

Sample ID: Sakibaru_1 
Forward In: 1.raw/Sakibaru_1_R1_001.fastq.gz 
Reverse In: 1.raw/Sakibaru_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Sakibaru_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Sakibaru_1_R2_001.fastq.gz 

Sample ID: Sakibaru_2 
Forward In: 1.raw/Sakibaru_2_R1_001.fastq.gz 
Reverse In: 1.raw/Sakibaru_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Sakibaru_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Sakibaru_2_R2_001.fastq.gz 

Sample ID: Sakibaru_3 
Forward In: 1.raw/Sakibaru_3_R1_001.fastq.gz 
Reverse In: 1.raw/Sakibaru_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Sakibaru_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Sakibaru_3_R2_001.fastq.gz 

Sample ID: Sani_1 
Forward In: 1.raw/Sani_1_R1_001.fastq.gz 
Reverse In: 1.raw/Sani_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Sani_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Sani_1_R2_001.fastq.gz 

Sample ID: Sani_2 
Forward In: 1.raw/Sani_2_R1_001.fastq.gz 
Reverse In: 1.raw/Sani_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Sani_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Sani_2_R2_001.fastq.gz 

Sample ID: Sani_3 
Forward In: 1.raw/Sani_3_R1_001.fastq.gz 
Reverse In: 1.raw/Sani_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Sani_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Sani_3_R2_001.fastq.gz 

Sample ID: Taen_1 
Forward In: 1.raw/Taen_1_R1_001.fastq.gz 
Reverse In: 1.raw/Taen_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Taen_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Taen_1_R2_001.fastq.gz 

Sample ID: Taen_2 
Forward In: 1.raw/Taen_2_R1_001.fastq.gz 
Reverse In: 1.raw/Taen_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Taen_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Taen_2_R2_001.fastq.gz 

Sample ID: Taen_3 
Forward In: 1.raw/Taen_3_R1_001.fastq.gz 
Reverse In: 1.raw/Taen_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Taen_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Taen_3_R2_001.fastq.gz 

Sample ID: Tatsugo_1 
Forward In: 1.raw/Tatsugo_1_R1_001.fastq.gz 
Reverse In: 1.raw/Tatsugo_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Tatsugo_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Tatsugo_1_R2_001.fastq.gz 

Sample ID: Tatsugo_2 
Forward In: 1.raw/Tatsugo_2_R1_001.fastq.gz 
Reverse In: 1.raw/Tatsugo_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Tatsugo_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Tatsugo_2_R2_001.fastq.gz 

Sample ID: Tatsugo_3 
Forward In: 1.raw/Tatsugo_3_R1_001.fastq.gz 
Reverse In: 1.raw/Tatsugo_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Tatsugo_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Tatsugo_3_R2_001.fastq.gz 

Sample ID: Yadon_1 
Forward In: 1.raw/Yadon_1_R1_001.fastq.gz 
Reverse In: 1.raw/Yadon_1_R2_001.fastq.gz 
Forward Out: 2.trimmed/Yadon_1_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Yadon_1_R2_001.fastq.gz 

Sample ID: Yadon_2 
Forward In: 1.raw/Yadon_2_R1_001.fastq.gz 
Reverse In: 1.raw/Yadon_2_R2_001.fastq.gz 
Forward Out: 2.trimmed/Yadon_2_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Yadon_2_R2_001.fastq.gz 

Sample ID: Yadon_3 
Forward In: 1.raw/Yadon_3_R1_001.fastq.gz 
Reverse In: 1.raw/Yadon_3_R2_001.fastq.gz 
Forward Out: 2.trimmed/Yadon_3_R1_001.fastq.gz 
Reverse Out: 2.trimmed/Yadon_3_R2_001.fastq.gz 
</pre></div>
</div>
</div>
<p>That seemed to work perfectly! We managed to loop over the list of raw sequence files, and since our metadata file and other lists are ordered in the same way, we can match the correct files together when passing through the loop. So, if we were to add our cutadapt command from above, but change <code class="docutils literal notranslate"><span class="pre">[1]</span></code> to <code class="docutils literal notranslate"><span class="pre">[i]</span></code>, we will trim off the primer sequences for all samples. However, rather than simply using the cutadapt command, we will add 2 additional things to our for loop. The first is to capture all the cutadapt text output, as it is quite elaborate. We don’t need to print this all out in the Console window. We can accomplish this using the <code class="docutils literal notranslate"><span class="pre">stdout</span> <span class="pre">=</span> <span class="pre">TRUE</span></code> and <code class="docutils literal notranslate"><span class="pre">stderr</span> <span class="pre">=</span> <span class="pre">TRUE</span></code> parameters in combination with the <code class="docutils literal notranslate"><span class="pre">system2()</span></code> function. The second is that we will capture the number of trimmed reads from the cutadapt output and add the number to <code class="docutils literal notranslate"><span class="pre">metadata</span></code>. This will allow us to visualise the number of reads lost during primer trimming in a stacked bar chart, which is what we’ll do afterwards.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span><span class="o">$</span><span class="n">trimmed_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NA</span>
<span class="c1"># Execute cutadapt</span>
<span class="kr">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">seq_along</span><span class="p">(</span><span class="n">raw_forward_reads</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Processing&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;-----------&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;/&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">raw_forward_reads</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;-----------\n&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="n">cutadapt_output</span><span class="w"> </span><span class="o">&lt;-</span><span class="nf">system2</span><span class="p">(</span><span class="n">cutadapt</span><span class="p">,</span><span class="w"> </span><span class="n">args</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">R1_primers</span><span class="p">,</span><span class="w"> </span><span class="n">R2_primers</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;--discard-untrimmed&quot;</span><span class="p">,</span>
<span class="w">                             </span><span class="s">&quot;--overlap 15&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;--no-indels&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;-e 2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;--cores=0&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                             </span><span class="s">&quot;-o&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">trimmed_forward_reads</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;-p&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">trimmed_reverse_reads</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
<span class="w">                             </span><span class="n">raw_forward_reads</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="n">raw_reverse_reads</span><span class="p">[</span><span class="n">i</span><span class="p">]),</span>
<span class="w">          </span><span class="n">stdout</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">stderr</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="w">  </span><span class="c1"># Extract trimmed read counts from output</span>
<span class="w">  </span><span class="n">trimmed_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;,&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">str_extract</span><span class="p">(</span>
<span class="w">    </span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;Pairs written \\(passing filters\\):&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">cutadapt_output</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;\\d[\\d,]*&quot;</span><span class="p">)))</span>
<span class="w">  </span><span class="c1"># Store extracted values in metadata</span>
<span class="w">  </span><span class="n">metadata</span><span class="o">$</span><span class="n">trimmed_reads</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">trimmed_reads</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">trimmed_reads</span><span class="p">,</span><span class="w"> </span><span class="kc">NA</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; for(i in seq_along(raw_forward_reads)) {
+   cat(&quot;Processing&quot;, &quot;-----------&quot;, i, &quot;/&quot;, length(raw_forward_reads), &quot;-----------\n&quot;)
+   cutadapt_output &lt;-system2(cutadapt, args = c(R1_primers, R2_primers, &quot;--discard-untrimmed&quot;,
+                                                &quot;--overlap 15&quot;, &quot;--no-indels&quot;, &quot;-e 2&quot;, &quot;--cores=0&quot;, 
+                                                &quot;-o&quot;, trimmed_forward_reads[i], &quot;-p&quot;, trimmed_reverse_reads[i],
+                                                raw_forward_reads[i], raw_reverse_reads[i]),
+                             stdout = TRUE, stderr = TRUE)
+   # Extract trimmed read counts from output
+   trimmed_reads &lt;- as.numeric(gsub(&quot;,&quot;, &quot;&quot;, str_extract(
+     grep(&quot;Pairs written \\(passing filters\\):&quot;, cutadapt_output, value = TRUE), &quot;\\d[\\d,]*&quot;)))
+   # Store extracted values in metadata
+   metadata$trimmed_reads[i] &lt;- ifelse(length(trimmed_reads) &gt; 0, trimmed_reads, NA)
+ }
Processing ----------- 1 / 45 -----------
Processing ----------- 2 / 45 -----------
Processing ----------- 3 / 45 -----------
Processing ----------- 4 / 45 -----------
Processing ----------- 5 / 45 -----------
Processing ----------- 6 / 45 -----------
Processing ----------- 7 / 45 -----------
Processing ----------- 8 / 45 -----------
Processing ----------- 9 / 45 -----------
Processing ----------- 10 / 45 -----------
Processing ----------- 11 / 45 -----------
Processing ----------- 12 / 45 -----------
Processing ----------- 13 / 45 -----------
Processing ----------- 14 / 45 -----------
Processing ----------- 15 / 45 -----------
Processing ----------- 16 / 45 -----------
Processing ----------- 17 / 45 -----------
Processing ----------- 18 / 45 -----------
Processing ----------- 19 / 45 -----------
Processing ----------- 20 / 45 -----------
Processing ----------- 21 / 45 -----------
Processing ----------- 22 / 45 -----------
Processing ----------- 23 / 45 -----------
Processing ----------- 24 / 45 -----------
Processing ----------- 25 / 45 -----------
Processing ----------- 26 / 45 -----------
Processing ----------- 27 / 45 -----------
Processing ----------- 28 / 45 -----------
Processing ----------- 29 / 45 -----------
Processing ----------- 30 / 45 -----------
Processing ----------- 31 / 45 -----------
Processing ----------- 32 / 45 -----------
Processing ----------- 33 / 45 -----------
Processing ----------- 34 / 45 -----------
Processing ----------- 35 / 45 -----------
Processing ----------- 36 / 45 -----------
Processing ----------- 37 / 45 -----------
Processing ----------- 38 / 45 -----------
Processing ----------- 39 / 45 -----------
Processing ----------- 40 / 45 -----------
Processing ----------- 41 / 45 -----------
Processing ----------- 42 / 45 -----------
Processing ----------- 43 / 45 -----------
Processing ----------- 44 / 45 -----------
Processing ----------- 45 / 45 -----------
</pre></div>
</div>
</div>
</section>
<section id="summarising-results">
<h3>4.3 Summarising results<a class="headerlink" href="#summarising-results" title="Link to this heading">#</a></h3>
<section id="quality-scores-after-primer-trimming">
<h4>4.3.1 Quality scores after primer trimming<a class="headerlink" href="#quality-scores-after-primer-trimming" title="Link to this heading">#</a></h4>
<p>One of the first things we can do is to run the <code class="docutils literal notranslate"><span class="pre">plotQualityProfile()</span></code> function again to determine differences in quality scores and read length before and after primer trimming.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step three: check and summarise cutadapt results</span>
<span class="c1"># Check quality scores after primer trimming</span>
<span class="n">forward_trimmed_qual_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="n">trimmed_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">aggregate</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Forward Trimmed Reads&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">theme</span><span class="p">(</span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span>
<span class="n">reverse_trimmed_qual_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="n">trimmed_reverse_reads</span><span class="p">,</span><span class="w"> </span><span class="n">aggregate</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Reverse Trimmed Reads&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">theme</span><span class="p">(</span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span>
<span class="n">trimmed_qual_plots</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggarrange</span><span class="p">(</span><span class="n">forward_trimmed_qual_plot</span><span class="p">,</span><span class="w"> </span><span class="n">reverse_trimmed_qual_plot</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;trimmed_read_quality_plot_aggregated.pdf&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trimmed_qual_plots</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="trimmed-quality-scores">
<img alt="_images/trimmed_quality_scores.png" src="_images/trimmed_quality_scores.png" />
<figcaption>
<p><span class="caption-number">Fig. 18 </span><span class="caption-text">: A plot of the combined quality scores of all forward and reverse sequence files after primer trimming</span><a class="headerlink" href="#trimmed-quality-scores" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="reads-retained-after-primer-trimming">
<h4>4.3.2 Reads retained after primer trimming<a class="headerlink" href="#reads-retained-after-primer-trimming" title="Link to this heading">#</a></h4>
<p>Besides the quality scores, we can also calculate the percentage of reads retained after primer trimming. Investigating if a similar proportion of reads are retained across samples, as well as the actual average proportion can help determine issues regarding our code or specific samples.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Determine percentage retained after primer trimming</span>
<span class="n">metadata</span><span class="o">$</span><span class="n">trimming_retained_pct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">((</span><span class="n">metadata</span><span class="o">$</span><span class="n">trimmed_reads</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>
<span class="n">metadata</span><span class="o">$</span><span class="n">trimming_retained_pct</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt; metadata$trimming_retained_pct
 [1] 88.1 88.7 88.1 88.1 88.0 87.8 88.7 89.0 88.3 88.7 88.1 88.2 87.7 62.1 24.0 42.2 36.7 44.0 84.3 87.7 88.1 88.2 88.7 88.1
[25] 88.1 88.2 87.8 88.2 88.1 88.1 87.7 88.5 88.3 88.0 87.9 87.6 88.6 88.6 88.5 89.1 89.4 89.4 88.4 88.1 88.2
</pre></div>
</div>
</div>
<p>Looking at the output, we can see that we observe the same value for sample Akakina_1 as was reported by cutadapt earlier, i.e., 88.1%. As mentioned before, it is always good to double-check if results are as expected! Besides this single value, we can see that most values are in the high 80s, except for a few. Let’s investigate this further.</p>
</section>
<section id="visualise-read-count-after-primer-trimming">
<h4>4.3.3 Visualise read count after primer trimming<a class="headerlink" href="#visualise-read-count-after-primer-trimming" title="Link to this heading">#</a></h4>
<p>Rather than looking at the list of numbers, it can be easier to interpret data by visualising it in a graph. In this case, we can plot a stacked bar chart with the raw read count and trimmed read count. Since the values are stored in our metadata dataframe, we can accomplish this easily using the <a class="reference external" href="https://ggplot2.tidyverse.org/">ggplot2</a> R package. Let’s create a simple stacked bar chart, as well as a more annotated one to show the power of R and ggplot2.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualise reads after primer trimming</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sampleID</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;red&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trimmed_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;green&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">)</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stacked_bar_raw_trimmed.png&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">last_plot</span><span class="p">(),</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="stacked-bar-raw-trimmed">
<img alt="_images/stacked_bar_raw_trimmed.png" src="_images/stacked_bar_raw_trimmed.png" />
<figcaption>
<p><span class="caption-number">Fig. 19 </span><span class="caption-text">: A simple stacked bar chart depicting the number of reads before and after primer trimming</span><a class="headerlink" href="#stacked-bar-raw-trimmed" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>As you can see, with only 4 lines of code, we managed to create a simple stacked bar chart and save it to our <code class="docutils literal notranslate"><span class="pre">0.metadata</span></code> folder. However, let’s make this look a little bit nicer with some additional formatting and annotations.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Formatting and annotations to stacked bar chart</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">reorder</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trimmed_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;trimmed_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="nf">comma</span><span class="p">(</span><span class="n">trimmed_reads</span><span class="p">),</span><span class="w"> </span><span class="s">&quot; (&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sprintf</span><span class="p">(</span><span class="s">&quot;%.1f%%&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">trimming_retained_pct</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">)),</span>
<span class="w">    </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.1</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_nudge</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">0.02</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;trimmed_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#469597&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#BBC6C8&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Raw read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Trimmed read count&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">comma</span><span class="p">,</span><span class="w"> </span><span class="n">expand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">expansion</span><span class="p">(</span><span class="n">mult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Read Counts Before and After Trimming&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Sample ID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Number of Reads&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">panel.grid.major.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">())</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stacked_bar_raw_trimmed_annotated.png&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">last_plot</span><span class="p">(),</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="stacked-bar-raw-trimmed-annotated">
<img alt="_images/stacked_bar_raw_trimmed_annotated.png" src="_images/stacked_bar_raw_trimmed_annotated.png" />
<figcaption>
<p><span class="caption-number">Fig. 20 </span><span class="caption-text">: A formatted and annotated stacked bar chart depicting the number of reads before and after primer trimming</span><a class="headerlink" href="#stacked-bar-raw-trimmed-annotated" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>From this plot, we can see that a similar proportion of reads were retained after primer trimming for all samples, but that some of our negative controls lost a larger proportion of reads. This observation is quite common and fits within the parameters of metabarcoding data, so let’s continue with the analysis.</p>
</section>
</section>
</section>
<section id="quality-filtering">
<h2>5. Quality filtering<a class="headerlink" href="#quality-filtering" title="Link to this heading">#</a></h2>
<section id="sample-processing">
<h3>5.1 Sample processing<a class="headerlink" href="#sample-processing" title="Link to this heading">#</a></h3>
<p>During the bioinformatic pipeline, it is critical to only retain high-quality reads to reduce the abundance and impact of spurious sequences. While we conducted some filtering during primer trimming, the quality score plot still indicates additional filtering is needed. There is an intrinsic error rate to all polymerases used during PCR amplification, as well as sequencing technologies. For example, the most frequently used polymerase during PCR is Taq, though lacks 3’ to 5’ exonuclease proofreading activity, resulting in relatively low replication fidelity. These errors will generate some portion of sequences that vary from their biological origin sequence. Such reads can substantially inflate metrics such as alpha diversity, especially in a denoising approach (more on this later). While it is near-impossible to remove all of these sequences bioinformatically, especially PCR errors, we will attempt to remove erroneous reads by filtering on base calling quality scores (the fourth line of a sequence record in a .fastq file).</p>
<p>For quality filtering, we will trim all sequences that do not adhere to a specific set of rules. We will be using the <code class="docutils literal notranslate"><span class="pre">filterAndTrim()</span></code> function in <a class="reference external" href="https://benjjneb.github.io/dada2/">dada2</a> for this step. Quality filtering parameters are not standardized, but rather specific for each library. For our tutorial data, we will filter out all sequences that do not adhere to a minimum and maximum length, have unassigned base calls (‘N’), and have a higher expected error than 2. Once we have filtered our data, we can check the quality using the <code class="docutils literal notranslate"><span class="pre">plotQualityProfile()</span></code> function again and compare it to the previous steps.</p>
<p>Before quality filtering, we need to set some additional variables, including the output folder (in case it does not yet exist) and a list of output files</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step four: quality filtering</span>
<span class="c1"># Set variables</span>
<span class="kr">if</span><span class="p">(</span><span class="o">!</span><span class="nf">dir.exists</span><span class="p">(</span><span class="s">&quot;3.filtered&quot;</span><span class="p">))</span><span class="w"> </span><span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;3.filtered&quot;</span><span class="p">)</span>
<span class="n">filtered_forward_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;3.filtered&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">basename</span><span class="p">(</span><span class="n">raw_forward_reads</span><span class="p">))</span>
<span class="n">filtered_reverse_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;3.filtered&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">basename</span><span class="p">(</span><span class="n">raw_reverse_reads</span><span class="p">))</span>
</pre></div>
</div>
<p>Once we have set the variables, we can run the <code class="docutils literal notranslate"><span class="pre">filterAndTrim()</span></code> function. This function takes a list of input forward reads (<code class="docutils literal notranslate"><span class="pre">trimmed_forward_reads</span></code>), a list of output forward reads (<code class="docutils literal notranslate"><span class="pre">filtered_forward_reads</span></code>), a list of input reverse reads (<code class="docutils literal notranslate"><span class="pre">trimmed_reverse_reads</span></code>), a list of output reverse reads (<code class="docutils literal notranslate"><span class="pre">filtered_reverse_reads</span></code>), and the filtering parameters, including <code class="docutils literal notranslate"><span class="pre">maxEE</span></code> for the applied quality filtering threshold, <code class="docutils literal notranslate"><span class="pre">rm.phix</span></code> to remove any reads that match the PhiX bacteriophage genome (added to metabarcoding libraries to increas complexity), <code class="docutils literal notranslate"><span class="pre">maxN</span></code> to remove sequences containing unambiguous basecalls, and the truncation and length settings <code class="docutils literal notranslate"><span class="pre">truncLen</span></code>, <code class="docutils literal notranslate"><span class="pre">truncQ</span></code>, and <code class="docutils literal notranslate"><span class="pre">minLen</span></code>.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># perform quality filtering</span>
<span class="n">filtered_out</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">filterAndTrim</span><span class="p">(</span><span class="n">trimmed_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">filtered_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">trimmed_reverse_reads</span><span class="p">,</span><span class="w"> </span><span class="n">filtered_reverse_reads</span><span class="p">,</span>
<span class="w">                              </span><span class="n">trimLeft</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">truncLen</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">225</span><span class="p">,</span><span class="m">216</span><span class="p">),</span><span class="w"> </span><span class="n">maxN</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">maxEE</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="n">truncQ</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span>
<span class="w">                              </span><span class="n">minLen</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="n">rm.phix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">compress</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">multithread</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">filtered_out</span></code> is a named list containing the filenames, as well as a column with initial read counts (<code class="docutils literal notranslate"><span class="pre">reads.in</span></code>) and a column with read counts after filtering (<code class="docutils literal notranslate"><span class="pre">reads.out</span></code>). As we did before, let’s add the filtered read count to our metadata file for record keeping.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># add filtering results to metadata file</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">filtered_out</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_R1_001\\.fastq\\.gz$&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">filtered_out</span><span class="p">))</span>
<span class="n">metadata</span><span class="o">$</span><span class="n">filtered_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filtered_out</span><span class="p">[</span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;reads.out&quot;</span><span class="p">]</span>
</pre></div>
</div>
</section>
<section id="visualising-results">
<h3>5.2 Visualising results<a class="headerlink" href="#visualising-results" title="Link to this heading">#</a></h3>
<section id="quality-scores">
<h4>5.2.1 Quality scores<a class="headerlink" href="#quality-scores" title="Link to this heading">#</a></h4>
<p>As before, we can visualise the quality scores to determine if filtering was successfull. However, unlike before, let’s generate plots for each sample separately, rather than create one aggregated plot, and sort them by location.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step five: check quality filtered files</span>
<span class="c1"># Check quality scores across filtering steps</span>
<span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">location</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">sampleLocation</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">plot_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">()</span>
<span class="w">  </span><span class="n">location_samples</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">filter</span><span class="p">(</span><span class="n">sampleLocation</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">location</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">pull</span><span class="p">(</span><span class="n">sampleID</span><span class="p">)</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="s">&quot;Analysing samples of location&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">location</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="w">  </span><span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">seq_along</span><span class="p">(</span><span class="n">location_samples</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="n">f_plot_raw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">location_samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;_R1_001.fastq.gz&quot;</span><span class="p">)))</span>
<span class="w">    </span><span class="n">f_plot_trimmed</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;2.trimmed&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">location_samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;_R1_001.fastq.gz&quot;</span><span class="p">)))</span>
<span class="w">    </span><span class="n">f_plot_filtered</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;3.filtered&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">location_samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;_R1_001.fastq.gz&quot;</span><span class="p">)))</span>
<span class="w">    </span><span class="n">r_plot_raw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;1.raw&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">location_samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;_R2_001.fastq.gz&quot;</span><span class="p">)))</span>
<span class="w">    </span><span class="n">r_plot_trimmed</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;2.trimmed/&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">location_samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;_R2_001.fastq.gz&quot;</span><span class="p">)))</span>
<span class="w">    </span><span class="n">r_plot_filtered</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotQualityProfile</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;3.filtered/&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">location_samples</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="s">&quot;_R2_001.fastq.gz&quot;</span><span class="p">)))</span>
<span class="w">    </span><span class="n">plot_list</span><span class="p">[[</span><span class="n">i</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">grid.arrange</span><span class="p">(</span><span class="n">f_plot_raw</span><span class="p">,</span><span class="w"> </span><span class="n">f_plot_trimmed</span><span class="p">,</span><span class="w"> </span><span class="n">f_plot_filtered</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                   </span><span class="n">r_plot_raw</span><span class="p">,</span><span class="w"> </span><span class="n">r_plot_trimmed</span><span class="p">,</span><span class="w"> </span><span class="n">r_plot_filtered</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">)</span>
<span class="w">  </span><span class="p">}</span>
<span class="w">  </span><span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">location</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;_quality_profile.png&quot;</span><span class="p">)),</span>
<span class="w">         </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">marrangeGrob</span><span class="p">(</span><span class="n">plot_list</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">plot_list</span><span class="p">),</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span>
<span class="w">         </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">30</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">plot_list</span><span class="p">),</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
<figure class="align-default" id="tracked-quality-scores">
<img alt="_images/quality_scores_tracked.png" src="_images/quality_scores_tracked.png" />
<figcaption>
<p><span class="caption-number">Fig. 21 </span><span class="caption-text">: Tracked quality scores for location Akakina across all filtering steps for the forward and reverse reads</span><a class="headerlink" href="#tracked-quality-scores" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>From this plot, we can track the quality scores for each sample per location across all filtering steps for the forward and reverse reads. As shown in the figure, after the <code class="docutils literal notranslate"><span class="pre">filterAndTrim()</span></code> function, all quality scores are very high, so we can continue with the analysis and further filtering is not necessary.</p>
</section>
<section id="visualise-read-counts">
<h4>5.2.2 Visualise read counts<a class="headerlink" href="#visualise-read-counts" title="Link to this heading">#</a></h4>
<p>Similar to what we did to check primer trimming, we can generate a stacked bar plot, now including the number of reads after quality filtering. Note that we first calculate the percentage of reads retained, as before, for printing, as well as the similar code used for plotting.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Determine percentage retained after quality filtering</span>
<span class="n">metadata</span><span class="o">$</span><span class="n">filtering_retained_pct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">((</span><span class="n">metadata</span><span class="o">$</span><span class="n">filtered_reads</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>

<span class="c1"># Formatting and annotations to stacked bar chart</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">reorder</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trimmed_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;trimmed_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">filtered_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;filtered_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="nf">comma</span><span class="p">(</span><span class="n">filtered_reads</span><span class="p">),</span><span class="w"> </span><span class="s">&quot; (&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sprintf</span><span class="p">(</span><span class="s">&quot;%.1f%%&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">filtering_retained_pct</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">)),</span>
<span class="w">    </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.1</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_nudge</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">0.02</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;trimmed_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#469597&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#BBC6C8&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;filtered_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#DDBEAA&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Filtered read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Raw read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Trimmed read count&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">comma</span><span class="p">,</span><span class="w"> </span><span class="n">expand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">expansion</span><span class="p">(</span><span class="n">mult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Read counts across filtering steps&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Sample ID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Number of Reads&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">panel.grid.major.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">())</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stacked_bar_raw_trimmed_filtered.png&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">last_plot</span><span class="p">(),</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="stacked-bar-raw-trimmed-filtered">
<img alt="_images/stacked_bar_raw_trimmed_filtered.png" src="_images/stacked_bar_raw_trimmed_filtered.png" />
<figcaption>
<p><span class="caption-number">Fig. 22 </span><span class="caption-text">: A formatted and annotated stacked bar chart depicting the number of reads across filtering steps</span><a class="headerlink" href="#stacked-bar-raw-trimmed-filtered" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
</section>
</section>
<section id="learning-error-rates">
<h2>6. Learning error rates<a class="headerlink" href="#learning-error-rates" title="Link to this heading">#</a></h2>
<p>Once we have filtered out the low quality data and retained only high quality reads, we can move forward with our bioinformatic pipeline, i.e., generating an error model by learning the specific error-signature of our dataset. Each sequencing run, even when all goes well, will ahve its own subtle variations to its error profile. This step tries to assess that for both the forward and reverse reads, so that we can generate amplicon sequence variants (ASVs) in the next step. We can use the function <code class="docutils literal notranslate"><span class="pre">learnErrors()</span></code> in dada2 to accomplish this step. I recommend setting the parameter <code class="docutils literal notranslate"><span class="pre">multithread</span> <span class="pre">=</span> <span class="pre">TRUE</span></code>, as this is quite a computationally heavy step.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step six: learn error rates for our data</span>
<span class="c1"># Generate an error model</span>
<span class="n">err_forward_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">learnErrors</span><span class="p">(</span><span class="n">filtered_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">multithread</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="n">err_reverse_reads</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">learnErrors</span><span class="p">(</span><span class="n">filtered_reverse_reads</span><span class="p">,</span><span class="w"> </span><span class="n">multithread</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">err_forward_reads</span> <span class="o">&lt;-</span> <span class="n">learnErrors</span><span class="p">(</span><span class="n">filtered_forward_reads</span><span class="p">,</span> <span class="n">multithread</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="mi">113168250</span> <span class="n">total</span> <span class="n">bases</span> <span class="ow">in</span> <span class="mi">502970</span> <span class="n">reads</span> <span class="kn">from</span><span class="w"> </span><span class="mi">5</span> <span class="n">samples</span> <span class="n">will</span> <span class="n">be</span> <span class="n">used</span> <span class="k">for</span> <span class="n">learning</span> <span class="n">the</span> <span class="n">error</span> <span class="n">rates</span><span class="o">.</span>
<span class="o">&gt;</span> <span class="n">err_reverse_reads</span> <span class="o">&lt;-</span> <span class="n">learnErrors</span><span class="p">(</span><span class="n">filtered_reverse_reads</span><span class="p">,</span> <span class="n">multithread</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="mi">108641520</span> <span class="n">total</span> <span class="n">bases</span> <span class="ow">in</span> <span class="mi">502970</span> <span class="n">reads</span> <span class="kn">from</span><span class="w"> </span><span class="mi">5</span> <span class="n">samples</span> <span class="n">will</span> <span class="n">be</span> <span class="n">used</span> <span class="k">for</span> <span class="n">learning</span> <span class="n">the</span> <span class="n">error</span> <span class="n">rates</span><span class="o">.</span>
</pre></div>
</div>
</div>
<p>Dada2 also incorporates a plotting function (<code class="docutils literal notranslate"><span class="pre">plotErrors()</span></code>) to visualise how well the estimated error rates match with up with the observed. The red line of the graph is the expected error rate based on the quality score, the black line represents the estimate, and the grey dots represent the observed error rates. In general, we want the error frequency (y-axis) to reduce with increased quality scores (y-axis) and that the observed error rates (grey dots) follow the estimated error rates (black line).</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot error rates</span>
<span class="n">error_reads_forward_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotErrors</span><span class="p">(</span><span class="n">err_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">nominalQ</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Error Forward&quot;</span><span class="p">)</span>
<span class="n">error_reads_reverse_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">plotErrors</span><span class="p">(</span><span class="n">err_reverse_reads</span><span class="p">,</span><span class="w"> </span><span class="n">nominalQ</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Error Reverse&quot;</span><span class="p">)</span>
<span class="n">error_plots</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggarrange</span><span class="p">(</span><span class="n">error_reads_forward_plot</span><span class="p">,</span><span class="w"> </span><span class="n">error_reads_reverse_plot</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;error_plots.pdf&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">error_plots</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="error-plots">
<img alt="_images/error_plots.png" src="_images/error_plots.png" />
<figcaption>
<p><span class="caption-number">Fig. 23 </span><span class="caption-text">: The expected (red line), estimated (black line), and observed (grey dots) error rates for our dataset.</span><a class="headerlink" href="#error-plots" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="generating-asvs">
<h2>7. Generating ASVs<a class="headerlink" href="#generating-asvs" title="Link to this heading">#</a></h2>
<section id="denoising-the-data">
<h3>7.1 Denoising the data<a class="headerlink" href="#denoising-the-data" title="Link to this heading">#</a></h3>
<p>Now that our data set is filtered and we’ve learned the error rates, we are ready for the next step in the bioinformatic pipeline, i.e., looking for <strong>biologically meaningful</strong> or <strong>biologically correct</strong> sequences. Two approaches exist to achieve this goal, including <strong>denoising</strong> and <strong>clustering</strong>. There is still an ongoing debate on what the best approach is to obtain these biologically meaningful sequences. For more information, these are two good papers to read: (<a class="reference external" href="https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13398">Brandt et al., 2021</a>) and (<a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04115-6">Antich et al., 2021</a>). For this workshop, we won’t be discussing this topic in much detail, but this is the basic idea…</p>
<p>When <strong>clustering</strong> the dataset, OTUs (Operational Taxonomic Units) will be generated by combining sequences that are similar to a set percentage level (traditionally 97%), with the most abundant sequence identified as the true sequence. When clustering at 97%, this means that if a sequence is more than 3% different than the generated OTU, a second OTU will be generated. The concept of OTU clustering was introduced in the 1960s and has been debated since. Clustering the dataset is usually used for identifying species in metabarcoding data.</p>
<p><strong>Denoising</strong>, on the other hand, attempts to identify all correct biological sequences through an algorithm. In short, denoising will cluster the data with a 100% threshold and tries to identify errors based on abundance differences and error profiles. The retained sequences are called <strong>ASVs (Amplicon Sequence Variants)</strong> or ZOTUs (Zero-radius Operation Taxonomic Unit). Denoising the dataset is usually used for identifying intraspecific variation in metabarcoding data. A schematic of both approaches can be found below.</p>
<figure class="align-default" id="denoising-and-clustering">
<img alt="_images/denoising_and_clustering.png" src="_images/denoising_and_clustering.png" />
<figcaption>
<p><span class="caption-number">Fig. 24 </span><span class="caption-text">: A schematic representation of denoising and clustering. Copyright: <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04115-6">Antich et al., 2021</a></span><a class="headerlink" href="#denoising-and-clustering" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>This difference in approach may seem small but has a very big impact on your final dataset!</p>
<p>When you denoise the dataset, it is expected that one species may have more than one ASV. This means that when the reference database is incomplete, or you plan to work with ASVs instead of taxonomic assignments, your diversity estimates will be highly inflated. When clustering the dataset, on the other hand, it is expected that an OTU may have more than one species assigned to it, meaning that you may lose some correct biological sequences that are present in your data by merging species with barcodes more similar than 97%. In other words, you will miss out on differentiating closely related species and intraspecific variation.</p>
<p>Since we are following the <a class="reference external" href="https://benjjneb.github.io/dada2/">dada2</a> pipeline, we will be denoising the data set and infer ASVs based on read abundance and the error profiles we have generated in the previous step. For more information, I recommend reading the <a class="reference external" href="https://www.nature.com/articles/nmeth.3869#methods">dada2 publication</a>.</p>
<div class="tip dropdown admonition">
<p class="admonition-title">Dereplication</p>
<p>In the latest versions of dada2, dereplicating the metabarcoding data, i.e., finding unique sequences, does not require a separate step, as it is implemented in the <code class="docutils literal notranslate"><span class="pre">dada()</span></code> function.</p>
</div>
<div class="tip dropdown admonition">
<p class="admonition-title">Pooling samples</p>
<p>Interestingly, ASVs can be inferred from each sample separately, but also by combining all samples together using the <code class="docutils literal notranslate"><span class="pre">pool</span> <span class="pre">=</span> <span class="pre">TRUE</span></code> parameter. Since this step is computationally intensive, we will not pool the samples for this workshop, but this is something you might want to do for your own data!</p>
</div>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step seven: generate ASVs</span>
<span class="c1"># Infer ASVs using the dada algorithm</span>
<span class="n">dada_forward</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dada</span><span class="p">(</span><span class="n">filtered_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">err</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">err_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">pool</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">multithread</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="n">dada_reverse</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">dada</span><span class="p">(</span><span class="n">filtered_reverse_reads</span><span class="p">,</span><span class="w"> </span><span class="n">err</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">err_reverse_reads</span><span class="p">,</span><span class="w"> </span><span class="n">pool</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">multithread</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">dada_forward</span> <span class="o">&lt;-</span> <span class="n">dada</span><span class="p">(</span><span class="n">filtered_forward_reads</span><span class="p">,</span> <span class="n">err</span> <span class="o">=</span> <span class="n">err_forward_reads</span><span class="p">,</span> <span class="n">pool</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">multithread</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="n">Sample</span> <span class="mi">1</span> <span class="o">-</span> <span class="mi">64904</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">8480</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">2</span> <span class="o">-</span> <span class="mi">103525</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13912</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">3</span> <span class="o">-</span> <span class="mi">64719</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">7727</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">4</span> <span class="o">-</span> <span class="mi">106362</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13776</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">5</span> <span class="o">-</span> <span class="mi">163460</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">22246</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">6</span> <span class="o">-</span> <span class="mi">171466</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">18547</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">7</span> <span class="o">-</span> <span class="mi">55583</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">6691</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">8</span> <span class="o">-</span> <span class="mi">104685</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10538</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">9</span> <span class="o">-</span> <span class="mi">124109</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13655</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">10</span> <span class="o">-</span> <span class="mi">105792</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">12078</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">11</span> <span class="o">-</span> <span class="mi">86312</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10089</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">12</span> <span class="o">-</span> <span class="mi">125966</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">14268</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">13</span> <span class="o">-</span> <span class="mi">11069</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">1118</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">14</span> <span class="o">-</span> <span class="mi">9735</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">1180</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">15</span> <span class="o">-</span> <span class="mi">5431</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">865</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">16</span> <span class="o">-</span> <span class="mi">8272</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">1083</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">17</span> <span class="o">-</span> <span class="mi">308</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">73</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">18</span> <span class="o">-</span> <span class="mi">5770</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">761</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">19</span> <span class="o">-</span> <span class="mi">51019</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">5270</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">20</span> <span class="o">-</span> <span class="mi">166483</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">18215</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">21</span> <span class="o">-</span> <span class="mi">109777</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">12073</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">22</span> <span class="o">-</span> <span class="mi">63818</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">6319</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">23</span> <span class="o">-</span> <span class="mi">143284</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">15101</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">24</span> <span class="o">-</span> <span class="mi">95323</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11985</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">25</span> <span class="o">-</span> <span class="mi">105557</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11235</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">26</span> <span class="o">-</span> <span class="mi">131517</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">14691</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">27</span> <span class="o">-</span> <span class="mi">113171</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11752</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">28</span> <span class="o">-</span> <span class="mi">68148</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">8922</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">29</span> <span class="o">-</span> <span class="mi">101575</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10303</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">30</span> <span class="o">-</span> <span class="mi">128532</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">14843</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">31</span> <span class="o">-</span> <span class="mi">84763</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">9613</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">32</span> <span class="o">-</span> <span class="mi">63680</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">6861</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">33</span> <span class="o">-</span> <span class="mi">95882</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10886</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">34</span> <span class="o">-</span> <span class="mi">75380</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">9105</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">35</span> <span class="o">-</span> <span class="mi">73502</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">9774</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">36</span> <span class="o">-</span> <span class="mi">56787</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">8489</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">37</span> <span class="o">-</span> <span class="mi">65363</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">7542</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">38</span> <span class="o">-</span> <span class="mi">104662</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11532</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">39</span> <span class="o">-</span> <span class="mi">109573</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11254</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">40</span> <span class="o">-</span> <span class="mi">68568</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10310</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">41</span> <span class="o">-</span> <span class="mi">60316</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">8631</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">42</span> <span class="o">-</span> <span class="mi">52152</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">8201</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">43</span> <span class="o">-</span> <span class="mi">81881</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10421</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">44</span> <span class="o">-</span> <span class="mi">84642</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10825</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">45</span> <span class="o">-</span> <span class="mi">114864</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13894</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="o">&gt;</span> <span class="n">dada_reverse</span> <span class="o">&lt;-</span> <span class="n">dada</span><span class="p">(</span><span class="n">filtered_reverse_reads</span><span class="p">,</span> <span class="n">err</span> <span class="o">=</span> <span class="n">err_reverse_reads</span><span class="p">,</span> <span class="n">pool</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">multithread</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="n">Sample</span> <span class="mi">1</span> <span class="o">-</span> <span class="mi">64904</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">9966</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">2</span> <span class="o">-</span> <span class="mi">103525</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">14506</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">3</span> <span class="o">-</span> <span class="mi">64719</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">9345</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">4</span> <span class="o">-</span> <span class="mi">106362</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">17586</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">5</span> <span class="o">-</span> <span class="mi">163460</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">22977</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">6</span> <span class="o">-</span> <span class="mi">171466</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">24066</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">7</span> <span class="o">-</span> <span class="mi">55583</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">7859</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">8</span> <span class="o">-</span> <span class="mi">104685</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">14191</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">9</span> <span class="o">-</span> <span class="mi">124109</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">18170</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">10</span> <span class="o">-</span> <span class="mi">105792</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">15100</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">11</span> <span class="o">-</span> <span class="mi">86312</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13004</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">12</span> <span class="o">-</span> <span class="mi">125966</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">16830</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">13</span> <span class="o">-</span> <span class="mi">11069</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">1433</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">14</span> <span class="o">-</span> <span class="mi">9735</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">2359</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">15</span> <span class="o">-</span> <span class="mi">5431</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">1737</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">16</span> <span class="o">-</span> <span class="mi">8272</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">2971</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">17</span> <span class="o">-</span> <span class="mi">308</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">151</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">18</span> <span class="o">-</span> <span class="mi">5770</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">1849</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">19</span> <span class="o">-</span> <span class="mi">51019</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">7735</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">20</span> <span class="o">-</span> <span class="mi">166483</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">20967</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">21</span> <span class="o">-</span> <span class="mi">109777</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13798</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">22</span> <span class="o">-</span> <span class="mi">63818</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">8731</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">23</span> <span class="o">-</span> <span class="mi">143284</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">19553</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">24</span> <span class="o">-</span> <span class="mi">95323</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13637</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">25</span> <span class="o">-</span> <span class="mi">105557</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13256</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">26</span> <span class="o">-</span> <span class="mi">131517</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">18209</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">27</span> <span class="o">-</span> <span class="mi">113171</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">14828</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">28</span> <span class="o">-</span> <span class="mi">68148</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11521</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">29</span> <span class="o">-</span> <span class="mi">101575</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11118</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">30</span> <span class="o">-</span> <span class="mi">128532</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">16911</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">31</span> <span class="o">-</span> <span class="mi">84763</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11360</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">32</span> <span class="o">-</span> <span class="mi">63680</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">8199</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">33</span> <span class="o">-</span> <span class="mi">95882</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13894</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">34</span> <span class="o">-</span> <span class="mi">75380</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">11080</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">35</span> <span class="o">-</span> <span class="mi">73502</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10961</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">36</span> <span class="o">-</span> <span class="mi">56787</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">9284</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">37</span> <span class="o">-</span> <span class="mi">65363</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">9197</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">38</span> <span class="o">-</span> <span class="mi">104662</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">15520</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">39</span> <span class="o">-</span> <span class="mi">109573</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">17623</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">40</span> <span class="o">-</span> <span class="mi">68568</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">10510</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">41</span> <span class="o">-</span> <span class="mi">60316</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">9580</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">42</span> <span class="o">-</span> <span class="mi">52152</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">8129</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">43</span> <span class="o">-</span> <span class="mi">81881</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">12792</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">44</span> <span class="o">-</span> <span class="mi">84642</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">13029</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
<span class="n">Sample</span> <span class="mi">45</span> <span class="o">-</span> <span class="mi">114864</span> <span class="n">reads</span> <span class="ow">in</span> <span class="mi">17908</span> <span class="n">unique</span> <span class="n">sequences</span><span class="o">.</span>
</pre></div>
</div>
</div>
</section>
<section id="id1">
<h3>7.2 Visualising results<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<p>Let’s build upon our stacked bar chart and include the read count after denoising. First, we need to extract the read count from the dada objects we created during denoising.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step eight: check generated ASVs</span>
<span class="c1"># Extract read count from dada objects and add to metadata</span>
<span class="n">dada_f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sapply</span><span class="p">(</span><span class="n">dada_forward</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">getUniques</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
<span class="n">dada_r</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sapply</span><span class="p">(</span><span class="n">dada_reverse</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">getUniques</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
</pre></div>
</div>
<p>Next, we need to rename the labels to match the sample IDs rather than file names.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Rename labels to match sample IDs using regex patterns</span>
<span class="nf">names</span><span class="p">(</span><span class="n">dada_f</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_R1_001\\.fastq\\.gz$&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">dada_f</span><span class="p">))</span>
<span class="nf">names</span><span class="p">(</span><span class="n">dada_r</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_R2_001\\.fastq\\.gz$&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">dada_r</span><span class="p">))</span>
</pre></div>
</div>
<p>Then, we need to find the minimum value between the read count for the forward and reverse reads.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Find the minimum value of read count and add to metadata</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span>
<span class="w">    </span><span class="n">denoised_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">pmin</span><span class="p">(</span>
<span class="w">      </span><span class="n">dada_f</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">dada_f</span><span class="p">))],</span>
<span class="w">      </span><span class="n">dada_r</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">dada_r</span><span class="p">))]))</span>
</pre></div>
</div>
<p>Once we have the read counts, we can calculate the percentage as before and plot the data.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1"># Determine percentage retained after denoising</span>
<span class="n">metadata</span><span class="o">$</span><span class="n">denoising_retained_pct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">((</span><span class="n">metadata</span><span class="o">$</span><span class="n">denoised_reads</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>

<span class="c1"># Formatting and annotations to stacked bar chart</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">reorder</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trimmed_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;trimmed_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">filtered_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;filtered_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">denoised_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;denoised_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="nf">comma</span><span class="p">(</span><span class="n">denoised_reads</span><span class="p">),</span><span class="w"> </span><span class="s">&quot; (&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sprintf</span><span class="p">(</span><span class="s">&quot;%.1f%%&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">denoising_retained_pct</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">)),</span>
<span class="w">    </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.1</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_nudge</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">0.02</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;trimmed_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#469597&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#BBC6C8&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;filtered_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#DDBEAA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;denoised_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#806491&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Denoised read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Filtered read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Raw read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Trimmed read count&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">comma</span><span class="p">,</span><span class="w"> </span><span class="n">expand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">expansion</span><span class="p">(</span><span class="n">mult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Read counts across filtering steps&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Sample ID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Number of Reads&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">panel.grid.major.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">())</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stacked_bar_raw_trimmed_filtered_denoised.png&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">last_plot</span><span class="p">(),</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="stacked-bar-raw-trimmed-filtered-denoised">
<img alt="_images/stacked_bar_raw_trimmed_filtered_denoised.png" src="_images/stacked_bar_raw_trimmed_filtered_denoised.png" />
<figcaption>
<p><span class="caption-number">Fig. 25 </span><span class="caption-text">: A formatted and annotated stacked bar chart depicting the number of reads across filtering steps, including denoising</span><a class="headerlink" href="#stacked-bar-raw-trimmed-filtered-denoised" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
</section>
<section id="merging-forward-and-reverse-reads">
<h2>8. Merging forward and reverse reads<a class="headerlink" href="#merging-forward-and-reverse-reads" title="Link to this heading">#</a></h2>
<section id="reconstruct-the-full-amplicon">
<h3>8.1 Reconstruct the full amplicon<a class="headerlink" href="#reconstruct-the-full-amplicon" title="Link to this heading">#</a></h3>
<p>Now that we have our ASVs, we can merge the forward and reverse reads in our data to reconstruct the full amplicon. We can merge reads using the <code class="docutils literal notranslate"><span class="pre">mergePairs()</span></code> function in the dada2 R package. Merging forward and reverse reads works by the Needleman-Wunsch algorithm. By default, a successful alignment requirs an overlap of at least 12 basepairs (<code class="docutils literal notranslate"><span class="pre">minOverlap</span></code>), as well as 0 mismatches (<code class="docutils literal notranslate"><span class="pre">maxMismatch</span></code>). However, for our data, we will set these parameters to 10 and 2, respectively. Also, we can set the <code class="docutils literal notranslate"><span class="pre">trimOverhang</span> <span class="pre">=</span> <span class="pre">TRUE</span></code> parameter in case any of the reads go passed their opposite primers (this should not be the case, however, given the amplicon length and the primer trimming we performed with cutadapt).</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step nine: merge forward and reverse reads</span>
<span class="n">merged_amplicons</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">mergePairs</span><span class="p">(</span><span class="n">dada_forward</span><span class="p">,</span><span class="w"> </span><span class="n">filtered_forward_reads</span><span class="p">,</span><span class="w"> </span><span class="n">dada_reverse</span><span class="p">,</span><span class="w"> </span><span class="n">filtered_reverse_reads</span><span class="p">,</span>
<span class="w">                               </span><span class="n">minOverlap</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">maxMismatch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">trimOverhang</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">merged_amplicons</span> <span class="o">&lt;-</span> <span class="n">mergePairs</span><span class="p">(</span><span class="n">dada_forward</span><span class="p">,</span> <span class="n">filtered_forward_reads</span><span class="p">,</span> <span class="n">dada_reverse</span><span class="p">,</span> <span class="n">filtered_reverse_reads</span><span class="p">,</span>
<span class="o">+</span>                                <span class="n">minOverlap</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">maxMismatch</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">trimOverhang</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="mi">63905</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">481</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">64011</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">526</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">102341</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">532</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">102506</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">590</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">63809</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">419</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">63910</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">466</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">104535</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">895</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">104806</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">998</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">160432</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1120</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">161048</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1318</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">169231</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1110</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">169546</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1250</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">54543</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">457</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">54699</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">516</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">103161</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">647</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">103453</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">744</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">122118</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">855</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">122531</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">981</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">103894</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">904</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">104250</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1034</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">84278</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">734</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">84613</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">835</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">124047</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1016</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">124343</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1130</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">11026</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">33</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">11026</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">33</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">9609</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">40</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">9609</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">40</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">5235</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">22</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">5315</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">24</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">7975</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">26</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">8103</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">29</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="mi">286</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">6</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">286</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">6</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="mi">5606</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">27</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">5637</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">30</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">50696</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">161</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">50830</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">169</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">164020</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1139</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">164469</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1303</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">108217</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">890</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">108476</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">987</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">63111</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">351</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">63260</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">397</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="mi">141414</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">866</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">141724</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">948</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">93352</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1092</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">93662</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1211</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">104131</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">699</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">104303</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">771</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">129513</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">783</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">129905</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">909</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">111377</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">645</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">111691</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">718</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">66827</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">606</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">67019</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">664</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">100425</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">520</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">100541</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">574</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">126806</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">856</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">127068</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">962</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">83417</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">555</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">83526</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">602</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">62608</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">423</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">62683</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">454</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">94431</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">575</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">94715</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">640</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">73538</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1054</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">73613</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1090</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">71536</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1024</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">71663</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1071</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">54618</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">959</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">54741</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">995</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">64036</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">548</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">64197</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">597</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">103004</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">830</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">103221</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">904</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">107782</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">802</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">108086</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">896</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">67776</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">312</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">67848</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">342</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">59906</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">252</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">60013</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">290</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">51656</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">257</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">51728</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">284</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">80233</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">747</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">80438</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">811</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">83003</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">790</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">83284</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">876</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
<span class="mi">112873</span> <span class="n">paired</span><span class="o">-</span><span class="n">reads</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1052</span> <span class="n">unique</span> <span class="n">pairings</span><span class="p">)</span> <span class="n">successfully</span> <span class="n">merged</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">113147</span> <span class="p">(</span><span class="ow">in</span> <span class="mi">1167</span> <span class="n">pairings</span><span class="p">)</span> <span class="nb">input</span><span class="o">.</span>
<span class="n">Duplicate</span> <span class="n">sequences</span> <span class="ow">in</span> <span class="n">merged</span> <span class="n">output</span><span class="o">.</span>
</pre></div>
</div>
</div>
</section>
<section id="id2">
<h3>8.2 Visualising results<a class="headerlink" href="#id2" title="Link to this heading">#</a></h3>
<p>Similarly to the previous steps, we can build upon the stacked bar chart to add the read counts after merging.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step ten: check merging success</span>
<span class="c1"># Extract read count from merged_amplicons and add to metadata</span>
<span class="n">merged_count</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sapply</span><span class="p">(</span><span class="n">merged_amplicons</span><span class="p">,</span><span class="w"> </span><span class="kr">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">getUniques</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
<span class="nf">names</span><span class="p">(</span><span class="n">merged_count</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_R1_001\\.fastq\\.gz$&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">merged_count</span><span class="p">))</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span><span class="n">merged_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">merged_count</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">merged_count</span><span class="p">))])</span>

<span class="c1"># Determine percentage retained after merging</span>
<span class="n">metadata</span><span class="o">$</span><span class="n">merging_retained_pct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">((</span><span class="n">metadata</span><span class="o">$</span><span class="n">merged_reads</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>

<span class="c1"># Formatting and annotations to stacked bar chart</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">reorder</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trimmed_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;trimmed_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">filtered_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;filtered_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">denoised_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;denoised_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">denoised_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;merged_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="nf">comma</span><span class="p">(</span><span class="n">merged_reads</span><span class="p">),</span><span class="w"> </span><span class="s">&quot; (&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sprintf</span><span class="p">(</span><span class="s">&quot;%.1f%%&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">merging_retained_pct</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">)),</span>
<span class="w">    </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.1</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_nudge</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">0.02</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;trimmed_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#469597&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#BBC6C8&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;filtered_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#DDBEAA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;denoised_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#806491&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;merged_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#2F70AF&quot;</span><span class="p">),</span><span class="w"> </span>
<span class="w">  </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Denoised read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Filtered read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Merged read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Raw read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Trimmed read count&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">comma</span><span class="p">,</span><span class="w"> </span><span class="n">expand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">expansion</span><span class="p">(</span><span class="n">mult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Read counts across filtering steps&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Sample ID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Number of Reads&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">panel.grid.major.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">())</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stacked_bar_raw_trimmed_filtered_denoised_merged.png&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">last_plot</span><span class="p">(),</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="stacked-bar-raw-trimmed-filtered-denoised-merged">
<img alt="_images/stacked_bar_raw_trimmed_filtered_denoised_merged.png" src="_images/stacked_bar_raw_trimmed_filtered_denoised_merged.png" />
<figcaption>
<p><span class="caption-number">Fig. 26 </span><span class="caption-text">: A formatted and annotated stacked bar chart depicting the number of reads across filtering steps, including denoising and merging</span><a class="headerlink" href="#stacked-bar-raw-trimmed-filtered-denoised-merged" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
</section>
<section id="count-table">
<h2>9. Count table<a class="headerlink" href="#count-table" title="Link to this heading">#</a></h2>
<p>Once reads are merged, we can generate a count table with the <code class="docutils literal notranslate"><span class="pre">makeSequenceTable()</span></code> function in the dada2 R package. Using the <code class="docutils literal notranslate"><span class="pre">dim()</span></code> function, we can investigate the dimensions of the table, which represents the number of samples (45) and number of ASVs (9,313).</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step 11: construct a count table</span>
<span class="n">count_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">makeSequenceTable</span><span class="p">(</span><span class="n">merged_amplicons</span><span class="p">)</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">count_table</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">dim</span><span class="p">(</span><span class="n">count_table</span><span class="p">)</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span>   <span class="mi">45</span> <span class="mi">9313</span>
</pre></div>
</div>
</div>
</section>
<section id="chimera-removal">
<h2>10. Chimera removal<a class="headerlink" href="#chimera-removal" title="Link to this heading">#</a></h2>
<p>As a final step in our bioinformatic pipeline, we need to remove chimeric sequences. Amplicon sequencing has the potential to generate chimeric reads, which can cause spurious inference of biological variation. Chimeric amplicons form when an incomplete DNA strand anneals to a different template and primes synthesis of a new template derived from two different biological sequences, or in other words chimeras are artefact sequences formed by two or more biological sequences incorrectly joined together. More information can be found in <a class="reference external" href="https://www.biorxiv.org/content/biorxiv/early/2016/09/09/074252.full.pdf">this paper</a> and a simple illustration can be found below.</p>
<figure class="align-default" id="chimeras">
<img alt="_images/chimeras.png" src="_images/chimeras.png" />
<figcaption>
<p><span class="caption-number">Fig. 27 </span><span class="caption-text">: A schematic representation of chimera formation. Copyright: <a class="reference external" href="https://genome.cshlp.org/content/21/3/494/F1.expansion.html">Genome Research</a></span><a class="headerlink" href="#chimeras" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Dada2 identifies likely chimeras by aligning each sequence with those that were recovered in greater abundance and then determining if any lower-abundance sequences can be made exactly by mixing left and right portions of two of the more abundant ones. Dada2 accomplishes this by the column names, which are currently our ASV sequences, and column sums to gather information on abundance.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step 12: removing chimeras</span>
<span class="n">count_table_nochimeras</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">removeBimeraDenovo</span><span class="p">(</span><span class="n">count_table</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;consensus&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">multithread</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">count_table_nochimeras</span> <span class="o">&lt;-</span> <span class="n">removeBimeraDenovo</span><span class="p">(</span><span class="n">count_table</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;consensus&quot;</span><span class="p">,</span> <span class="n">multithread</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span>
<span class="n">Identified</span> <span class="mi">999</span> <span class="n">bimeras</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">9313</span> <span class="nb">input</span> <span class="n">sequences</span><span class="o">.</span>
</pre></div>
</div>
</div>
</section>
<section id="read-count-tracking">
<h2>11. Read count tracking<a class="headerlink" href="#read-count-tracking" title="Link to this heading">#</a></h2>
<p>Now that we have finalised our bioinformatic pipeline, we can finish the stacked bar chart by incorporating the read counts after chimera removal.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step 13: check chimera removal numbers</span>
<span class="c1"># Extract read count per sample and add to metadata</span>
<span class="n">non_chimera_count</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rowSums</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">)</span>
<span class="nf">names</span><span class="p">(</span><span class="n">non_chimera_count</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_R1_001\\.fastq\\.gz$&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">non_chimera_count</span><span class="p">))</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span><span class="n">nonchimera_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">non_chimera_count</span><span class="p">[</span><span class="nf">match</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">non_chimera_count</span><span class="p">))])</span>

<span class="c1"># Determine percentage retained after merging</span>
<span class="n">metadata</span><span class="o">$</span><span class="n">nonchimera_retained_pct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">((</span><span class="n">metadata</span><span class="o">$</span><span class="n">nonchimera_reads</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)</span>

<span class="c1"># Formatting and annotations to stacked bar chart</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">reorder</span><span class="p">(</span><span class="n">sampleID</span><span class="p">,</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trimmed_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;trimmed_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">filtered_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;filtered_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">denoised_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;denoised_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">denoised_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;merged_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_bar</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">nonchimera_reads</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;nonchimera_reads&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_text</span><span class="p">(</span><span class="nf">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">raw_reads_f</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="nf">comma</span><span class="p">(</span><span class="n">nonchimera_reads</span><span class="p">),</span><span class="w"> </span><span class="s">&quot; (&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">sprintf</span><span class="p">(</span><span class="s">&quot;%.1f%%&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">nonchimera_retained_pct</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;)&quot;</span><span class="p">)),</span>
<span class="w">    </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.1</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">position_nudge</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">metadata</span><span class="o">$</span><span class="n">raw_reads_f</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">0.02</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;trimmed_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#469597&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;raw_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#BBC6C8&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;filtered_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#DDBEAA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;denoised_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#806491&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;merged_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#2F70AF&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;nonchimera_reads&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#5F021F&quot;</span><span class="p">),</span><span class="w"> </span>
<span class="w">  </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Denoised read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Filtered read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Merged read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Non chimera read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Raw read count&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Trimmed read count&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_y_continuous</span><span class="p">(</span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">comma</span><span class="p">,</span><span class="w"> </span><span class="n">expand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">expansion</span><span class="p">(</span><span class="n">mult</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Read counts across filtering steps&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Sample ID&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Number of Reads&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">panel.grid.major.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">())</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;0.metadata&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;stacked_bar_raw_trimmed_filtered_denoised_merged_nonchimera.png&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">last_plot</span><span class="p">(),</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</pre></div>
</div>
<figure class="align-default" id="stacked-bar-raw-trimmed-filtered-denoised-merged-nonchimera">
<img alt="_images/stacked_bar_raw_trimmed_filtered_denoised_merged_nonchimera.png" src="_images/stacked_bar_raw_trimmed_filtered_denoised_merged_nonchimera.png" />
<figcaption>
<p><span class="caption-number">Fig. 28 </span><span class="caption-text">: A formatted and annotated stacked bar chart depicting the number of reads across filtering steps, including denoising, merging, and chimera removal</span><a class="headerlink" href="#stacked-bar-raw-trimmed-filtered-denoised-merged-nonchimera" title="Link to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="saving-output-files">
<h2>12. Saving output files<a class="headerlink" href="#saving-output-files" title="Link to this heading">#</a></h2>
<p>Before we end the day, it is best to reformat our output and save the files.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## Step 13: Saving output files</span>
<span class="c1"># Create directory</span>
<span class="kr">if</span><span class="p">(</span><span class="o">!</span><span class="nf">dir.exists</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">))</span><span class="w"> </span><span class="nf">dir.create</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">)</span>
<span class="c1"># Export metadata file</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">metadata</span><span class="p">,</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;metadata.txt&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">quote</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
<span class="c1"># Change ASV IDs</span>
<span class="n">asv_ids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="s">&quot;ASV_&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">seq_along</span><span class="p">(</span><span class="nf">colnames</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">)))</span>
<span class="nf">names</span><span class="p">(</span><span class="n">asv_ids</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">colnames</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">)</span>
<span class="c1"># Export ASV sequences</span>
<span class="nf">uniquesToFasta</span><span class="p">(</span><span class="nf">getUniques</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">),</span><span class="w"> </span><span class="n">fout</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">,</span><span class="s">&quot;asvs.fasta&quot;</span><span class="p">),</span>
<span class="w">               </span><span class="n">ids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">asv_ids</span><span class="p">[</span><span class="nf">colnames</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">)])</span>
<span class="c1"># Export count table</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">asv_ids</span><span class="p">[</span><span class="nf">colnames</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">)]</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gsub</span><span class="p">(</span><span class="s">&quot;_R1_001\\.fastq\\.gz$&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">))</span>
<span class="n">count_table_nochimeras</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">))</span>
<span class="n">count_table_nochimeras</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">cbind</span><span class="p">(</span><span class="n">sampleID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">),</span><span class="w"> </span><span class="n">count_table_nochimeras</span><span class="p">)</span>
<span class="nf">write.table</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">count_table_nochimeras</span><span class="p">),</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;count_table.txt&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">quote</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;\t&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</div>
<p>Today, we have generated three files, including:</p>
<ol class="arabic simple">
<li><p>an updated metadata file containing read count information across the different steps of the bioinformatic pipeline</p></li>
<li><p>an ASV fasta file containing a list of all ASVs</p></li>
<li><p>a count table containing information on read abundance for each ASV per sample</p></li>
</ol>
<p><strong>That’s it for today, we will see you tomorrow when we will be creating custom reference databases and assigning a taxonomic ID to our ASVs!</strong></p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="setup.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Setting up the environment</p>
      </div>
    </a>
    <a class="right-next"
       href="taxonomyassignment.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Taxonomy assignment</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setting-up-the-r-environment">1. Setting up the R environment</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exploring-the-starting-files">2. Exploring the starting files</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-metadata-file">2.1 Sample metadata file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sequence-files">2.2 Sequence files</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#data-format">2.2.1 Data format</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#renaming-files">2.2.2 Renaming files</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#initial-data-checks">3. Initial data checks</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#raw-data-quality-scores">3.1 Raw data quality scores</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#read-count-per-sample">3.2 Read count per sample</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#removing-primer-sequences">4. Removing primer sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-single-sample-example">4.1 A single sample example</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#batch-trimming">4.2 Batch trimming</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summarising-results">4.3 Summarising results</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-scores-after-primer-trimming">4.3.1 Quality scores after primer trimming</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#reads-retained-after-primer-trimming">4.3.2 Reads retained after primer trimming</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#visualise-read-count-after-primer-trimming">4.3.3 Visualise read count after primer trimming</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-filtering">5. Quality filtering</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#sample-processing">5.1 Sample processing</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#visualising-results">5.2 Visualising results</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#quality-scores">5.2.1 Quality scores</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#visualise-read-counts">5.2.2 Visualise read counts</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-error-rates">6. Learning error rates</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#generating-asvs">7. Generating ASVs</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#denoising-the-data">7.1 Denoising the data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">7.2 Visualising results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#merging-forward-and-reverse-reads">8. Merging forward and reverse reads</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reconstruct-the-full-amplicon">8.1 Reconstruct the full amplicon</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">8.2 Visualising results</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-table">9. Count table</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#chimera-removal">10. Chimera removal</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#read-count-tracking">11. Read count tracking</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#saving-output-files">12. Saving output files</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gert-Jan Jeunen
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025-2030.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>