

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Phylogenetic tree &#8212; Japan eDNA bootcamp 2025</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'phylogenetictree';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Statistical analysis" href="statisticalanalysis.html" />
    <link rel="prev" title="Data curation" href="datacuration.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Japan eDNA bootcamp 2025 - Home"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Japan eDNA bootcamp 2025 - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Japan eDNA bootcamp 2025
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setting up the environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinformaticanalysis.html">Bioinformatic analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="taxonomyassignment.html">Taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="datacuration.html">Data curation</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Phylogenetic tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="statisticalanalysis.html">Statistical analysis</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Fphylogenetictree.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/phylogenetictree.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Phylogenetic tree</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-sequence-alignment">2. Multiple Sequence Alignment</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#beauti-2">3. BEAUti 2</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#beast-2">4. BEAST 2</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tracer">5. Tracer</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#treeannotator">6. TreeAnnotator</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#figtree">7. FigTree</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualisation-in-r">8. Visualisation in R</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="phylogenetic-tree">
<h1>Phylogenetic tree<a class="headerlink" href="#phylogenetic-tree" title="Permalink to this heading">#</a></h1>
<section id="introduction">
<h2>1. Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading">#</a></h2>
<p>Before conducting the statistical analysis, there is one last output file we can create to help us explore our data: a phylogenetic tree of the ASV sequences. Phylogenetic relatedness is commonly used to inform downstream analyses, especially the calculation of phylogeny-aware distnances.</p>
<p>There are many different ways of building phylogenetic trees, ranging from fast heuristic distance-based methods with potential low accuracy (neighbour-joining; NJ) to slow probabilistic model-based highly-accurate methods (Bayesian Inference) and options in between (maximum likelyhood [ML] and maximum parsimony [MP]). The discussion and amount of research conducted on best approaches is vast and, unfortunately, outside the scope of this tutorial. Today, we will show you how to build Bayesian phylogenetic trees using <a class="reference external" href="https://www.beast2.org/">BEAST 2</a> (<strong>B</strong>ayesian <strong>E</strong>volutionary <strong>A</strong>nalysis by <strong>S</strong>ampling <strong>T</strong>rees).</p>
<p>BEAST 2 is a cross-platform program for Bayesian phylogenetic analysis of molecular sequences. It estiamtes rooted, time-measured phylogenies using strict or relaxed molecular clock models. It can be used as a method of reocnstructing phylogenies, but is also a framework for testing evolutionary hypotheses without conditioning on a single tree topology. BEAST 2 uses Markov chain Monte Carlo (MCMC) to average over tree space, so that each tree is weighted proportional to its posterior probability. BEAST 2 includes a graphical user-interface for setting up standard analyses and a suit of programs for analysing the results.</p>
<div class="warning admonition">
<p class="admonition-title">R and GUI</p>
<p>While BEAST 2 comes with a graphical user interface (GUI), we will still have to conduct the first couple of steps of the analysis in R.</p>
</div>
</section>
<section id="multiple-sequence-alignment">
<h2>2. Multiple Sequence Alignment<a class="headerlink" href="#multiple-sequence-alignment" title="Permalink to this heading">#</a></h2>
<p>To start the analysis, we have to align all our ASV sequences through a multiple sequence alignment (MSA). There are multiple programs capable of creating multiple sequence alignments. Since our pipeline is mainly based in the R environment, we’ll use the <code class="docutils literal notranslate"><span class="pre">AlignSeqs()</span></code> function of the <a class="reference external" href="https://www2.decipher.codes/Tutorials.html">DECIPHER R package</a>.</p>
<p>First, let’s start RStudio, open a new R script, save it as <strong>phylogenetic_tree.R</strong> in the <code class="docutils literal notranslate"><span class="pre">bootcamp</span></code> directory, prepare the R environment.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## set up the working environment</span>
<span class="nf">setwd</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="nf">Sys.getenv</span><span class="p">(</span><span class="s">&quot;HOME&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;Desktop&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;bootcamp&quot;</span><span class="p">))</span>
<span class="nf">list.files</span><span class="p">()</span>

<span class="c1">## load necessary R packages</span>
<span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">pkg</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Biostrings&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;DECIPHER&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ape&quot;</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="n">character.only</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> </span><span class="nf">install.packages</span><span class="p">(</span><span class="n">pkg</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="n">character.only</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;version:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">packageVersion</span><span class="p">(</span><span class="n">pkg</span><span class="p">)),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Once we have loaded all the necessary R packages and set the current working directory, we can read the ASV sequence file (<strong>asvs_filtered.fasta</strong>) into R.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## read data into R using Biostrings</span>
<span class="n">sequences</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">readDNAStringSet</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;asvs_filtered.fasta&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>Next, we can generate the MSA using the <code class="docutils literal notranslate"><span class="pre">AlignSeqs()</span></code> function.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## generate alignment using DECIPHER</span>
<span class="n">alignment</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">AlignSeqs</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span><span class="w"> </span><span class="n">anchor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">iterations</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">refinements</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Determining</span> <span class="n">distance</span> <span class="n">matrix</span> <span class="n">based</span> <span class="n">on</span> <span class="n">shared</span> <span class="mi">8</span><span class="o">-</span><span class="n">mers</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">15.77</span> <span class="n">secs</span>

<span class="n">Clustering</span> <span class="n">into</span> <span class="n">groups</span> <span class="n">by</span> <span class="n">similarity</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">0.33</span> <span class="n">secs</span>

<span class="n">Aligning</span> <span class="n">Sequences</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">6.92</span> <span class="n">secs</span>

<span class="n">Iteration</span> <span class="mi">1</span> <span class="n">of</span> <span class="mi">5</span><span class="p">:</span>

<span class="n">Determining</span> <span class="n">distance</span> <span class="n">matrix</span> <span class="n">based</span> <span class="n">on</span> <span class="n">alignment</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">1.02</span> <span class="n">secs</span>

<span class="n">Reclustering</span> <span class="n">into</span> <span class="n">groups</span> <span class="n">by</span> <span class="n">similarity</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">0.25</span> <span class="n">secs</span>

<span class="n">Realigning</span> <span class="n">Sequences</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">2.51</span> <span class="n">secs</span>

<span class="n">Iteration</span> <span class="mi">2</span> <span class="n">of</span> <span class="mi">5</span><span class="p">:</span>

<span class="n">Determining</span> <span class="n">distance</span> <span class="n">matrix</span> <span class="n">based</span> <span class="n">on</span> <span class="n">alignment</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">1.04</span> <span class="n">secs</span>

<span class="n">Reclustering</span> <span class="n">into</span> <span class="n">groups</span> <span class="n">by</span> <span class="n">similarity</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">0.17</span> <span class="n">secs</span>

<span class="n">Realigning</span> <span class="n">Sequences</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">0.37</span> <span class="n">secs</span>

<span class="n">Iteration</span> <span class="mi">3</span> <span class="n">of</span> <span class="mi">5</span><span class="p">:</span>

<span class="n">Determining</span> <span class="n">distance</span> <span class="n">matrix</span> <span class="n">based</span> <span class="n">on</span> <span class="n">alignment</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">1.06</span> <span class="n">secs</span>

<span class="n">Reclustering</span> <span class="n">into</span> <span class="n">groups</span> <span class="n">by</span> <span class="n">similarity</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">0.18</span> <span class="n">secs</span>

<span class="n">Realigning</span> <span class="n">Sequences</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">0.2</span> <span class="n">secs</span>

<span class="n">Alignment</span> <span class="n">converged</span> <span class="o">-</span> <span class="n">skipping</span> <span class="n">remaining</span> <span class="n">iterations</span><span class="o">.</span>

<span class="n">Refining</span> <span class="n">the</span> <span class="n">alignment</span><span class="p">:</span>
  <span class="o">|===================================================================================================================|</span> <span class="mi">100</span><span class="o">%</span>

<span class="n">Time</span> <span class="n">difference</span> <span class="n">of</span> <span class="mf">0.2</span> <span class="n">secs</span>

<span class="n">Alignment</span> <span class="n">converged</span> <span class="o">-</span> <span class="n">skipping</span> <span class="n">remaining</span> <span class="n">refinements</span><span class="o">.</span>
</pre></div>
</div>
</div>
<p>Once we have created the MSA, we need to export the alignment. It is important to export the alignment in <strong>nexus format</strong>, as this is required for the next step. We can use the <code class="docutils literal notranslate"><span class="pre">write.nexus.data()</span></code> function of the <a class="reference external" href="https://cran.r-project.org/web/packages/ape/index.html">ape R package</a>.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## export alignment in nexus format using ape</span>
<span class="nf">write.nexus.data</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;6.phylotree&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;asvs_aligned.nex&quot;</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="beauti-2">
<h2>3. BEAUti 2<a class="headerlink" href="#beauti-2" title="Permalink to this heading">#</a></h2>
<p>After we exported the MSA into nexus format, we need to generate an xml document that can be imported into BEAST 2. Luckily, the BEAST 2 developers have created BEAUti 2 (<strong>B</strong>ayesian <strong>E</strong>volutionary <strong>A</strong>nalysis <strong>Uti</strong>lity), a GUI that can import a nexus document (the one we just created in the previous step: <strong>asvs_aligned.nex</strong>) and export the necessary xml document for BEAST 2. After opening the BEAUti 2 app, we need to click the <code class="docutils literal notranslate"><span class="pre">File</span> <span class="pre">--&gt;</span> <span class="pre">Import</span> <span class="pre">Alignment</span></code> button. Next, we need to navigate to our nexus alignment and click on <code class="docutils literal notranslate"><span class="pre">Open</span></code>.</p>
<figure class="align-default" id="beauti-alignment-import">
<img alt="_images/beauti_import_alignment.png" src="_images/beauti_import_alignment.png" />
<figcaption>
<p><span class="caption-number">Fig. 51 </span><span class="caption-text">: Screenshot showing the process of importing a nexus alignment into BEAUti 2.</span><a class="headerlink" href="#beauti-alignment-import" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Once the nexus document is imported, we can change the name under <code class="docutils literal notranslate"><span class="pre">Clock</span> <span class="pre">Model</span></code> and <code class="docutils literal notranslate"><span class="pre">Tree</span></code> in the <code class="docutils literal notranslate"><span class="pre">Partitions</span></code> panel (starting window after import). We will set the names from the file name to <code class="docutils literal notranslate"><span class="pre">clock</span></code> and <code class="docutils literal notranslate"><span class="pre">tree</span></code> respectively to keep things simple.</p>
<figure class="align-default" id="beauti-change-names">
<img alt="_images/beauti_change_names.png" src="_images/beauti_change_names.png" />
<figcaption>
<p><span class="caption-number">Fig. 52 </span><span class="caption-text">: Screenshot showing how to change names in BEAUti 2.</span><a class="headerlink" href="#beauti-change-names" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Next, we will change the substitution model (<code class="docutils literal notranslate"><span class="pre">Subst</span> <span class="pre">Model</span></code>) in the <code class="docutils literal notranslate"><span class="pre">Site</span> <span class="pre">Model</span></code> panel from <code class="docutils literal notranslate"><span class="pre">JC69</span></code> to <code class="docutils literal notranslate"><span class="pre">HKY</span></code> and leave the <code class="docutils literal notranslate"><span class="pre">Kappa</span></code> and <code class="docutils literal notranslate"><span class="pre">Frequencies</span></code> variables to the default settings.</p>
<figure class="align-default" id="beauti-substitution-model">
<img alt="_images/beauti_substitution_model.png" src="_images/beauti_substitution_model.png" />
<figcaption>
<p><span class="caption-number">Fig. 53 </span><span class="caption-text">: Screenshot showing how to change the substitution model in BEAUti 2.</span><a class="headerlink" href="#beauti-substitution-model" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Then, we will set the starting tree to <code class="docutils literal notranslate"><span class="pre">Cluster</span> <span class="pre">Tree</span></code> and cluster type to <code class="docutils literal notranslate"><span class="pre">UPGMA</span></code> in the <code class="docutils literal notranslate"><span class="pre">Starting</span> <span class="pre">tree</span></code> panel, while leaving the remaining variables to the default values. This panel is not shown by default in BEAUti 2 and requires us to click on the <code class="docutils literal notranslate"><span class="pre">View</span> <span class="pre">--&gt;</span> <span class="pre">Show</span> <span class="pre">Starting</span> <span class="pre">tree</span> <span class="pre">panel</span></code> button.</p>
<figure class="align-default" id="beauti-show-starting-tree-panel">
<img alt="_images/beauti_show_starting_tree_panel.png" src="_images/beauti_show_starting_tree_panel.png" />
<figcaption>
<p><span class="caption-number">Fig. 54 </span><span class="caption-text">: Screenshot showing how to show the starting tree panel in BEAUti 2.</span><a class="headerlink" href="#beauti-show-starting-tree-panel" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>After enabling the <code class="docutils literal notranslate"><span class="pre">Starting</span> <span class="pre">tree</span></code> panel, we can set the starting tree to <code class="docutils literal notranslate"><span class="pre">Cluster</span> <span class="pre">Tree</span></code> and Cluster Type to <code class="docutils literal notranslate"><span class="pre">upgma</span></code>.</p>
<figure class="align-default" id="beauti-set-starting-tree">
<img alt="_images/beauti_set_starting_tree.png" src="_images/beauti_set_starting_tree.png" />
<figcaption>
<p><span class="caption-number">Fig. 55 </span><span class="caption-text">: Screenshot showing how to set the starting tree to Cluster Tree and type to UPGMA in BEAUti 2.</span><a class="headerlink" href="#beauti-set-starting-tree" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The final setting we will change is the <code class="docutils literal notranslate"><span class="pre">Chain</span> <span class="pre">Length</span></code> in the <code class="docutils literal notranslate"><span class="pre">MCMC</span></code> panel, which we will set to 10^8 (add one more 0 at the end of the default value).</p>
<figure class="align-default" id="beauti-set-chain-length">
<img alt="_images/beauti_set_chain_length.png" src="_images/beauti_set_chain_length.png" />
<figcaption>
<p><span class="caption-number">Fig. 56 </span><span class="caption-text">: Screenshot showing how to set the chain length in BEAUti 2.</span><a class="headerlink" href="#beauti-set-chain-length" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Once the <code class="docutils literal notranslate"><span class="pre">Chain</span> <span class="pre">Length</span></code> is set, we can export the data to an <strong>xml file format</strong> by pressing the <code class="docutils literal notranslate"><span class="pre">File</span> <span class="pre">--&gt;</span> <span class="pre">Save</span></code> button.</p>
<figure class="align-default" id="beauti-save-xml">
<img alt="_images/beauti_save_xml.png" src="_images/beauti_save_xml.png" />
<figcaption>
<p><span class="caption-number">Fig. 57 </span><span class="caption-text">: Screenshot showing how to save the xml file in BEAUti 2.</span><a class="headerlink" href="#beauti-save-xml" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>For this tutorial, we will save the xml file in the <code class="docutils literal notranslate"><span class="pre">6.phylotree</span></code> subdirectory and save the file as <strong>phylo_asvs.xml</strong>.</p>
<figure class="align-default" id="beauti-save-location">
<img alt="_images/beauti_save_location.png" src="_images/beauti_save_location.png" />
<figcaption>
<p><span class="caption-number">Fig. 58 </span><span class="caption-text">: Screenshot showing the save location of the xml file for this tutorial.</span><a class="headerlink" href="#beauti-save-location" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="beast-2">
<h2>4. BEAST 2<a class="headerlink" href="#beast-2" title="Permalink to this heading">#</a></h2>
<p>After we’ve created the xml file using BEAUti 2, we can start the BEAST 2 application.</p>
<p>BEAST 2 requires the xml file as <code class="docutils literal notranslate"><span class="pre">Input</span> <span class="pre">file</span></code>. We can also set the <code class="docutils literal notranslate"><span class="pre">Random</span> <span class="pre">number</span> <span class="pre">seed</span></code> to <code class="docutils literal notranslate"><span class="pre">2025</span></code>. Finally we can set the <code class="docutils literal notranslate"><span class="pre">Threads</span></code> and <code class="docutils literal notranslate"><span class="pre">Instances</span></code> within the <code class="docutils literal notranslate"><span class="pre">Run</span> <span class="pre">options</span></code> to the maximum values of <code class="docutils literal notranslate"><span class="pre">20</span></code> for each.</p>
<div class="warning admonition">
<p class="admonition-title">Random number seed</p>
<p>It is important to remember and report the <code class="docutils literal notranslate"><span class="pre">Random</span> <span class="pre">number</span> <span class="pre">seed</span></code> to make results reproducible.</p>
</div>
<div class="tip admonition">
<p class="admonition-title">Beagle</p>
<p>I highly recommend installing Beagle when using BEAST 2 to generate phylogenetic trees. Creating Bayesian trees is a very computationally intensive process and can take a long time to complete (hours to days). Beagle allows BEAST to make use of the GPU cores on your system and significantly reduces computation times.</p>
</div>
<figure class="align-default" id="beast-settings">
<img alt="_images/beast_settings.png" src="_images/beast_settings.png" />
<figcaption>
<p><span class="caption-number">Fig. 59 </span><span class="caption-text">: Screenshot showing the BEAST 2 settings.</span><a class="headerlink" href="#beast-settings" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Once you have set the correct settings for BEAST 2, we can start the process by pressing the <code class="docutils literal notranslate"><span class="pre">Run</span></code> button.</p>
<div class="tip admonition">
<p class="admonition-title">BEAST 2 output files</p>
<p>Since BEAST 2 will take several hours to complete for this dataset, we have provided all output files in the <code class="docutils literal notranslate"><span class="pre">6.phylotree</span></code> subdirectory, including <strong>phylo_asvs.log</strong> (BEAST 2 log file to check results) and <strong>phylo_asvs-tree.trees</strong> (output trees generated by BEAST 2).</p>
</div>
</section>
<section id="tracer">
<h2>5. Tracer<a class="headerlink" href="#tracer" title="Permalink to this heading">#</a></h2>
<p>After BEAST 2 completes, we can check the log file using the <a class="reference external" href="https://www.beast2.org/tracer-2/">Tracer</a> application. Tracer is a graphical tool for visualisation and diagnostics of MCMC (BEAST 2) output. A full explanation on the interpretation of the graphs is, unfortunately, outside the scope of this workshop. For now, we will only focus on the <code class="docutils literal notranslate"><span class="pre">ESS</span></code> values in the left-side panel. Ideally, they should be above 200.</p>
<p>After opening the Tracer application, we can click the <code class="docutils literal notranslate"><span class="pre">File</span> <span class="pre">--&gt;</span> <span class="pre">Import</span> <span class="pre">Trace</span> <span class="pre">File</span></code> button.</p>
<figure class="align-default" id="import-trace">
<img alt="_images/import_trace.png" src="_images/import_trace.png" />
<figcaption>
<p><span class="caption-number">Fig. 60 </span><span class="caption-text">: Screenshot showing the import trace file in Tracer.</span><a class="headerlink" href="#import-trace" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>We can navigate to the <code class="docutils literal notranslate"><span class="pre">6.phylotree</span></code> subdirectory and select the <strong>asvs_aligned.log</strong> file.</p>
<figure class="align-default" id="investigate-trace">
<img alt="_images/investigate_trace.png" src="_images/investigate_trace.png" />
<figcaption>
<p><span class="caption-number">Fig. 61 </span><span class="caption-text">: Screenshot showing the trace file results in Tracer.</span><a class="headerlink" href="#investigate-trace" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>From the results of the log file, we can see that multiple parameters, including very important ones such as <code class="docutils literal notranslate"><span class="pre">posterior</span></code>, <code class="docutils literal notranslate"><span class="pre">likelihood</span></code>, and <code class="docutils literal notranslate"><span class="pre">treeLikelihood</span></code>, have much lower ESS values than 200. Ideally, we would either run the MCMC for longer (current chain length: 3 x 10^7) or tweak some of the parameters (note: the current run took ~2 days on my laptop). We’ve included the trace file (<strong>asvs_aligned.xml.state</strong>), as well as the xml file (<strong>asvs_aligned.xml</strong>) in case you would like to optimise this tree. For the tutorial, we will go ahead with finding the best supported tree. <strong>However, please keep in mind that the tree might not be as accurate as it can be!</strong></p>
</section>
<section id="treeannotator">
<h2>6. TreeAnnotator<a class="headerlink" href="#treeannotator" title="Permalink to this heading">#</a></h2>
<p>Once we have checked the log file using Tracer, we can find the best supported tree using the <a class="reference external" href="https://www.beast2.org/treeannotator/">TreeAnnotator</a> application. TreeAnnotator assists in summarising the information from the sample of trees produced by BEAST 2 onto a single “target” tree. The summary information includes the posterior probabilities of the nodes in the target tree, the posterior estimates, and HPD limits of the node heights.</p>
<p>When opening TreeAnnotator, we will set the <code class="docutils literal notranslate"><span class="pre">Node</span> <span class="pre">heights</span></code> parameter to <code class="docutils literal notranslate"><span class="pre">Mean</span> <span class="pre">heights</span></code>, as well as select the <code class="docutils literal notranslate"><span class="pre">Input</span> <span class="pre">Tree</span> <span class="pre">File</span></code>, which is our <strong>asvs_aligned-tree.trees</strong> file. Finally, we will set the <code class="docutils literal notranslate"><span class="pre">Output</span> <span class="pre">File</span></code> parameter in the same folder with name <strong>asvs_aligned-tree.tree</strong> and select the <code class="docutils literal notranslate"><span class="pre">Low</span> <span class="pre">memory</span></code> option. Then, we can press the <code class="docutils literal notranslate"><span class="pre">Run</span></code> button.</p>
<figure class="align-default" id="treeannotator-settings">
<img alt="_images/treeannotator_settings.png" src="_images/treeannotator_settings.png" />
<figcaption>
<p><span class="caption-number">Fig. 62 </span><span class="caption-text">: Screenshot showing the TreeAnnotator settings.</span><a class="headerlink" href="#treeannotator-settings" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>After TreeAnnotator has finished finding the best supported tree, we can close the application.</p>
<figure class="align-default" id="treeannotator-results">
<img alt="_images/treeannotator_results.png" src="_images/treeannotator_results.png" />
<figcaption>
<p><span class="caption-number">Fig. 63 </span><span class="caption-text">: Screenshot showing the TreeAnnotator results.</span><a class="headerlink" href="#treeannotator-results" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="figtree">
<h2>7. FigTree<a class="headerlink" href="#figtree" title="Permalink to this heading">#</a></h2>
<p>After exporting the best-supported tree using TreeAnnotator, we can visualise the tree quality using <a class="reference external" href="https://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>.</p>
<p>After opening the FigTree application, we can click the <code class="docutils literal notranslate"><span class="pre">File</span> <span class="pre">--&gt;</span> <span class="pre">Open</span></code> button and select the <strong>asvs_aligned-tree.tree</strong> file that was generated by TreeAnnotator. Here, we can select <code class="docutils literal notranslate"><span class="pre">Node</span> <span class="pre">Labels</span></code> in the left hand-side column and set <code class="docutils literal notranslate"><span class="pre">Display</span></code> to <code class="docutils literal notranslate"><span class="pre">posterior</span></code> to determine how well branches are supported within the phylogenetic tree.</p>
<figure class="align-default" id="figtree-numbers">
<img alt="_images/figtree_numbers.png" src="_images/figtree_numbers.png" />
<figcaption>
<p><span class="caption-number">Fig. 64 </span><span class="caption-text">: Screenshot showing the posterior values of our phylogenetic tree in FigTree.</span><a class="headerlink" href="#figtree-numbers" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>However, given how many branches this phylogenetic tree has, it might be easier to set the <code class="docutils literal notranslate"><span class="pre">Colour</span> <span class="pre">by</span></code> parameter in <code class="docutils literal notranslate"><span class="pre">Appearance</span></code> to <code class="docutils literal notranslate"><span class="pre">posterior</span></code>.</p>
<figure class="align-default" id="figtree-colours">
<img alt="_images/figtree_colours.png" src="_images/figtree_colours.png" />
<figcaption>
<p><span class="caption-number">Fig. 65 </span><span class="caption-text">: Screenshot showing the posterior values of our phylogenetic tree in branch colours in FigTree.</span><a class="headerlink" href="#figtree-colours" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>As expected, given the log details we explored in the Tracer application, many branches show low posterior values and are, hence, not well supported. <strong>This is something we would definitely need to fix for a scientific publication, but let’s move forward for this tutorial.</strong></p>
</section>
<section id="visualisation-in-r">
<h2>8. Visualisation in R<a class="headerlink" href="#visualisation-in-r" title="Permalink to this heading">#</a></h2>
<p>While the Bayesian phylogenetic tree we created enables us to explore <strong>P</strong>hylogenetic <strong>D</strong>iversity (PD) in our data (a topic we will explore during the Statistical Analysis), phylogenetic trees are also frequently included as one of the visualisations in metabarcoding publications. Today, we will show you how to generate a circular phylogenetic tree and append other data types to the phylogenetic tree. This will enable us to visualise differences in ASV detection between sample locations. The R package we will be using to generate this figure is <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/ggtree.html">ggtree</a>, a fantastic R package developed by Prof. Guangchaung Yu. We will only scratch the surface of what this R package is capable of, so for more information, I recommend <a class="reference external" href="https://yulab-smu.top/treedata-book/author.html">this book</a> and the <a class="reference external" href="https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.12628">accompanying publication</a> to start with.</p>
<p>First, let’s open a new R script and save it as <strong>phylotree_visualisation.R</strong> in the <code class="docutils literal notranslate"><span class="pre">bootcamp</span></code> directory. We can, then, set the working environment, load the necessary R packages, and read the data into R.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## set up the working environment</span>
<span class="nf">setwd</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="nf">Sys.getenv</span><span class="p">(</span><span class="s">&quot;HOME&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;Desktop&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;bootcamp&quot;</span><span class="p">))</span>
<span class="nf">list.files</span><span class="p">()</span>

<span class="c1">## load necessary R packages</span>
<span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">pkg</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;ape&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;tidyverse&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ggtree&quot;</span><span class="p">))</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="kr">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="n">character.only</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> </span><span class="nf">install.packages</span><span class="p">(</span><span class="n">pkg</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="n">character.only</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="w">  </span><span class="nf">cat</span><span class="p">(</span><span class="n">pkg</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;version:&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">packageVersion</span><span class="p">(</span><span class="n">pkg</span><span class="p">)),</span><span class="w"> </span><span class="s">&quot;\n&quot;</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1">## read data into R</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;metadata_filtered.txt&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;\t&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">comment.char</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">count_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.table</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;count_table_filtered.txt&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;\t&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">comment.char</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;&#39;</span><span class="p">)</span>
<span class="n">phylo_tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.nexus</span><span class="p">(</span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;6.phylotree&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;asvs_aligned-tree.tree&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>To plot presence-absence data per location on the phylogenetic tree, we first have to reformat the data, by grouping samples per location and transforming the data to presence-absence. Then, we have to set 0 values to NA and set 1 values to the location name. This last step is essential for creating the heatmap we will plot over the phylogenetic tree.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## reformat data</span>
<span class="n">metadata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">metadata</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span><span class="n">sample</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">.</span><span class="p">))</span>
<span class="n">count_location</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">count_table</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">t</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">rownames_to_column</span><span class="p">(</span><span class="s">&quot;sample&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">left_join</span><span class="p">(</span><span class="n">metadata</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">select</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span><span class="w"> </span><span class="n">sampleLocation</span><span class="p">),</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;sample&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">select</span><span class="p">(</span><span class="o">-</span><span class="n">sample</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">sampleLocation</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">summarise</span><span class="p">(</span><span class="nf">across</span><span class="p">(</span><span class="nf">everything</span><span class="p">(),</span><span class="w"> </span><span class="n">sum</span><span class="p">,</span><span class="w"> </span><span class="n">na.rm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">mutate</span><span class="p">(</span><span class="nf">across</span><span class="p">(</span><span class="o">-</span><span class="n">sampleLocation</span><span class="p">,</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="nf">if_else</span><span class="p">(</span><span class="n">.x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)))</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">column_to_rownames</span><span class="p">(</span><span class="s">&quot;sampleLocation&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">t</span><span class="p">()</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">()</span>
<span class="n">count_location</span><span class="p">[</span><span class="n">count_location</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NA</span>
<span class="n">column_names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">count_location</span><span class="p">)</span>
<span class="kr">for</span><span class="w"> </span><span class="p">(</span><span class="n">col</span><span class="w"> </span><span class="kr">in</span><span class="w"> </span><span class="n">column_names</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">  </span><span class="n">count_location</span><span class="p">[[</span><span class="n">col</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ifelse</span><span class="p">(</span><span class="n">count_location</span><span class="p">[[</span><span class="n">col</span><span class="p">]]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="p">,</span><span class="w"> </span><span class="n">count_location</span><span class="p">[[</span><span class="n">col</span><span class="p">]])</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Once we have the data formatted, we can plot the graph using the ggtree R package.</p>
<div class="highlight-R notranslate"><div class="highlight"><pre><span></span><span class="c1">## plot tree</span>
<span class="n">sample_colors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Akakina&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#EAE0DA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Asan&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#A0C3D2&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Ashiken&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#E7C8C0&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Edateku&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#B5C7A4&quot;</span><span class="p">,</span>
<span class="w">                   </span><span class="s">&quot;Gamo&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#D4BBDD&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Heda&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#F2D7A7&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Imaizaki&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#A2D9D1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Kurasaki&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#DFA7A7&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Sakibaru&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#C5D0E6&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                   </span><span class="s">&quot;Sani&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#D9C5A0&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Taen&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#B7D9C4&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Tatsugo&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#E3B7D1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Yadon&quot;</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;#A8C6D9&quot;</span><span class="p">)</span>
<span class="n">circ1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggtree</span><span class="p">(</span><span class="n">phylo_tree</span><span class="p">,</span><span class="w"> </span><span class="n">layout</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;fan&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">open.angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;grey30&quot;</span><span class="p">)</span>
<span class="n">circ2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rotate_tree</span><span class="p">(</span><span class="n">circ1</span><span class="p">,</span><span class="w"> </span><span class="m">280</span><span class="p">)</span>
<span class="n">circ3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">gheatmap</span><span class="p">(</span><span class="n">circ2</span><span class="p">,</span><span class="w"> </span><span class="n">count_location</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">offset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">-0.0272</span><span class="p">,</span>
<span class="w">                  </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">font.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sample_colors</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;top&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1.385</span><span class="p">,</span><span class="w"> </span><span class="m">1.45</span><span class="p">,</span><span class="w"> </span><span class="m">1.51</span><span class="p">,</span><span class="w"> </span><span class="m">1.575</span><span class="p">,</span><span class="w"> </span><span class="m">1.64</span><span class="p">,</span><span class="w"> </span><span class="m">1.705</span><span class="p">,</span><span class="w"> </span><span class="m">1.77</span><span class="p">,</span>
<span class="w">                            </span><span class="m">1.83</span><span class="p">,</span><span class="w"> </span><span class="m">1.895</span><span class="p">,</span><span class="w"> </span><span class="m">1.955</span><span class="p">,</span><span class="w"> </span><span class="m">2.02</span><span class="p">,</span><span class="w"> </span><span class="m">2.085</span><span class="p">),</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.1</span><span class="p">)</span>
<span class="n">circ3</span>
<span class="nf">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">file.path</span><span class="p">(</span><span class="s">&quot;4.results&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;phylogenetic_tree.png&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">circ3</span><span class="p">,</span><span class="w"> </span><span class="n">bg</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;white&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span><span class="w"> </span>
</pre></div>
</div>
<figure class="align-default" id="id1">
<img alt="_images/phylogenetic_tree.png" src="_images/phylogenetic_tree.png" />
<figcaption>
<p><span class="caption-number">Fig. 66 </span><span class="caption-text">: Phylogenetic tree with overlaid heatmap for ASV detection per location using the ggtree R package.</span><a class="headerlink" href="#id1" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<div class="tip admonition">
<p class="admonition-title">Additional customisation</p>
<p>Maybe try yourself to change this phylogenetic tree plot by colouring branches based on the assigned phylum. When trying, it would be best to subset the data and prune the tree to only include high confidence BLAST-hits.</p>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="datacuration.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Data curation</p>
      </div>
    </a>
    <a class="right-next"
       href="statisticalanalysis.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Statistical analysis</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multiple-sequence-alignment">2. Multiple Sequence Alignment</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#beauti-2">3. BEAUti 2</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#beast-2">4. BEAST 2</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#tracer">5. Tracer</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#treeannotator">6. TreeAnnotator</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#figtree">7. FigTree</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#visualisation-in-r">8. Visualisation in R</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gert-Jan Jeunen
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025-2030.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>