

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Taxonomy assignment &#8212; Japan eDNA bootcamp 2025</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'taxonomyassignment';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Data curation" href="datacuration.html" />
    <link rel="prev" title="Bioinformatic analysis" href="bioinformaticanalysis.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Japan eDNA bootcamp 2025 - Home"/>
    <script>document.write(`<img src="_static/logo.png" class="logo__image only-dark" alt="Japan eDNA bootcamp 2025 - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Japan eDNA bootcamp 2025
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setting up the environment</a></li>
<li class="toctree-l1"><a class="reference internal" href="bioinformaticanalysis.html">Bioinformatic analysis</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Taxonomy assignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="datacuration.html">Data curation</a></li>
<li class="toctree-l1"><a class="reference internal" href="phylogenetictree.html">Phylogenetic tree</a></li>
<li class="toctree-l1"><a class="reference internal" href="statisticalanalysis.html">Statistical analysis</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2Ftaxonomyassignment.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/taxonomyassignment.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Taxonomy assignment</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#online-data-repositories">2. Online data repositories</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#crabs">3. CRABS</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow">3.1 Workflow</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#help-documentation">3.2 Help documentation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-download-data">3.3 Step 1: download data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-import-data">3.4 Step 2: import data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-extract-amplicons-through-in-silico-pcr">3.5 Step 3: extract amplicons through <em>in silico</em> PCR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-extract-amplicons-through-pairwise-global-alignments">3.6 Step 4: extract amplicons through pairwise global alignments</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-dereplicate-the-database">3.7 Step 5: dereplicate the database</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-filter-the-reference-database">3.8 Step 6: filter the reference database</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-7-export-the-reference-database">3.9 Step 7: export the reference database</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-coi-reference-database">4. The COI reference database</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#blast">5. BLAST</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#online-blast">5.1 Online BLAST</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#local-blast">5.2 Local BLAST</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#blast-output-parsing">5.3 BLAST output parsing</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="taxonomy-assignment">
<h1>Taxonomy assignment<a class="headerlink" href="#taxonomy-assignment" title="Permalink to this heading">#</a></h1>
<section id="introduction">
<h2>1. Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading">#</a></h2>
<p>As we have previously discussed, eDNA metabarcoding is the process whereby we amplify a specific gene region of the taxonomic group of interest from a DNA extract obtained from an environmental sample. Once the DNA is amplified, a library is prepared for sequencing. Up to this point, we have prepared the data for the bioinformatic pipeline and executed the <a class="reference external" href="https://benjjneb.github.io/dada2/">dada2</a> bioinformatic pipeline. Yesterday, we ended up with a count table (<strong>count_table.txt</strong>) and a list of ASV sequences (<strong>asvs.fasta</strong>) containing all the biologically-relevant sequences.</p>
<p>The next step in the bioinformatic pipeline is to assign a taxonomic ID to each ASV to determine what species were detected through our eDNA metabarcoding analysis.</p>
<p>There are 4 basic strategies to taxonomy classification according to <a class="reference external" href="https://onlinelibrary.wiley.com/doi/epdf/10.1111/1755-0998.13407">Hleap et al., 2021</a>, as well as an endless variations on each of these four strategies:</p>
<ol class="arabic simple">
<li><p><strong>Sequence Similarity</strong> (SS)</p></li>
<li><p><strong>Sequence Composition</strong> (SC)</p></li>
<li><p><strong>Phylogenetic</strong> (Ph)</p></li>
<li><p><strong>Probabilistic</strong> (Pr)</p></li>
</ol>
<figure class="align-default" id="taxonomy-strategies">
<img alt="_images/taxonomy_strategies.png" src="_images/taxonomy_strategies.png" />
<figcaption>
<p><span class="caption-number">Fig. 29 </span><span class="caption-text">: The four basic strategies to taxonomy classification. Copyright: <a class="reference external" href="https://onlinelibrary.wiley.com/doi/epdf/10.1111/1755-0998.13407">Hleap et al., 2021</a></span><a class="headerlink" href="#taxonomy-strategies" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Of the four abovementioned strategies to taxonomy classification, <strong>Sequency Similarity</strong> and <strong>Sequence Composition</strong> are most-frequently used in the metabarcoding research community. Multiple comparative experiments have been conducted to determine the most optimal approach to assign a taxonomic ID to metabarcoding data, though an optimal method is yet to be found. One example of a comparative study is the one from <a class="reference external" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0470-z">Bokulich et al., 2018</a>.</p>
<figure class="align-default" id="taxonomy-comparison">
<img alt="_images/taxonomy_comparison.png" src="_images/taxonomy_comparison.png" />
<figcaption>
<p><span class="caption-number">Fig. 30 </span><span class="caption-text">: A comparison of taxonomy classifiers. Copyright: <a class="reference external" href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0470-z">Bokulich et al., 2018</a></span><a class="headerlink" href="#taxonomy-comparison" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>While there are multiple ways to assign a taxonomic ID to a sequence, all methods require a reference database to match ASVs against a set of sequences with known taxonomic IDs. So, before we start with taxonomy assignment, let’s create our own local curated reference database using <a class="reference external" href="https://github.com/gjeunen/reference_database_creator">CRABS</a>, which stands for <strong>C</strong>reating <strong>R</strong>eference <strong>d</strong>atabases for <strong>A</strong>mplicon-<strong>B</strong>ased <strong>S</strong>equencing. For more information on the sofware, I recommend reading <a class="reference external" href="https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13741">the publication</a>.</p>
</section>
<section id="online-data-repositories">
<h2>2. Online data repositories<a class="headerlink" href="#online-data-repositories" title="Permalink to this heading">#</a></h2>
<p>Several reference databases are available online. The most notable ones being NCBI, EMBL, BOLD, SILVA, and RDP. However, recent research has indicated a need to use custom curated reference databases to increase the accuracy of taxonomy assignment. While certain gene-specific or primer-specific reference databases are available (RDP: 16S microbial; MIDORI: COI eukaryotes; etc.), this essential data source is missing in most instances. We will, therefore, show you how to build your own custom curated reference database using CRABS.</p>
<div class="attention admonition">
<p class="admonition-title">Attention</p>
<p>Since the online COI reference material is quite large, it being the standard metazoan barcode gene, we will not build a COI reference database during this tutorial. Instead, we will use a pre-assembled COI database, which is already placed in your <code class="docutils literal notranslate"><span class="pre">5.refdb</span></code> folder. However, to show you how to build your own database, we will go through the steps to build a 16S Chondrichthyes (sharks and rays) database, as the online reference material is small. Therefore, the steps can be accomplished quite quickly.</p>
</div>
</section>
<section id="crabs">
<h2>3. CRABS<a class="headerlink" href="#crabs" title="Permalink to this heading">#</a></h2>
<section id="workflow">
<h3>3.1 Workflow<a class="headerlink" href="#workflow" title="Permalink to this heading">#</a></h3>
<p>The current version of CRABS v1.7.7 includes 7 steps, namely:</p>
<ol class="arabic simple">
<li><p>download data from online repositories,</p></li>
<li><p>import downloaded data into CRABS format,</p></li>
<li><p>extract amplicons from imported data,</p></li>
<li><p>retrieve amplicons without primer-binding regions,</p></li>
<li><p>curate and subset the local database,</p></li>
<li><p>export the local database in various taxonomic classifier formats, and</p></li>
<li><p>basic visualisations to explore the local reference database.</p></li>
</ol>
<figure class="align-default" id="crabs-workflow">
<img alt="_images/crabs_workflow.png" src="_images/crabs_workflow.png" />
<figcaption>
<p><span class="caption-number">Fig. 31 </span><span class="caption-text">: The CRABS workflow to create your own reference database. Copyright: <a class="reference external" href="https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13741">Jeunen et al., 2023</a></span><a class="headerlink" href="#crabs-workflow" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="help-documentation">
<h3>3.2 Help documentation<a class="headerlink" href="#help-documentation" title="Permalink to this heading">#</a></h3>
<p>To bring up the help documentation for CRABS, we can use the <code class="docutils literal notranslate"><span class="pre">--help</span></code> or <code class="docutils literal notranslate"><span class="pre">-h</span></code> command.</p>
<div class="warning admonition">
<p class="admonition-title">Running CLI programs in RStudio</p>
<p>CRABS is a CLI program written in Python, similar to cutadapt. However, rather than the <code class="docutils literal notranslate"><span class="pre">system2()</span></code> function we employed yesterday to run cutadapt, we will execute CRABS’ code into the Terminal window, rather than the Console window. For the remainder of this section, all code will be run in this fashion, unless stated otherwise.</p>
<p>So, let’s start RStudio and open a new text file, rather than an R script, and save it as <strong>taxonomy_assignment.txt</strong> in the <code class="docutils literal notranslate"><span class="pre">bootcamp</span></code> folder.</p>
<figure class="align-default" id="rstudio-text">
<img alt="_images/RStudio_text.png" src="_images/RStudio_text.png" />
<figcaption>
<p><span class="caption-number">Fig. 32 </span><span class="caption-text">: Opening up a text file in RStudio</span><a class="headerlink" href="#rstudio-text" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Now, let’s copy-paste the code to bring up the help documentation into the text file and run it in the Terminal window. You can run the code either by copy-pasting it in the Terminal window or by pressing <code class="docutils literal notranslate"><span class="pre">command</span> <span class="pre">+</span> <span class="pre">option</span> <span class="pre">+</span> <span class="pre">return</span></code> for MacOS and <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">alt</span> <span class="pre">+</span> <span class="pre">enter</span></code> for Windows and Linux.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>-h
</pre></div>
</div>
<figure class="align-default" id="crabs-help">
<img alt="_images/crabs_help.png" src="_images/crabs_help.png" />
<figcaption>
<p><span class="caption-number">Fig. 33 </span><span class="caption-text">: The CRABS help documentation</span><a class="headerlink" href="#crabs-help" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</div>
</section>
<section id="step-1-download-data">
<h3>3.3 Step 1: download data<a class="headerlink" href="#step-1-download-data" title="Permalink to this heading">#</a></h3>
<p>As a first step to create your local custom reference database, you will need to download data from online repositories or use barcode sequences that you have generated yourself. CRABS can download data from the 7 most widely-used online repositories, including BOLD, EMBL, GreenGenes, MIDORI2, MitoFish, NCBI, and SILVA.</p>
<p>For this tutorial, however, we will only focus on downloading shark sequences of the 16S gene from the NCBI online repository, as this database is the most widely-used in metabarcoding research, as well as the biggest online repository. To download data from the NCBI servers, we need to provide several parameters to CRABS. Maybe the trickiest to understand is the <code class="docutils literal notranslate"><span class="pre">--query</span></code> parameter. However, we can get the input string for this parameter from the NCBI website, as shown in the screenshot below.</p>
<figure class="align-default" id="crabs-ncbi-download">
<img alt="_images/crabs_ncbi_download.png" src="_images/crabs_ncbi_download.png" />
<figcaption>
<p><span class="caption-number">Fig. 34 </span><span class="caption-text">: A screenshot of the NCBI website showing where to find the input value for the <code class="docutils literal notranslate"><span class="pre">--query</span></code> parameter</span><a class="headerlink" href="#crabs-ncbi-download" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>Other parameters that are required are an output file name (<code class="docutils literal notranslate"><span class="pre">--output</span></code>), your email for identification by NCBI’s servers (<code class="docutils literal notranslate"><span class="pre">--email</span></code>), and the database you wish to download from (<code class="docutils literal notranslate"><span class="pre">--database</span></code>), which for metabarcoding data is usually <code class="docutils literal notranslate"><span class="pre">nucleotide</span></code>.</p>
<p>Before we can execute the code, we need to set our working directory. However, since we are working in the Terminal window, we need to provide slightly different commands, as this is not an R environment. To print the current working directory, we can use the <code class="docutils literal notranslate"><span class="pre">pwd</span></code> command. You should now see an output string in the Terminal window telling you the current working directory.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">pwd</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">gjeunen</span>
</pre></div>
</div>
</div>
<figure class="align-default" id="print-working-directory">
<img alt="_images/print_working_directory.png" src="_images/print_working_directory.png" />
<figcaption>
<p><span class="caption-number">Fig. 35 </span><span class="caption-text">: A screenshot of RStudio when running the <code class="docutils literal notranslate"><span class="pre">pwd</span></code> command in the Terminal window, executed by pressing <code class="docutils literal notranslate"><span class="pre">command</span> <span class="pre">+</span> <span class="pre">option</span> <span class="pre">+</span> <span class="pre">return</span></code> for MacOS or <code class="docutils literal notranslate"><span class="pre">ctrl</span> <span class="pre">+</span> <span class="pre">alt</span> <span class="pre">+</span> <span class="pre">enter</span></code> for Linux and Windows.</span><a class="headerlink" href="#print-working-directory" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>To change the current working directory, we can use the <code class="docutils literal notranslate"><span class="pre">cd</span></code> command, which stands for <strong>change directory</strong>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span><span class="w"> </span>Desktop/bootcamp/5.refdb/
</pre></div>
</div>
<div class="warning admonition">
<p class="admonition-title">Navigating directories in the Terminal</p>
<p>The code above is for MacOS and Linux, slight alterations might be required for Windows users!</p>
</div>
<p>Now that we have navigated to the <code class="docutils literal notranslate"><span class="pre">5.refdb</span></code> folder, let’s print the current working directory again to make sure we’re in the folder we expect.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">pwd</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">gjeunen</span><span class="o">/</span><span class="n">Desktop</span><span class="o">/</span><span class="n">bootcamp</span><span class="o">/</span><span class="mf">5.</span><span class="n">refdb</span>
</pre></div>
</div>
</div>
<p>We can also list the current files using the <code class="docutils literal notranslate"><span class="pre">ls</span></code> command, which stands for <strong>list</strong>. We can provide a couple of additional parameters to this command to format the output.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ls<span class="w"> </span>-ltr
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>    <span class="mi">17850320</span> <span class="n">Apr</span>  <span class="mi">1</span> <span class="mi">04</span><span class="p">:</span><span class="mi">26</span> <span class="n">taxdb</span><span class="o">.</span><span class="n">bti</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>   <span class="mi">169532428</span> <span class="n">Apr</span>  <span class="mi">1</span> <span class="mi">04</span><span class="p">:</span><span class="mi">26</span> <span class="n">taxdb</span><span class="o">.</span><span class="n">btd</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>    <span class="mi">85700608</span> <span class="n">Apr</span>  <span class="mi">1</span> <span class="mi">04</span><span class="p">:</span><span class="mi">26</span> <span class="n">taxonomy4blast</span><span class="o">.</span><span class="n">sqlite3</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>   <span class="mi">201193865</span> <span class="n">Apr</span>  <span class="mi">2</span> <span class="mi">19</span><span class="p">:</span><span class="mi">26</span> <span class="n">nodes</span><span class="o">.</span><span class="n">dmp</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>   <span class="mi">265924736</span> <span class="n">Apr</span>  <span class="mi">2</span> <span class="mi">19</span><span class="p">:</span><span class="mi">27</span> <span class="n">names</span><span class="o">.</span><span class="n">dmp</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>  <span class="mi">1129185908</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">02</span><span class="p">:</span><span class="mi">32</span> <span class="n">coi_clean</span><span class="o">.</span><span class="n">txt</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>   <span class="mi">207099442</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="n">nsq</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>    <span class="mi">10683728</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="n">nog</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>    <span class="mi">32051236</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="n">nin</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>   <span class="mi">158567715</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="n">nhr</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>    <span class="mi">12942204</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="n">nto</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>    <span class="mi">50335744</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="n">ntf</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>    <span class="mi">32051096</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="ow">not</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>    <span class="mi">67982376</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="n">nos</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>   <span class="mi">132980736</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">03</span><span class="p">:</span><span class="mi">15</span> <span class="n">coi_blast</span><span class="o">.</span><span class="n">ndb</span>
<span class="o">-</span><span class="n">rw</span><span class="o">-</span><span class="n">r</span><span class="o">--</span><span class="n">r</span><span class="o">--</span>  <span class="mi">1</span> <span class="n">gjeunen</span>  <span class="n">staff</span>  <span class="mi">2542038129</span> <span class="n">Apr</span>  <span class="mi">3</span> <span class="mi">12</span><span class="p">:</span><span class="mi">33</span> <span class="n">nucl_gb</span><span class="o">.</span><span class="n">accession2taxid</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
<p>You should see a list of 16 files, all related to the COI reference database we will be using later.</p>
<p>For now, let’s download the 16S shark barcodes from NCBI.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--download-ncbi<span class="w"> </span>--query<span class="w"> </span><span class="s1">&#39;(16S[All Fields] AND (&quot;Chondrichthyes&quot;[Organism] OR Chondrichthyes[All Fields])) AND mitochondrion[filter]&#39;</span><span class="w"> </span>--email<span class="w"> </span>gjeunen@gmail.com<span class="w"> </span>--output<span class="w"> </span>ncbi_sharks.fasta<span class="w"> </span>--database<span class="w"> </span>nucleotide
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
/// CRABS | v1.7.7

|            Function | Download NCBI database
|Retrieving NCBI info | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|         Downloading | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:01:13
|             Results | Number of sequences downloaded: 3868/3868 (100.0%)
</pre></div>
</div>
</div>
<p>Let’s look at the top 10 lines of the output file to see what CRABS downloaded. We can use the <code class="docutils literal notranslate"><span class="pre">head</span></code> command for this and pass the <code class="docutils literal notranslate"><span class="pre">-n</span> <span class="pre">10</span></code> parameter to specify 10 lines.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>ncbi_sharks.fasta
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">MZ562568</span><span class="mf">.1</span> <span class="n">Fontitrygon</span> <span class="n">garouaensis</span> <span class="n">isolate</span> <span class="n">GN19480</span> <span class="n">mitochondrion</span><span class="p">,</span> <span class="n">partial</span> <span class="n">genome</span>
<span class="n">GCTAGCGTAGCTTAAACAAAAGCATAGCATTGAAGATGCTAAGATAAAAATTAACCTTTTTCGCAAGCAT</span>
<span class="n">GAAGGTTTGGTCCTGGCCTCAATATTAGTTCTAACTTGATTTACACATGCAAGTCTCAGCATCCTGGTGA</span>
<span class="n">GAACGCCCTACTTAAACCAAAAATTTAGACAGGAGCTGGTATCAGGTACACCATACAGGTAGCCCACAAC</span>
<span class="n">ACCTCGCTCAGCCACACCCCCAAGGGAACTCAGCAGTGACTAACATTGTTCCATAAGCGTAAGCTTGAGC</span>
<span class="n">CAATCAAAGTTAAGAGAGTTGGTTAATCTCGTGCCAGCCACCGCGGTTATACGAGTAACACAAATTAATA</span>
<span class="n">TTTTACGGCGTTAAGGGTGATTAAAAATAAACTTCCCAAAAATAGAGTTAATACATCATCAAACCGTCAT</span>
<span class="n">ACGTTTACATGCTCAAAAACATCACTCACGAAAGTAACTCTACATAAAAAAGACTTTTTGACCTCACGAT</span>
<span class="n">AGTTAAGATCCAAACTAGGATTAGATACCCTATTATGCTTAACCATAAACATTGTCATAAAAATTTACCT</span>
<span class="n">TAATATTACCGCCCGAGTACTACGAGCGCTAGCTTAAAACCCAAAGGACTTGGCGGTGCTCCAAACCCCC</span>
</pre></div>
</div>
</div>
<p>From the output, we can see that CRABS has downloaded the original format from the NCBI servers, which in this case is a multi-line fasta document.</p>
</section>
<section id="step-2-import-data">
<h3>3.4 Step 2: import data<a class="headerlink" href="#step-2-import-data" title="Permalink to this heading">#</a></h3>
<p>Since CRABS can download data from multiple online repositories, each with their own format, we need to first reformat the data to CRABS format before we can continue building our reference database. This reformatting can be done using the <code class="docutils literal notranslate"><span class="pre">--import</span></code> function, which will create a single line text document for each sequence, that contains a unique ID, the taxonomic lineage for that sequence, plus the sequence itself. The CRABS taxonomic lineage is based on the NCBI taxonomy database and requires 3 files to be downloaded using the <code class="docutils literal notranslate"><span class="pre">--download-taxonomy</span></code> function. However, to speed things up, these 3 files have already been provided to you.</p>
<p>For those interested, you can download the three NCBI files using the following command (do not execute this command for the workshop, as the files are already in your folder!).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--download-taxonomy
</pre></div>
</div>
<p>Before importing, we need to unzip one of the three taxonomy files, as it was too big to share (note: this won’t be necessary when downloading the files through CRABS).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gunzip<span class="w"> </span>nucl_gb.accession2taxid.gz
</pre></div>
</div>
<p>Once unzipped, we can import the downloaded NCBI data into CRABS format. To import data, we have to provide several parameters, including the format <code class="docutils literal notranslate"><span class="pre">--import-format</span></code>, the three taxonomy files (<code class="docutils literal notranslate"><span class="pre">--names</span></code>, <code class="docutils literal notranslate"><span class="pre">--nodes</span></code>, and <code class="docutils literal notranslate"><span class="pre">--acc2tax</span></code>), the input file (<code class="docutils literal notranslate"><span class="pre">--input</span></code>), an output file (<code class="docutils literal notranslate"><span class="pre">--output</span></code>), and optionally the taxonomic ranks to be included in the taxonomic lineage (<code class="docutils literal notranslate"><span class="pre">--ranks</span></code>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--import<span class="w"> </span>--import-format<span class="w"> </span>ncbi<span class="w"> </span>--names<span class="w"> </span>names.dmp<span class="w"> </span>--nodes<span class="w"> </span>nodes.dmp<span class="w"> </span>--acc2tax<span class="w"> </span>nucl_gb.accession2taxid<span class="w"> </span>--input<span class="w"> </span>ncbi_sharks.fasta<span class="w"> </span>--output<span class="w"> </span>ncbi_sharks.txt
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
/// CRABS | v1.7.7

|            Function | Import sequence data into CRABS format
| Read data to memory | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:03:31
|Phylogenetic lineage | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|Fill missing lineage | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|             Results | Imported 3868 out of 3868 sequences into CRABS format (100.0%)
</pre></div>
</div>
</div>
<p>When we now look at the first line of the output file, we can see how CRABS has changed the structure from a multi-line fasta to a single line tab-delimited text file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">1</span><span class="w"> </span>ncbi_sharks.txt
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">MZ562568</span>        <span class="n">Fontitrygon</span> <span class="n">garouaensis</span> <span class="mi">3355173</span> <span class="n">Metazoa</span> <span class="n">Chordata</span>        <span class="n">Chondrichthyes</span>  <span class="n">Myliobatiformes</span> <span class="n">Dasyatidae</span>      <span class="n">Fontitrygon</span> <span class="n">Fontitrygon</span> <span class="n">garouaensis</span>  <span class="n">GCTAGCGTAGCTTAAACAAAAGCATAGCATTGAAGATGCTAAGATAAAAATTAACCTTTTTCGCAAGCATGAAGGTTTGGTCCTGGCCTCAATATTAGTTCTAACTTGATTTACACATGCAAGTCTCAGCATCCTGGTGAGAA</span><span class="o">...</span>
</pre></div>
</div>
</div>
</section>
<section id="step-3-extract-amplicons-through-in-silico-pcr">
<h3>3.5 Step 3: extract amplicons through <em>in silico</em> PCR<a class="headerlink" href="#step-3-extract-amplicons-through-in-silico-pcr" title="Permalink to this heading">#</a></h3>
<p>Once we have formatted the data, we can start with a first curation step of the reference database, i.e., extracting the amplicon sequences. We will be extracting the amplicon region from each sequence through an <em>in silico</em> PCR analysis. With this analysis, we will locate the forward and reverse primer binding regions and extract the sequence in between. This will significantly reduce file sizes when larger data sets are downloaded, while also ensuring that only the necessary information is kept.</p>
<p>We can use CRABS’ <code class="docutils literal notranslate"><span class="pre">--in-silico-pcr</span></code> function for this step, which requires us to provide information on the following parameters, the input file name (<code class="docutils literal notranslate"><span class="pre">--input</span></code>), the output file name (<code class="docutils literal notranslate"><span class="pre">--output</span></code>), the forward primer sequence (<code class="docutils literal notranslate"><span class="pre">--forward</span></code>), and the reverse primer sequence (<code class="docutils literal notranslate"><span class="pre">--reverse</span></code>). The primers for the 16S gene for which we are building a reference database are F: 5’-GACCCTATGGAGCTTTAGAC-3’ and R: 5’-CGCTGTTATCCCTADRGTAACT-3’. Note that we do not need to reverse complement the reverse primer when we provide this information to CRABS, who will do this for us before running the <em>in silico</em> PCR analysis.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--in-silico-pcr<span class="w"> </span>--input<span class="w"> </span>ncbi_sharks.txt<span class="w"> </span>--output<span class="w"> </span>ncbi_sharks_amplicon.txt<span class="w"> </span>--forward<span class="w"> </span>GACCCTATGGAGCTTTAGAC<span class="w"> </span>--reverse<span class="w"> </span>CGCTGTTATCCCTADRGTAACT
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>|  Transform to fasta | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|       In silico PCR | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|      Exporting data | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|             Results | Extracted 3529 amplicons from 3868 sequences (91.24%)
</pre></div>
</div>
</div>
<p>From the output, we can see that we downloaded and imported 3,868 16S shark sequences from NCBI and CRABS managed to find the primer binding regions in 3,529 (91.24%) of them. So, nearly all.</p>
</section>
<section id="step-4-extract-amplicons-through-pairwise-global-alignments">
<h3>3.6 Step 4: extract amplicons through pairwise global alignments<a class="headerlink" href="#step-4-extract-amplicons-through-pairwise-global-alignments" title="Permalink to this heading">#</a></h3>
<p>Amplicons in the originally downloaded sequences might be missed during the in silico PCR analysis when one or both primer-binding regions are not incorporated in the online deposited sequence. This can happen when the reference barcode is generated with the same primer set or if the deposited sequence is incomplete in the primer-binding regions (denoted as “N” in the sequence). To retrieve those missed amplicons, we can use the already-retrieved amplicons as seed sequences in a Pairwise Global Alignment (PGA) analysis.</p>
<p>We can use CRABS’ <code class="docutils literal notranslate"><span class="pre">--pairwise-global-alignment</span></code> function for this step, which requires an input file name (<code class="docutils literal notranslate"><span class="pre">--input</span></code>), an amplicon file name (<code class="docutils literal notranslate"><span class="pre">--amplicons</span></code>), an output file name (<code class="docutils literal notranslate"><span class="pre">--output</span></code>), the forward (<code class="docutils literal notranslate"><span class="pre">--forward</span></code>) and reverse (<code class="docutils literal notranslate"><span class="pre">--reverse</span></code>) primer sequences, the minimum threshold value for the alignment similarity (<code class="docutils literal notranslate"><span class="pre">--percent-identity</span></code>), and the minimum threshold value for the alignment coverage of the amplicon (<code class="docutils literal notranslate"><span class="pre">--coverage</span></code>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--pairwise-global-alignment<span class="w"> </span>--input<span class="w"> </span>ncbi_sharks.txt<span class="w"> </span>--amplicons<span class="w"> </span>ncbi_sharks_amplicon.txt<span class="w"> </span>--output<span class="w"> </span>ncbi_sharks_pga.txt<span class="w"> </span>--forward<span class="w"> </span>GACCCTATGGAGCTTTAGAC<span class="w"> </span>--reverse<span class="w"> </span>CGCTGTTATCCCTADRGTAACT<span class="w"> </span>--percent-identity<span class="w"> </span><span class="m">0</span>.9<span class="w"> </span>--coverage<span class="w"> </span><span class="m">90</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
/// CRABS | v1.7.7

|            Function | Retrieve amplicons without primer-binding regions
|         Import data | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|  Transform to fasta | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|  Pairwise alignment | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:-- 0:00:00
|             Results | Retrieved 42 amplicons without primer-binding regions from 339 sequences
</pre></div>
</div>
</div>
<p>The output shows us that by aligning the remaining sequences to the amplicons, we have retrieved another 42 amplicons for our reference database. The remaining 297 sequences we downloaded, albeit from the 16S gene region, are likely not covering at least 90% of the amplicon region of the 16S gene or are covering a different part of the 16S gene altogether.</p>
</section>
<section id="step-5-dereplicate-the-database">
<h3>3.7 Step 5: dereplicate the database<a class="headerlink" href="#step-5-dereplicate-the-database" title="Permalink to this heading">#</a></h3>
<p>Now that we have retrieved all amplicons from the downloaded sequences, we can remove duplicate sequences for each species, i.e., dereplicating the reference database. This step helps reduce the file size even more, plus is essential for most taxonomic classifiers that restrict the output to N number of best hits (such as the most widely used classifier BLAST).</p>
<p>To dereplicate the database, we can use CRABS’ <code class="docutils literal notranslate"><span class="pre">--dereplicate</span></code> function, which requires an input file name (<code class="docutils literal notranslate"><span class="pre">--input</span></code>) and an output file name (<code class="docutils literal notranslate"><span class="pre">--output</span></code>). By default, CRABS will retain all unique barcodes for each species, though you can change this to only keeping one barcode per species or only unique barcodes regardless if shared between different species using the <code class="docutils literal notranslate"><span class="pre">--dereplication-method</span></code> parameter.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--dereplicate<span class="w"> </span>--input<span class="w"> </span>ncbi_sharks_pga.txt<span class="w"> </span>--output<span class="w"> </span>ncbi_sharks_derep.txt
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
/// CRABS | v1.7.7

|            Function | Dereplicate CRABS database
|         Import data | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|      Exporting data | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|             Results | Written 746 unique sequences to ncbi_sharks_derep.txt out of 3571 initial sequences (20.89%)
</pre></div>
</div>
</div>
<p>The output shows us that a lot of duplicate data was in our database and that we can reduce the reference database by 79.11% without losing any information.</p>
</section>
<section id="step-6-filter-the-reference-database">
<h3>3.8 Step 6: filter the reference database<a class="headerlink" href="#step-6-filter-the-reference-database" title="Permalink to this heading">#</a></h3>
<p>The final curation step for our custom reference database is to clean up the database using a variety of parameters, including minimum length, maximum length, maximum number of ambiguous base calls, environmental sequences, sequences for which the species name is not provided, and sequences with unspecified taxonomic levels. To accomplish this, we can use the <code class="docutils literal notranslate"><span class="pre">--filter</span></code> function, for which we need to provide an input file name (<code class="docutils literal notranslate"><span class="pre">--input</span></code>), an output file name (<code class="docutils literal notranslate"><span class="pre">--output</span></code>), and values for the different parameters.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--filter<span class="w"> </span>--input<span class="w"> </span>ncbi_sharks_derep.txt<span class="w"> </span>--output<span class="w"> </span>ncbi_sharks_clean.txt<span class="w"> </span>--minimum-length<span class="w"> </span><span class="m">150</span><span class="w"> </span>--maximum-length<span class="w"> </span><span class="m">250</span><span class="w"> </span>--environmental<span class="w"> </span>--no-species-id<span class="w"> </span>--maximum-n<span class="w"> </span><span class="m">1</span><span class="w"> </span>--rank-na<span class="w"> </span><span class="m">2</span>
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
/// CRABS | v1.7.7

|            Function | Filter CRABS database
| Included parameters | &quot;--minimum-length&quot;, &quot;--maximum-length&quot;, &quot;--maximum-n&quot;, &quot;--environmental&quot;, &quot;--no-species-id&quot;, &quot;--rank-na&quot;
|         Import data | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|      Exporting data | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|             Results | Written 728 filtered sequences to ncbi_sharks_clean.txt out of 746 initial sequences (97.59%)
|                     | Maximum ambiguous bases filter: 18 sequences not passing filter (2.41%)
</pre></div>
</div>
</div>
<p>The output shows that we only discarded 18 sequences, which contained too many ambiguous bases in the barcode sequence. These ambiguous bases are likely originating from unclean Sanger sequencing outputs, where a high confidence basecall cannot be made.</p>
</section>
<section id="step-7-export-the-reference-database">
<h3>3.9 Step 7: export the reference database<a class="headerlink" href="#step-7-export-the-reference-database" title="Permalink to this heading">#</a></h3>
<p>Now that we have created a custom curated reference database containing 728 shark sequences for the 16S rRNA gene in 6 simple steps, all that is left to do is to export the database in a format that can be used by a taxonomic classifier. As mentioned before, a huge variety of taxonomic classifiers exists, each requiring their own specific file format. Luckily, CRABS is capable of exporting your custom reference database to multiple taxonomic classifiers, including SINTAX, RDP, QIIME, DADA2, IDTAXA, and BLAST.</p>
<p>To show you how this works, we will export the current shark database to SINTAX format, which is a kmer based classifier implemented in USEARCH and VSEARCH. This requires the <code class="docutils literal notranslate"><span class="pre">--export</span></code> function in CRABS and needs an input file name (<code class="docutils literal notranslate"><span class="pre">--input</span></code>), an output file name (<code class="docutils literal notranslate"><span class="pre">--output</span></code>), and the format in which the database needs to be exported (<code class="docutils literal notranslate"><span class="pre">--export-format</span></code>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>crabs<span class="w"> </span>--export<span class="w"> </span>--input<span class="w"> </span>ncbi_sharks_clean.txt<span class="w"> </span>--output<span class="w"> </span>ncbi_sharks_sintax.fasta<span class="w"> </span>--export-format<span class="w"> </span>sintax
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
/// CRABS | v1.7.7

|            Function | Export CRABS database to SINTAX format
|         Import data | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|      Exporting data | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|             Results | Written 728 sequences to ncbi_sharks_sintax.fasta out of 728 initial sequences (100.0%)
</pre></div>
</div>
</div>
<p>We can now have a look at the first 2 lines of the newly-created document to see what the SINTAX format looks like.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">2</span><span class="w"> </span>ncbi_sharks_sintax.fasta
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">MZ562568</span><span class="p">;</span><span class="n">tax</span><span class="o">=</span><span class="n">d</span><span class="p">:</span><span class="n">Metazoa</span><span class="p">,</span><span class="n">p</span><span class="p">:</span><span class="n">Chordata</span><span class="p">,</span><span class="n">c</span><span class="p">:</span><span class="n">Chondrichthyes</span><span class="p">,</span><span class="n">o</span><span class="p">:</span><span class="n">Myliobatiformes</span><span class="p">,</span><span class="n">f</span><span class="p">:</span><span class="n">Dasyatidae</span><span class="p">,</span><span class="n">g</span><span class="p">:</span><span class="n">Fontitrygon</span><span class="p">,</span><span class="n">s</span><span class="p">:</span><span class="n">Fontitrygon_garouaensis</span>
<span class="n">ACTTAAATTATTTCTAACCAAACTCCTACCTCCGGGCATAAACCAAAAAAGAATCTTAATTTAACCTGTTTTTGGTTGGGGCGACCAAGGGGAAAAATAAAACCCCCTTATCGAACGTGTGAACAGTCACTCAAAAATTAGAACTACCACTCTAATTAATAGAAAATCTAACGAATAATGACCCAGGAATACATTTCCTGATCATTGGACCA</span>
</pre></div>
</div>
</div>
<p>The SINTAX format is a 2-line fasta document, with the header on the first line containing the full taxonomic lineage of the barcode, and the sequence on the second line.</p>
<p><strong>That’s all that it takes to generate your own reference database. We are now ready to assign a taxonomy to our ASVs that were generated during the bioinformatic pipeline in the previous section!</strong></p>
</section>
</section>
<section id="the-coi-reference-database">
<h2>4. The COI reference database<a class="headerlink" href="#the-coi-reference-database" title="Permalink to this heading">#</a></h2>
<p>For our tutorial dataset, we will be using a COI reference database I generated earlier using CRABS (<strong>coi_blast</strong>). This database was generated using COI barcodes from NCBI, BOLD, and MIRODI2, as well as supplemented by local barcodes provided by students of this class. This reference database is formatted according to BLAST specifications and holds 2,670,924 high quality, unique, COI reference barcodes. In case you are interested, we have also provided you the unformatted COI reference database in CRABS format (<strong>coi_clean.txt</strong>), if you’d like to export the database in a different format and trial other taxonomic classifiers for this dataset or your own data.</p>
</section>
<section id="blast">
<h2>5. BLAST<a class="headerlink" href="#blast" title="Permalink to this heading">#</a></h2>
<section id="online-blast">
<h3>5.1 Online BLAST<a class="headerlink" href="#online-blast" title="Permalink to this heading">#</a></h3>
<p>The taxonomic classifier we will be using for this tutorial is BLAST (<strong>B</strong>asic <strong>L</strong>ocal <strong>A</strong>lignment <strong>S</strong>earch <strong>T</strong>ool). BLAST is probably the most well-known classifier, as well as the most widely used one. As the name suggests, BLAST assigns a taxonomic ID to a sequence through best hits from local alignments against a reference database. Hence, when thinking about the four basic strategies for taxonomy assignment we introduced earlier, BLAST belongs to the <strong>Sequence Similarity</strong> (SS) category.</p>
<p>The popularity of BLAST likely stems from its direct implementation into the <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">NCBI online reference database repository</a>. To check out the website, we will copy-paste the first 5 sequences in our ASV file (<strong>4.results/asvs.fasta</strong>) and perform a <strong>Nucleotide BLAST</strong>, as shown in the screenshot below.</p>
<figure class="align-default" id="online-blast-search">
<img alt="_images/online_blast_search.png" src="_images/online_blast_search.png" />
<figcaption>
<p><span class="caption-number">Fig. 36 </span><span class="caption-text">: A screenshot of the setup for an online BLAST search</span><a class="headerlink" href="#online-blast-search" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>And below is a screenshot of the BLAST results for ASV_1.</p>
<figure class="align-default" id="online-blast-results">
<img alt="_images/online_blast_results.png" src="_images/online_blast_results.png" />
<figcaption>
<p><span class="caption-number">Fig. 37 </span><span class="caption-text">: A screenshot of the online BLAST results for ASV_1</span><a class="headerlink" href="#online-blast-results" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>We will explore the results on the website more in depth during the workshop, but the most important numbers are provided in the screenshot above, including the ASV ID (<strong>Results for</strong>), the taxonomic ID the sequence best matches to (<strong>Scientific Name</strong>), the percentage of how much the reference sequence provides coverage to the query sequence (<strong>Query Cover</strong>), the expected value is a parameter that describes the number of hits one can “expect” to see by chance when searching a database of a particular size and describes the random background noise (<strong>E value</strong>), and the percent identity value of how well the query and reference match over the covered region (<strong>Per. Ident</strong>).</p>
</section>
<section id="local-blast">
<h3>5.2 Local BLAST<a class="headerlink" href="#local-blast" title="Permalink to this heading">#</a></h3>
<p>While the website is extremely handy, it will be time-consuming if we have to check all 8,314 ASVs in our <strong>4.results/asvs.fasta</strong> file. Plus, besides being time-consuming, the NCBI servers also limit the number of sequences you can BLAST per time and might restrict server access if you are taking up too much bandwidth. Luckily, BLAST is also available via the Command Line Interface, called <a class="reference external" href="https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html">BLAST+</a>, and can be run locally on curated reference databases without restrictions. Furthermore, since CRABS can download data from multiple online repositories, as well as incorporate in house barcodes, running BLAST locally will aid in increased taxonomic assignment accuracy by incorporating a larger number of barcodes.</p>
<p>When we conduct BLAST via the command line (function: <code class="docutils literal notranslate"><span class="pre">blastn</span></code>), we can specify that we want to search against the custom local reference database by specifying the local database name to the <code class="docutils literal notranslate"><span class="pre">-db</span></code> parameter. Our query sequences are provided through the <code class="docutils literal notranslate"><span class="pre">-query</span></code> parameter. We can also set the maximum number of target sequences we get back, the minimum percent identity threshold, and the minimum query coverage threshold through the <code class="docutils literal notranslate"><span class="pre">-max_target_seqs</span></code>, <code class="docutils literal notranslate"><span class="pre">-perc_identity</span></code>, and <code class="docutils literal notranslate"><span class="pre">-qcov_hsp_perc</span></code> parameters. Finally, the output file (parameter <code class="docutils literal notranslate"><span class="pre">-out</span></code>) can take on different formats, but the most common is a tab-delimited structure, which is invoked by <code class="docutils literal notranslate"><span class="pre">-outfmt</span> <span class="pre">&quot;6&quot;</span></code>. Table C1 in <a class="reference external" href="https://scicomp.ethz.ch/public/manual/BLAST/BLAST.pdf">this document</a> provides an overview of all the different options to the BLAST+ search application.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>blastn<span class="w"> </span>-db<span class="w"> </span>coi_blast<span class="w"> </span>-query<span class="w"> </span>../4.results/asvs.fasta<span class="w"> </span>-out<span class="w"> </span>../4.results/blast_results.txt<span class="w"> </span>-max_target_seqs<span class="w"> </span><span class="m">50</span><span class="w"> </span>-perc_identity<span class="w"> </span><span class="m">50</span><span class="w"> </span>-qcov_hsp_perc<span class="w"> </span><span class="m">50</span><span class="w"> </span>-outfmt<span class="w"> </span><span class="s2">&quot;6 qaccver saccver staxid sscinames length pident mismatch qcovs evalue bitscore qstart qend sstart send gapopen&quot;</span>
</pre></div>
</div>
<p>We can inspect the BLAST output using the <code class="docutils literal notranslate"><span class="pre">head</span></code> command.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>../4.results/blast_results.txt
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ASV_1</span>   <span class="n">FJ859351</span>        <span class="mi">296587</span>  <span class="n">Micromonas</span> <span class="n">commoda</span>      <span class="mi">313</span>     <span class="mf">100.000</span> <span class="mi">0</span>       <span class="mi">100</span>     <span class="mf">9.88e-164</span>       <span class="mi">579</span>     <span class="mi">1</span>       <span class="mi">313</span>     <span class="mi">1</span>   <span class="mi">313</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955062</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">312</span>     <span class="mf">100.000</span> <span class="mi">0</span>       <span class="mi">99</span>      <span class="mf">3.55e-163</span>       <span class="mi">577</span>     <span class="mi">2</span>       <span class="mi">313</span>     <span class="mi">1</span>   <span class="mi">312</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955050</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">312</span>     <span class="mf">99.038</span>  <span class="mi">3</span>       <span class="mi">99</span>      <span class="mf">3.58e-158</span>       <span class="mi">560</span>     <span class="mi">2</span>       <span class="mi">313</span>     <span class="mi">1</span>   <span class="mi">312</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955056</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">312</span>     <span class="mf">99.038</span>  <span class="mi">3</span>       <span class="mi">99</span>      <span class="mf">3.58e-158</span>       <span class="mi">560</span>     <span class="mi">2</span>       <span class="mi">313</span>     <span class="mi">1</span>   <span class="mi">312</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955053</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">312</span>     <span class="mf">98.718</span>  <span class="mi">4</span>       <span class="mi">99</span>      <span class="mf">1.66e-156</span>       <span class="mi">555</span>     <span class="mi">2</span>       <span class="mi">313</span>     <span class="mi">1</span>   <span class="mi">312</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955057</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">312</span>     <span class="mf">98.718</span>  <span class="mi">4</span>       <span class="mi">99</span>      <span class="mf">1.66e-156</span>       <span class="mi">555</span>     <span class="mi">2</span>       <span class="mi">313</span>     <span class="mi">1</span>   <span class="mi">312</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955054</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">312</span>     <span class="mf">98.397</span>  <span class="mi">5</span>       <span class="mi">99</span>      <span class="mf">7.74e-155</span>       <span class="mi">549</span>     <span class="mi">2</span>       <span class="mi">313</span>     <span class="mi">1</span>   <span class="mi">312</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955060</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">311</span>     <span class="mf">96.785</span>  <span class="mi">10</span>      <span class="mi">99</span>      <span class="mf">6.07e-146</span>       <span class="mi">520</span>     <span class="mi">3</span>       <span class="mi">313</span>     <span class="mi">2</span>   <span class="mi">312</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955063</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">311</span>     <span class="mf">93.891</span>  <span class="mi">19</span>      <span class="mi">99</span>      <span class="mf">6.20e-131</span>       <span class="mi">470</span>     <span class="mi">2</span>       <span class="mi">312</span>     <span class="mi">1</span>   <span class="mi">311</span>      <span class="mi">0</span>
<span class="n">ASV_1</span>   <span class="n">AY955052</span>        <span class="mi">38833</span>   <span class="n">Micromonas</span> <span class="n">pusilla</span>      <span class="mi">313</span>     <span class="mf">86.581</span>  <span class="mi">38</span>      <span class="mi">99</span>      <span class="mf">1.41e-92</span>        <span class="mi">342</span>     <span class="mi">2</span>       <span class="mi">312</span>     <span class="mi">1</span>   <span class="mi">311</span>      <span class="mi">4</span>
</pre></div>
</div>
</div>
<p>From the output, we can see that we have a tab-delimited text file with the same number of columns provided as the argument we passed to the <code class="docutils literal notranslate"><span class="pre">-outfmt</span></code> parameter for the <code class="docutils literal notranslate"><span class="pre">blastn</span></code> function. The most important columns to note are the first one representing the ASV ID, the second column representing the reference barcode ID, the fourth column representing the species ID of the reference barcode, the sixth column representing the percent similarity of the alignment between our ASV and the reference barcode, and the eight column representing the coverage of the amplicon over which the alignment spans. We can also see that we have a multiple hits per ASV, each above the similarity and coverage threshold of 50 (as specified by <code class="docutils literal notranslate"><span class="pre">-perc_identity</span></code> and <code class="docutils literal notranslate"><span class="pre">-qcov_hsp_perc</span></code>), and with a maximum of 50 hits as specified by the <code class="docutils literal notranslate"><span class="pre">-max_target_seqs</span></code> parameter.</p>
</section>
<section id="blast-output-parsing">
<h3>5.3 BLAST output parsing<a class="headerlink" href="#blast-output-parsing" title="Permalink to this heading">#</a></h3>
<p>One of the last things needed to finalise taxonomy assignment for our ASVs after receiving the BLAST results, is to parse the output file. Currently, we have an output file with potentially mutliple BLAST hits, that could have identical or differing percent identity and query coverage values, as well as be assigned to the same species or different species.</p>
<p>A frequently-used approach to parse BLAST outputs is to calculate the <strong>M</strong>ost <strong>R</strong>ecent <strong>C</strong>ommon <strong>A</strong>ncestor or MRCA. Today, we will be using a small script I have created that takes in a BLAST output file, finds the MRCA among all hits per ASV, creates the taxonomic lineage for the MRCA, and, finally, provides the list of matching species ID’s that was used to create the MRCA. To accomplish this, we will be using the <code class="docutils literal notranslate"><span class="pre">--mrca</span></code> function of ALEX (<strong>A</strong>ncestor <strong>L</strong>ink <strong>EXplorer</strong>), which requires us to provide the blast input file (<code class="docutils literal notranslate"><span class="pre">--blast-input</span></code>), the count table (<code class="docutils literal notranslate"><span class="pre">--table-input</span></code>) so that ALEX can write the MRCA results in the same order and facilitate import into R downstream, as well as the names.dmp (<code class="docutils literal notranslate"><span class="pre">--names-input</span></code>) and nodes.dmp (<code class="docutils literal notranslate"><span class="pre">--nodes-input</span></code>) NCBI taxonomy files to create the taxonomic lineage. Finally, we need to provide an output file name using the <code class="docutils literal notranslate"><span class="pre">--output</span></code> parameter. Those two files are already in our <code class="docutils literal notranslate"><span class="pre">5.refdb</span></code> folder, as we used them for CRABS as well. However, if you want to download them separately, you can do so through ALEX as well via the <code class="docutils literal notranslate"><span class="pre">--download-taxonomy</span></code> function.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>alex<span class="w"> </span>--mrca<span class="w"> </span>--blast-input<span class="w"> </span>../4.results/blast_results.txt<span class="w"> </span>--table-input<span class="w"> </span>../4.results/count_table.txt<span class="w"> </span>--names-input<span class="w"> </span>names.dmp<span class="w"> </span>--nodes-input<span class="w"> </span>nodes.dmp<span class="w"> </span>--output<span class="w"> </span>../4.results/blast_parsed.txt
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
/// ALEX | v0.1.0

|            Function | Calculate MRCA from BLAST result
|       Extracting... | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|       Extracting... | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|       Extracting... | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|       Extracting... | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:03
|       Extracting... | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:04
|       Extracting... | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|       Extracting... | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
|       Extracting... | ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 0:00:00
</pre></div>
</div>
</div>
<p>Finally, let’s use the <code class="docutils literal notranslate"><span class="pre">head</span></code> command once more to inspect the parsed results. We will print out the first 5 ASVs, so that we can compare them to the online BLAST results.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>head<span class="w"> </span>-n<span class="w"> </span><span class="m">10</span><span class="w"> </span>../4.results/blast_parsed.txt
</pre></div>
</div>
<div class="note dropdown admonition">
<p class="admonition-title">Output</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#OTU ID kingdom phylum  class   order   family  genus   species pident  qcov    matching species IDs</span>
<span class="n">ASV_1</span>   <span class="n">Viridiplantae</span>   <span class="n">Chlorophyta</span>     <span class="n">Mamiellophyceae</span> <span class="n">Mamiellales</span>     <span class="n">Mamiellaceae</span>    <span class="n">Micromonas</span>      <span class="n">NA</span>      <span class="mf">100.0</span>   <span class="mi">100</span>     <span class="n">Micromonas</span> <span class="n">commoda</span>
<span class="n">ASV_2</span>   <span class="n">Viridiplantae</span>   <span class="n">Chlorophyta</span>     <span class="n">Mamiellophyceae</span> <span class="n">Mamiellales</span>     <span class="n">Mamiellaceae</span>    <span class="n">Micromonas</span>      <span class="n">Micromonas_pusilla</span>      <span class="mf">100.099</span>      <span class="n">Micromonas</span> <span class="n">pusilla</span>
<span class="n">ASV_3</span>   <span class="n">Metazoa</span> <span class="n">Arthropoda</span>      <span class="n">Insecta</span> <span class="n">Diptera</span> <span class="n">Lonchaeidae</span>     <span class="n">Lonchaea</span>        <span class="n">Lonchaea_fraxina</span>        <span class="mf">78.966</span>  <span class="mi">90</span>      <span class="n">Lonchaea</span> <span class="n">fraxina</span>
<span class="n">ASV_4</span>   <span class="n">Viridiplantae</span>   <span class="n">Chlorophyta</span>     <span class="n">Mamiellophyceae</span> <span class="n">Mamiellales</span>     <span class="n">Mamiellaceae</span>    <span class="n">Micromonas</span>      <span class="n">Micromonas_pusilla</span>      <span class="mf">100.099</span>      <span class="n">Micromonas</span> <span class="n">pusilla</span>
<span class="n">ASV_5</span>   <span class="n">NA</span>      <span class="n">NA</span>      <span class="n">Phaeophyceae</span>    <span class="n">Dictyotales</span>     <span class="n">Dictyotaceae</span>    <span class="n">Canistrocarpus</span>  <span class="n">Canistrocarpus_crispatus</span>        <span class="mf">82.253</span>  <span class="mi">93</span>  <span class="n">Canistrocarpus</span> <span class="n">crispatus</span>
</pre></div>
</div>
</div>
<p><strong>That’s it for creating your own local reference database and taxonomy assignment. Next up is data curation, followed by the generation of phylogenetic trees, and statistical analysis of metabarcoding data!</strong></p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="bioinformaticanalysis.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Bioinformatic analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="datacuration.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Data curation</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">1. Introduction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#online-data-repositories">2. Online data repositories</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#crabs">3. CRABS</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow">3.1 Workflow</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#help-documentation">3.2 Help documentation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-download-data">3.3 Step 1: download data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-import-data">3.4 Step 2: import data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-extract-amplicons-through-in-silico-pcr">3.5 Step 3: extract amplicons through <em>in silico</em> PCR</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-extract-amplicons-through-pairwise-global-alignments">3.6 Step 4: extract amplicons through pairwise global alignments</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-dereplicate-the-database">3.7 Step 5: dereplicate the database</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-filter-the-reference-database">3.8 Step 6: filter the reference database</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-7-export-the-reference-database">3.9 Step 7: export the reference database</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-coi-reference-database">4. The COI reference database</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#blast">5. BLAST</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#online-blast">5.1 Online BLAST</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#local-blast">5.2 Local BLAST</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#blast-output-parsing">5.3 BLAST output parsing</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Gert-Jan Jeunen
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2025-2030.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>